
<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pypath.inputs.biomart &#8212; pypath 0.16.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=77553476" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/css/pypathdocs.css?v=b64fc3cb" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6" />
  <script src="../../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=365ca57ee442770a23c6"></script>

    <script src="../../../_static/documentation_options.js?v=7de89ed4"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pypath/inputs/biomart';</script>
    <link rel="icon" href="../../../_static/omnipath-logo.svg"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">
  

<a class="navbar-brand logo" href="../../../contents.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/omnipath-logo.svg" class="logo__image only-light" alt="Pypath"/>
    <script>document.write(`<img src="../../../_static/omnipath-logo.svg" class="logo__image only-dark" alt="Pypath"/>`);</script>
  
  
    <p class="title logo__title">Pypath</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          
 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">
<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Twitter</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">PyPI</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">OmniPath</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Saez Lab</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">
 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Twitter</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">PyPI</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">OmniPath</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Saez Lab</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">





<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../contents.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../index.html" class="nav-link">Module code</a></li>
    
    
    <li class="breadcrumb-item"><i class="fa-solid fa-ellipsis"></i></li>
    
    
    <li class="breadcrumb-item"><a href="../inputs.html" class="nav-link">pypath.inputs</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pypath.input...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <h1>Source code for pypath.inputs.biomart</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">#</span>
<span class="c1">#  This file is part of the `pypath` python module</span>
<span class="c1">#</span>
<span class="c1">#  Copyright 2014-2023</span>
<span class="c1">#  EMBL, EMBL-EBI, Uniklinik RWTH Aachen, Heidelberg University</span>
<span class="c1">#</span>
<span class="c1">#  Authors: see the file `README.rst`</span>
<span class="c1">#  Contact: Dénes Türei (turei.denes@gmail.com)</span>
<span class="c1">#</span>
<span class="c1">#  Distributed under the GPLv3 License.</span>
<span class="c1">#  See accompanying file LICENSE.txt or copy at</span>
<span class="c1">#      https://www.gnu.org/licenses/gpl-3.0.html</span>
<span class="c1">#</span>
<span class="c1">#  Website: https://pypath.omnipathdb.org/</span>
<span class="c1">#</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">from</span> <span class="nn">future.utils</span> <span class="kn">import</span> <span class="n">iteritems</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">collections</span>

<span class="kn">import</span> <span class="nn">pypath.share.session</span> <span class="k">as</span> <span class="nn">session_mod</span>
<span class="kn">import</span> <span class="nn">pypath.share.common</span> <span class="k">as</span> <span class="nn">common</span>
<span class="kn">import</span> <span class="nn">pypath.share.curl</span> <span class="k">as</span> <span class="nn">curl</span>
<span class="kn">import</span> <span class="nn">pypath.resources.urls</span> <span class="k">as</span> <span class="nn">urls</span>
<span class="kn">import</span> <span class="nn">pypath.utils.taxonomy</span> <span class="k">as</span> <span class="nn">taxonomy</span>

<span class="n">_logger</span> <span class="o">=</span> <span class="n">session_mod</span><span class="o">.</span><span class="n">Logger</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;biomart_input&#39;</span><span class="p">)</span>


<span class="c1"># for mouse homologues: Filter name = &quot;with_mmusculus_homolog&quot;</span>
<span class="n">_filter_xml_template</span> <span class="o">=</span> <span class="s1">&#39;&lt;Filter name=&quot;</span><span class="si">%s</span><span class="s1">&quot; excluded=&quot;0&quot;/&gt;&#39;</span>
<span class="n">_attr_xml_template</span> <span class="o">=</span> <span class="s1">&#39;&lt;Attribute name=&quot;</span><span class="si">%s</span><span class="s1">&quot; /&gt;&#39;</span>


<div class="viewcode-block" id="biomart_query">
<a class="viewcode-back" href="../../../api/pypath.inputs.biomart.biomart_query.html#pypath.inputs.biomart.biomart_query">[docs]</a>
<span class="k">def</span> <span class="nf">biomart_query</span><span class="p">(</span>
        <span class="n">attrs</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">filters</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">transcript</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">peptide</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">gene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">dataset</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;hsapiens_gene_ensembl&#39;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Query the Ensembl Biomart web service.</span>
<span class="sd">    Use https://www.ensembl.org/biomart/martview/ to check for attribute</span>
<span class="sd">    and dataset names.</span>

<span class="sd">    Args</span>
<span class="sd">        attrs:</span>
<span class="sd">            One or more Ensembl attribute names.</span>
<span class="sd">        filters:</span>
<span class="sd">            One or more Ensembl filter names.</span>
<span class="sd">        transcript:</span>
<span class="sd">            Include Ensembl transcript IDs in the result.</span>
<span class="sd">        peptide:</span>
<span class="sd">            Include Ensembl peptide IDs in the result.</span>
<span class="sd">        gene:</span>
<span class="sd">            Include Ensembl gene IDs in the result.</span>
<span class="sd">        dataset:</span>
<span class="sd">            An Ensembl dataset name.</span>

<span class="sd">    Yields:</span>
<span class="sd">        Named tuples with the requested attributes for each record returned</span>
<span class="sd">        by Ensembl Biomart.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">_attrs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="n">gene</span><span class="p">:</span>

        <span class="n">_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;ensembl_gene_id&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">transcript</span><span class="p">:</span>

        <span class="n">_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;ensembl_transcript_id&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">peptide</span><span class="p">:</span>

        <span class="n">_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;ensembl_peptide_id&#39;</span><span class="p">)</span>

    <span class="n">_attrs</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">attrs</span><span class="p">))</span>
    <span class="n">filters</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">filters</span><span class="p">)</span>

    <span class="n">record</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;EnsemblRecord&#39;</span><span class="p">,</span>
        <span class="n">_attrs</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">_logger</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
        <span class="s1">&#39;Downloading data from Ensembl Biomart: &#39;</span>
        <span class="s1">&#39;dataset=`</span><span class="si">%s</span><span class="s1">`, &#39;</span>
        <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">&#39;</span>
        <span class="s1">&#39;attributes=`</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="p">(</span>
            <span class="n">dataset</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="s1">&#39;filters=`</span><span class="si">%s</span><span class="s1">`, &#39;</span> <span class="o">%</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filters</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">filters</span> <span class="k">else</span>
                <span class="s1">&#39;&#39;</span>
            <span class="p">),</span>
            <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">_attrs</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">rewsp</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\n\s+&#39;</span><span class="p">)</span>

    <span class="n">xml_template_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">common</span><span class="o">.</span><span class="n">DATA</span><span class="p">,</span> <span class="s1">&#39;ensembl_biomart_query.xml&#39;</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">xml_template_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>

        <span class="n">xml_template</span> <span class="o">=</span> <span class="n">fp</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

    <span class="n">filter_part</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">_filter_xml_template</span> <span class="o">%</span> <span class="n">_filter</span>
        <span class="k">for</span> <span class="n">_filter</span> <span class="ow">in</span> <span class="n">filters</span>
    <span class="p">)</span>
    <span class="n">attr_part</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">_attr_xml_template</span> <span class="o">%</span> <span class="n">_attr</span>
        <span class="k">for</span> <span class="n">_attr</span> <span class="ow">in</span> <span class="n">_attrs</span>
    <span class="p">)</span>

    <span class="n">xml_query</span> <span class="o">=</span> <span class="n">xml_template</span> <span class="o">%</span> <span class="p">(</span>
        <span class="n">dataset</span><span class="p">,</span>
        <span class="n">filter_part</span><span class="p">,</span>
        <span class="n">attr_part</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">xml_query</span> <span class="o">=</span> <span class="n">rewsp</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">xml_query</span><span class="p">)</span>

    <span class="n">biomart_url</span> <span class="o">=</span> <span class="n">urls</span><span class="o">.</span><span class="n">urls</span><span class="p">[</span><span class="s1">&#39;ensembl&#39;</span><span class="p">][</span><span class="s1">&#39;biomart_url&#39;</span><span class="p">]</span> <span class="o">%</span> <span class="n">xml_query</span>

    <span class="n">c</span> <span class="o">=</span> <span class="n">curl</span><span class="o">.</span><span class="n">Curl</span><span class="p">(</span><span class="n">biomart_url</span><span class="p">,</span> <span class="n">large</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

    <span class="n">success</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">result</span><span class="p">:</span>

        <span class="n">_line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\r</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">success</span> <span class="o">=</span> <span class="n">success</span> <span class="ow">or</span> <span class="n">_line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;[success]&#39;</span>

        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">_line</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">_fields</span><span class="p">):</span>

            <span class="k">yield</span> <span class="n">record</span><span class="p">(</span><span class="o">*</span><span class="n">_line</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">success</span><span class="p">:</span>

        <span class="n">_logger</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Error: Interrupted transfer while downlading data &#39;</span>
            <span class="s1">&#39;from Ensembl Biomart (missing `success` tag).&#39;</span>
        <span class="p">)</span></div>



<div class="viewcode-block" id="biomart_homology">
<a class="viewcode-back" href="../../../api/pypath.inputs.biomart.biomart_homology.html#pypath.inputs.biomart.biomart_homology">[docs]</a>
<span class="k">def</span> <span class="nf">biomart_homology</span><span class="p">(</span>
        <span class="n">source_organism</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">target_organism</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">10090</span><span class="p">,</span>
        <span class="n">extra_fields</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;external_gene_name&#39;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves orthology data from Ensembl Biomart.</span>

<span class="sd">    Returns</span>
<span class="sd">        List of named tuples filtered to genes of the source organism having</span>
<span class="sd">        orthologues in the target organism, with homology related fields.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">ensure_organism</span><span class="p">(</span><span class="n">organism</span><span class="p">):</span>

        <span class="n">organism_ensembl</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">organism_ensembl</span><span class="p">:</span>

            <span class="n">msg</span> <span class="o">=</span> <span class="s1">&#39;Could not find Ensembl taxon ID for `</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>
            <span class="n">_log</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">organism_ensembl</span>


    <span class="n">source_organism</span> <span class="o">=</span> <span class="n">ensure_organism</span><span class="p">(</span><span class="n">source_organism</span><span class="p">)</span>
    <span class="n">target_organism</span> <span class="o">=</span> <span class="n">ensure_organism</span><span class="p">(</span><span class="n">target_organism</span><span class="p">)</span>

    <span class="n">homolog_attrs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;homolog_ensembl_peptide&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_ensembl_gene&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_orthology_type&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_orthology_confidence&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_canonical_transcript_protein&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_associated_gene_name&#39;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="n">homolog_attrs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">target_organism</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">attr</span> <span class="ow">in</span> <span class="n">homolog_attrs</span>
    <span class="p">]</span> <span class="o">+</span> <span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">extra_fields</span><span class="p">)</span>

    <span class="n">filters</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;with_</span><span class="si">%s</span><span class="s1">_homolog&#39;</span> <span class="o">%</span> <span class="n">target_organism</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span>
        <span class="n">biomart_query</span><span class="p">(</span>
            <span class="n">attrs</span> <span class="o">=</span> <span class="n">homolog_attrs</span><span class="p">,</span>
            <span class="n">filters</span> <span class="o">=</span> <span class="n">filters</span><span class="p">,</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">peptide</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_gene_ensembl&#39;</span> <span class="o">%</span> <span class="n">source_organism</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span></div>




<div class="viewcode-block" id="biomart_microarray_types">
<a class="viewcode-back" href="../../../api/pypath.inputs.biomart.biomart_microarray_types.html#pypath.inputs.biomart.biomart_microarray_types">[docs]</a>
<span class="k">def</span> <span class="nf">biomart_microarray_types</span><span class="p">(</span><span class="n">organism</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves a list of available microarray types for an organism.</span>

<span class="sd">    Args</span>
<span class="sd">        organism:</span>
<span class="sd">            Name or ID of an organism.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">organism</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>

    <span class="n">url</span> <span class="o">=</span> <span class="n">urls</span><span class="o">.</span><span class="n">urls</span><span class="p">[</span><span class="s1">&#39;ensembl&#39;</span><span class="p">][</span><span class="s1">&#39;arraytypes&#39;</span><span class="p">]</span> <span class="o">%</span> <span class="n">organism</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">curl</span><span class="o">.</span><span class="n">Curl</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">c</span><span class="o">.</span><span class="n">result</span><span class="p">)</span>

    <span class="n">_</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">r</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span>
                <span class="n">r</span><span class="p">[</span><span class="s1">&#39;vendor&#39;</span><span class="p">],</span>
                <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;[-_]&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;array&#39;</span><span class="p">]),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">result</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">result</span></div>



<div class="viewcode-block" id="biomart_microarray">
<a class="viewcode-back" href="../../../api/pypath.inputs.biomart.biomart_microarray.html#pypath.inputs.biomart.biomart_microarray">[docs]</a>
<span class="k">def</span> <span class="nf">biomart_microarray</span><span class="p">(</span>
        <span class="n">array_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">gene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">transcript</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">peptide</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">organism</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">9606</span>
    <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Microarray probe identifier mappings.</span>

<span class="sd">    Args</span>
<span class="sd">        array_type:</span>
<span class="sd">            The microarray model, as shown on the BioMart</span>
<span class="sd">            webpage, or the corresponding code. For a full list of available</span>
<span class="sd">            identifiers see the ``biomart_microarray_types``.</span>
<span class="sd">        gene:</span>
<span class="sd">            Include the mapping to Ensembl gene IDs.</span>
<span class="sd">        transcript:</span>
<span class="sd">            Include the mapping to Ensembl transcript IDs.</span>
<span class="sd">        peptide:</span>
<span class="sd">            Include the mapping to Ensembl peptide IDs.</span>
<span class="sd">        organism:</span>
<span class="sd">            Name or ID of an organism.</span>

<span class="sd">    Returns</span>
<span class="sd">        A dictionary with Ensembl gene, transcript and peptide IDs as keys</span>
<span class="sd">        and sets of microarray probe IDs as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">organism</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>
    <span class="n">array_types</span> <span class="o">=</span> <span class="n">biomart_microarray_types</span><span class="p">(</span><span class="n">organism</span> <span class="o">=</span> <span class="n">organism</span><span class="p">)</span>
    <span class="n">array_types</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">at</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">at</span> <span class="ow">in</span> <span class="n">array_types</span>
    <span class="p">}</span>

    <span class="n">_array_type</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">array_type</span><span class="o">.</span>
        <span class="n">lower</span><span class="p">()</span><span class="o">.</span>
        <span class="n">replace</span><span class="p">(</span><span class="s1">&#39;probe&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span>
        <span class="n">strip</span><span class="p">()</span><span class="o">.</span>
        <span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">_array_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">array_types</span><span class="p">:</span>

        <span class="n">msg</span> <span class="o">=</span> <span class="s1">&#39;No such array type in Ensembl BioMart: `</span><span class="si">%s</span><span class="s1">` (</span><span class="si">%s</span><span class="s1">).&#39;</span> <span class="o">%</span> <span class="p">(</span>
            <span class="n">array_type</span><span class="p">,</span>
            <span class="n">_array_type</span>
        <span class="p">)</span>
        <span class="n">_logger</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">attrs</span> <span class="o">=</span> <span class="p">[</span><span class="n">_array_type</span><span class="p">]</span>
    <span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_gene_ensembl&#39;</span> <span class="o">%</span> <span class="n">organism</span>

    <span class="n">biomart_result</span> <span class="o">=</span> <span class="n">biomart_query</span><span class="p">(</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="n">attrs</span><span class="p">,</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="n">transcript</span><span class="p">,</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span>
        <span class="n">peptide</span> <span class="o">=</span> <span class="n">peptide</span><span class="p">,</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
    <span class="n">_locals</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span>
    <span class="n">ensembl_attrs</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span>
        <span class="n">attr</span>
        <span class="k">for</span> <span class="n">attr</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;transcript&#39;</span><span class="p">,</span> <span class="s1">&#39;peptide&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">_locals</span><span class="p">[</span><span class="n">attr</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">biomart_result</span><span class="p">:</span>

        <span class="n">array_probe_id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">_array_type</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">array_probe_id</span><span class="p">:</span>

            <span class="k">for</span> <span class="n">ensembl_attr</span> <span class="ow">in</span> <span class="n">ensembl_attrs</span><span class="p">:</span>

                <span class="n">ensembl_id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="s1">&#39;ensembl_</span><span class="si">%s</span><span class="s1">_id&#39;</span> <span class="o">%</span> <span class="n">ensembl_attr</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">ensembl_id</span><span class="p">:</span>

                    <span class="n">result</span><span class="p">[</span><span class="n">ensembl_id</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">array_probe_id</span><span class="p">)</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>



<div class="viewcode-block" id="biomart_microarrays">
<a class="viewcode-back" href="../../../api/pypath.inputs.biomart.biomart_microarrays.html#pypath.inputs.biomart.biomart_microarrays">[docs]</a>
<span class="k">def</span> <span class="nf">biomart_microarrays</span><span class="p">(</span>
        <span class="n">organism</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">vendor</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">gene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">transcript</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">peptide</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Microarray probe identifier mappings for multiple microarrays.</span>
<span class="sd">    Retrieves probe mappings for all array types for one organism,</span>
<span class="sd">    optionally limited to one or more array vendors. Note: depending</span>
<span class="sd">    on the number of array models, it can take minutes to download</span>
<span class="sd">    the data.</span>

<span class="sd">    Args</span>
<span class="sd">        organism:</span>
<span class="sd">            Name or ID of an organism.</span>
<span class="sd">        vendor:</span>
<span class="sd">            One or more vendors. None means all vendors. For</span>
<span class="sd">            human, possible values are AFFY, ILLUMINA, AGILENT, CODELINK</span>
<span class="sd">            and PHALANX.</span>
<span class="sd">        gene:</span>
<span class="sd">            Include the mapping to Ensembl gene IDs.</span>
<span class="sd">        transcript:</span>
<span class="sd">            Include the mapping to Ensembl transcript IDs.</span>
<span class="sd">        peptide:</span>
<span class="sd">            Include the mapping to Ensembl peptide IDs.</span>

<span class="sd">    Returns</span>
<span class="sd">        A dictionary with Ensembl gene, transcript and peptide IDs as keys</span>
<span class="sd">        and sets of tuples with microarray types and probe IDs as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">record</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;Probe&#39;</span><span class="p">,</span>
        <span class="p">(</span>
            <span class="s1">&#39;array&#39;</span><span class="p">,</span>
            <span class="s1">&#39;probe&#39;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">array_types</span> <span class="o">=</span> <span class="n">biomart_microarray_types</span><span class="p">(</span><span class="n">organism</span> <span class="o">=</span> <span class="n">organism</span><span class="p">)</span>
    <span class="n">vendor</span> <span class="o">=</span> <span class="p">{</span><span class="n">v</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">common</span><span class="o">.</span><span class="n">to_set</span><span class="p">(</span><span class="n">vendor</span><span class="p">)}</span>

    <span class="n">result</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">at</span> <span class="ow">in</span> <span class="n">array_types</span><span class="p">:</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">vendor</span> <span class="ow">or</span> <span class="n">at</span><span class="p">[</span><span class="s1">&#39;vendor&#39;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">vendor</span><span class="p">:</span>

            <span class="n">probe_map</span> <span class="o">=</span> <span class="n">biomart_microarray</span><span class="p">(</span>
                <span class="n">array_type</span> <span class="o">=</span> <span class="n">at</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">],</span>
                <span class="n">organism</span> <span class="o">=</span> <span class="n">organism</span><span class="p">,</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span>
                <span class="n">transcript</span> <span class="o">=</span> <span class="n">transcript</span><span class="p">,</span>
                <span class="n">peptide</span> <span class="o">=</span> <span class="n">peptide</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">probe_ids</span> <span class="ow">in</span> <span class="n">iteritems</span><span class="p">(</span><span class="n">probe_map</span><span class="p">):</span>

                <span class="k">for</span> <span class="n">probe_id</span> <span class="ow">in</span> <span class="n">probe_ids</span><span class="p">:</span>

                    <span class="n">result</span><span class="p">[</span><span class="n">gene_id</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
                        <span class="n">record</span><span class="p">(</span>
                            <span class="n">array</span> <span class="o">=</span> <span class="n">at</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">),</span>
                            <span class="n">probe</span> <span class="o">=</span> <span class="n">probe_id</span><span class="p">,</span>
                        <span class="p">)</span>
                    <span class="p">)</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../index.html">Welcome to pypath!</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../webservice.html">Webservice</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../releasehistory.html">Release history</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../reference.html">Reference</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../notebooks/manual.html">The pypath book</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">
  <p class="copyright">
    
      © Copyright 2023, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Melih Darcan, Erva Ulusoy, Tennur Kilic, Omer Kaan Vural, Ahmet Rifaioglu, Charles Tapley Hoyt.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">
  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item"><p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.14.3.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>