
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pypath.utils.orthology &#8212; pypath 0.16.10 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=77553476" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/css/pypathdocs.css?v=b64fc3cb" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../../../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../../../_static/documentation_options.js?v=9e32664a"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pypath/utils/orthology';</script>
    <link rel="icon" href="../../../_static/omnipath-logo.svg"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../contents.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/omnipath-logo.svg" class="logo__image only-light" alt="Pypath"/>
    <script>document.write(`<img src="../../../_static/omnipath-logo.svg" class="logo__image only-dark" alt="Pypath"/>`);</script>
  
  
    <p class="title logo__title">Pypath</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Twitter</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">PyPI</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">OmniPath</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Saez Lab</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Twitter</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">PyPI</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">OmniPath</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Saez Lab</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../contents.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../index.html" class="nav-link">Module code</a></li>
    
    
    <li class="breadcrumb-item"><a href="../../pypath.html" class="nav-link">pypath</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pypath.utils...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pypath.utils.orthology</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">#</span>
<span class="c1">#  This file is part of the `pypath` python module</span>
<span class="c1">#</span>
<span class="c1">#  Copyright 2014-2023</span>
<span class="c1">#  EMBL, EMBL-EBI, Uniklinik RWTH Aachen, Heidelberg University</span>
<span class="c1">#</span>
<span class="c1">#  Authors: see the file `README.rst`</span>
<span class="c1">#  Contact: Dénes Türei (turei.denes@gmail.com)</span>
<span class="c1">#</span>
<span class="c1">#  Distributed under the GPLv3 License.</span>
<span class="c1">#  See accompanying file LICENSE.txt or copy at</span>
<span class="c1">#      https://www.gnu.org/licenses/gpl-3.0.html</span>
<span class="c1">#</span>
<span class="c1">#  Website: https://pypath.omnipathdb.org/</span>
<span class="c1">#</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">from</span> <span class="nn">future.utils</span> <span class="kn">import</span> <span class="n">iteritems</span>
<span class="kn">from</span> <span class="nn">past.builtins</span> <span class="kn">import</span> <span class="n">xrange</span><span class="p">,</span> <span class="nb">range</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Literal</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">functools</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">importlib</span> <span class="k">as</span> <span class="nn">imp</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">import</span> <span class="nn">abc</span>
<span class="kn">import</span> <span class="nn">inspect</span>
<span class="kn">import</span> <span class="nn">types</span> <span class="k">as</span> <span class="nn">_types</span>

<span class="kn">import</span> <span class="nn">timeloop</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">import</span> <span class="nn">pypath.utils.mapping</span> <span class="k">as</span> <span class="nn">mapping</span>
<span class="kn">import</span> <span class="nn">pypath.share.common</span> <span class="k">as</span> <span class="nn">common</span>
<span class="kn">import</span> <span class="nn">pypath_common._constants</span> <span class="k">as</span> <span class="nn">_const</span>
<span class="kn">import</span> <span class="nn">pypath.internals.intera</span> <span class="k">as</span> <span class="nn">intera</span>
<span class="kn">import</span> <span class="nn">pypath.resources.urls</span> <span class="k">as</span> <span class="nn">urls</span>
<span class="kn">import</span> <span class="nn">pypath.share.curl</span> <span class="k">as</span> <span class="nn">curl</span>
<span class="kn">import</span> <span class="nn">pypath.inputs.uniprot</span> <span class="k">as</span> <span class="nn">uniprot_input</span>
<span class="kn">import</span> <span class="nn">pypath.inputs.uniprot_db</span> <span class="k">as</span> <span class="nn">uniprot_db</span>
<span class="kn">import</span> <span class="nn">pypath.inputs.homologene</span> <span class="k">as</span> <span class="nn">homologene_input</span>
<span class="kn">import</span> <span class="nn">pypath.inputs.oma</span> <span class="k">as</span> <span class="nn">oma_input</span>
<span class="kn">import</span> <span class="nn">pypath.inputs.biomart</span> <span class="k">as</span> <span class="nn">biomart</span>
<span class="kn">import</span> <span class="nn">pypath.utils.seq</span> <span class="k">as</span> <span class="nn">_se</span>
<span class="kn">import</span> <span class="nn">pypath.share.session</span> <span class="k">as</span> <span class="nn">session</span>
<span class="kn">import</span> <span class="nn">pypath.share.settings</span> <span class="k">as</span> <span class="nn">settings</span>
<span class="kn">import</span> <span class="nn">pypath.utils.taxonomy</span> <span class="k">as</span> <span class="nn">taxonomy</span>
<span class="kn">import</span> <span class="nn">pypath.share.cache</span> <span class="k">as</span> <span class="nn">cache_mod</span>

<span class="n">_orthology_cleanup_timeloop</span> <span class="o">=</span> <span class="n">timeloop</span><span class="o">.</span><span class="n">Timeloop</span><span class="p">()</span>
<span class="n">_orthology_cleanup_timeloop</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="mi">9999</span><span class="p">)</span>

<span class="n">_logger</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;orthology&#39;</span><span class="p">)</span>
<span class="n">_log</span> <span class="o">=</span> <span class="n">_logger</span><span class="o">.</span><span class="n">_log</span>


<div class="viewcode-block" id="OrthologMeta">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OrthologMeta.html#pypath.utils.orthology.OrthologMeta">[docs]</a>
<span class="k">class</span> <span class="nc">OrthologMeta</span><span class="p">(</span><span class="n">abc</span><span class="o">.</span><span class="n">ABCMeta</span><span class="p">):</span>


    <span class="k">def</span> <span class="fm">__new__</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span>
            <span class="n">name</span><span class="p">,</span>
            <span class="n">supercls</span><span class="p">,</span>
            <span class="n">attrs</span><span class="p">,</span>
            <span class="n">resource</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">fields</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;id&#39;</span><span class="p">,),</span>
        <span class="p">):</span>

        <span class="n">Base</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">Base&#39;</span><span class="p">,</span> <span class="n">fields</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">new_method</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

            <span class="k">return</span> <span class="n">Base</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">sig</span> <span class="o">=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">signature</span><span class="p">(</span><span class="n">new_method</span><span class="p">)</span>
        <span class="n">new_params</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">inspect</span><span class="o">.</span><span class="n">Parameter</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="n">inspect</span><span class="o">.</span><span class="n">Parameter</span><span class="o">.</span><span class="n">POSITIONAL_OR_KEYWORD</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">fields</span>
        <span class="p">]</span>
        <span class="n">new_method</span><span class="o">.</span><span class="n">__signature__</span> <span class="o">=</span> <span class="n">sig</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
            <span class="n">parameters</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span>
                <span class="p">(</span><span class="n">sig</span><span class="o">.</span><span class="n">parameters</span><span class="p">[</span><span class="s1">&#39;cls&#39;</span><span class="p">],),</span>
                <span class="n">new_params</span><span class="p">,</span>
            <span class="p">))</span>
        <span class="p">)</span>

        <span class="n">clazz</span> <span class="o">=</span> <span class="n">_types</span><span class="o">.</span><span class="n">new_class</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">(</span><span class="n">Base</span><span class="p">,))</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="n">clazz</span><span class="p">,</span> <span class="s1">&#39;__new__&#39;</span><span class="p">,</span> <span class="n">new_method</span><span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="n">clazz</span><span class="p">,</span> <span class="s1">&#39;_resource&#39;</span><span class="p">,</span> <span class="n">resource</span><span class="p">)</span>
        <span class="n">clazz</span><span class="o">.</span><span class="vm">__module__</span> <span class="o">=</span> <span class="vm">__name__</span>

        <span class="k">return</span> <span class="n">clazz</span></div>



<div class="viewcode-block" id="OrthologBase">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OrthologBase.html#pypath.utils.orthology.OrthologBase">[docs]</a>
<span class="k">class</span> <span class="nc">OrthologBase</span><span class="p">:</span>


    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span>


    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;&lt;Ortholog </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1"> (</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_resource</span><span class="si">}</span><span class="s1">)&gt;&#39;</span>


    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="fm">__str__</span><span class="p">()</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="fm">__str__</span><span class="p">()</span>


    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="fm">__hash__</span><span class="p">()</span></div>



<div class="viewcode-block" id="OmaOrtholog">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OmaOrtholog.html#pypath.utils.orthology.OmaOrtholog">[docs]</a>
<span class="k">class</span> <span class="nc">OmaOrtholog</span><span class="p">(</span></div>

        <span class="n">OrthologBase</span><span class="p">,</span>
        <span class="n">metaclass</span> <span class="o">=</span> <span class="n">OrthologMeta</span><span class="p">,</span>
        <span class="n">resource</span> <span class="o">=</span> <span class="s1">&#39;OMA&#39;</span><span class="p">,</span>
        <span class="n">fields</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;rel_type&#39;</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">),</span>
    <span class="p">):</span> <span class="k">pass</span>



<div class="viewcode-block" id="EnsemblOrtholog">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.EnsemblOrtholog.html#pypath.utils.orthology.EnsemblOrtholog">[docs]</a>
<span class="k">class</span> <span class="nc">EnsemblOrtholog</span><span class="p">(</span></div>

        <span class="n">OrthologBase</span><span class="p">,</span>
        <span class="n">metaclass</span> <span class="o">=</span> <span class="n">OrthologMeta</span><span class="p">,</span>
        <span class="n">resource</span> <span class="o">=</span> <span class="s1">&#39;Ensembl&#39;</span><span class="p">,</span>
        <span class="n">fields</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;types&#39;</span><span class="p">,</span> <span class="s1">&#39;hc&#39;</span><span class="p">),</span>
    <span class="p">):</span> <span class="k">pass</span>


<div class="viewcode-block" id="HomologeneOrtholog">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.HomologeneOrtholog.html#pypath.utils.orthology.HomologeneOrtholog">[docs]</a>
<span class="k">class</span> <span class="nc">HomologeneOrtholog</span><span class="p">(</span></div>

        <span class="n">OrthologBase</span><span class="p">,</span>
        <span class="n">metaclass</span> <span class="o">=</span> <span class="n">OrthologMeta</span><span class="p">,</span>
        <span class="n">resource</span> <span class="o">=</span> <span class="s1">&#39;HomoloGene&#39;</span><span class="p">,</span>
    <span class="p">):</span> <span class="k">pass</span>


<span class="n">OrthologyTableKey</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span>
    <span class="s1">&#39;OrthologyTableKey&#39;</span><span class="p">,</span>
    <span class="p">(</span><span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;only_swissprot&#39;</span><span class="p">,</span> <span class="s1">&#39;resource&#39;</span><span class="p">,</span> <span class="s1">&#39;id_type&#39;</span><span class="p">),</span>
<span class="p">)</span>

<div class="viewcode-block" id="OrthologyManager">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OrthologyManager.html#pypath.utils.orthology.OrthologyManager">[docs]</a>
<span class="k">class</span> <span class="nc">OrthologyManager</span><span class="p">(</span><span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="p">):</span>

    <span class="n">TRANSLATION_PARAM</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s1">&#39;oma&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homologene&#39;</span><span class="p">,</span>
        <span class="s1">&#39;oma_rel_type&#39;</span><span class="p">,</span>
        <span class="s1">&#39;oma_score&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ensembl&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ensembl_hc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ensembl_types&#39;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">RESOURCE_PARAM</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;oma&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;rel_type&#39;</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">),</span>
        <span class="s1">&#39;ensembl&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;hc&#39;</span><span class="p">,</span> <span class="s1">&#39;types&#39;</span><span class="p">),</span>
        <span class="s1">&#39;homologene&#39;</span><span class="p">:</span> <span class="p">(),</span>
    <span class="p">}</span>

<div class="viewcode-block" id="OrthologyManager.__init__">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OrthologyManager.html#pypath.utils.orthology.OrthologyManager.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">cleanup_period</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
            <span class="n">lifetime</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">):</span>

        <span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;orthology&#39;</span><span class="p">)</span>


        <span class="nd">@_orthology_cleanup_timeloop</span><span class="o">.</span><span class="n">job</span><span class="p">(</span>
            <span class="n">interval</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">timedelta</span><span class="p">(</span>
                <span class="n">seconds</span> <span class="o">=</span> <span class="n">cleanup_period</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">def</span> <span class="nf">_cleanup</span><span class="p">():</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_remove_expired</span><span class="p">()</span>


        <span class="n">_orthology_cleanup_timeloop</span><span class="o">.</span><span class="n">start</span><span class="p">(</span><span class="n">block</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">lifetime</span> <span class="o">=</span> <span class="n">lifetime</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tables</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">expiry</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_param</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">TRANSLATION_PARAM</span><span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span><span class="s1">&#39;OrthologyManager has been created.&#39;</span><span class="p">)</span></div>



    <span class="k">def</span> <span class="nf">reload</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="n">modname</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__module__</span>
        <span class="n">mod</span> <span class="o">=</span> <span class="nb">__import__</span><span class="p">(</span><span class="n">modname</span><span class="p">,</span> <span class="n">fromlist</span><span class="o">=</span><span class="p">[</span><span class="n">modname</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]])</span>
        <span class="n">imp</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
        <span class="n">new</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mod</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;__class__&#39;</span><span class="p">,</span> <span class="n">new</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">which_table</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">resource</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;oma&#39;</span><span class="p">,</span> <span class="s1">&#39;homologene&#39;</span><span class="p">,</span> <span class="s1">&#39;ensembl&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;oma&#39;</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="p">):</span>

        <span class="n">loc</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span>
        <span class="n">key</span> <span class="o">=</span> <span class="n">OrthologyTableKey</span><span class="p">(</span><span class="o">**</span><span class="p">{</span>
            <span class="n">f</span><span class="p">:</span> <span class="n">loc</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">OrthologyTableKey</span><span class="o">.</span><span class="n">_fields</span>
        <span class="p">})</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">expiry</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">:</span>

            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>


    <span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">globals</span><span class="p">()[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">key</span><span class="o">.</span><span class="n">resource</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">Orthology&#39;</span><span class="p">](</span>
            <span class="n">target</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">only_swissprot</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
        <span class="p">)</span>


<div class="viewcode-block" id="OrthologyManager.translate">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OrthologyManager.html#pypath.utils.orthology.OrthologyManager.translate">[docs]</a>
    <span class="nd">@common</span><span class="o">.</span><span class="n">ignore_unhashable</span>
    <span class="nd">@functools</span><span class="o">.</span><span class="n">lru_cache</span><span class="p">(</span><span class="n">maxsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">1e5</span><span class="p">))</span>
    <span class="k">def</span> <span class="nf">translate</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">identifiers</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">oma</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">oma_rel_type</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">oma_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translate one or more identifiers by orthologous gene pairs.</span>

<span class="sd">        Args:</span>
<span class="sd">            identifiers:</span>
<span class="sd">                One or more identifers of the source organism, of ID type</span>
<span class="sd">                `id_type`.</span>
<span class="sd">            target:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">            source:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">            id_type:</span>
<span class="sd">                The identifier type to use.</span>
<span class="sd">            only_swissprot:</span>
<span class="sd">                Use only SwissProt IDs.</span>
<span class="sd">            oma</span>
<span class="sd">                Use orthology information from the Orthologous Matrix (OMA).</span>
<span class="sd">                Currently this is the recommended source for orthology data.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use orthology information from NCBI HomoloGene.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use orthology information from Ensembl.</span>
<span class="sd">            oma_rel_type:</span>
<span class="sd">                Restrict relations to certain types.</span>
<span class="sd">            oma_score:</span>
<span class="sd">                Lower threshold for similarity metric.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only the high confidence orthology relations from Ensembl.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                Ensembl orthology relation types to use. Possible values are</span>
<span class="sd">                `one2one`, `one2many` and `many2many`. By default only</span>
<span class="sd">                `one2one` is used.</span>
<span class="sd">            full_records:</span>
<span class="sd">                Include not only the identifiers, but also some properties of</span>
<span class="sd">                the orthology relationships.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Set of identifiers of orthologous genes or proteins in the</span>
<span class="sd">            target taxon.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="n">param</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_translation_param</span><span class="p">(</span><span class="nb">locals</span><span class="p">())</span>
        <span class="n">proc</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">full_records</span> <span class="k">else</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">resource</span><span class="p">,</span> <span class="n">keys</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">RESOURCE_PARAM</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">param</span><span class="p">[</span><span class="n">resource</span><span class="p">]:</span>

                <span class="k">continue</span>

            <span class="n">table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">which_table</span><span class="p">(</span>
                <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
                <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
                <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
                <span class="n">resource</span> <span class="o">=</span> <span class="n">resource</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">result</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="n">table</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
                    <span class="n">identifiers</span><span class="p">,</span>
                    <span class="n">full_records</span> <span class="o">=</span> <span class="n">full_records</span><span class="p">,</span>
                    <span class="o">**</span><span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">param</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">keys</span><span class="p">},</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span></div>



<div class="viewcode-block" id="OrthologyManager.get_dict">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OrthologyManager.html#pypath.utils.orthology.OrthologyManager.get_dict">[docs]</a>
    <span class="k">def</span> <span class="nf">get_dict</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">oma</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">oma_rel_type</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">oma_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">set</span><span class="p">[</span><span class="n">OrthologBase</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a dictionary for one source organism and ID type.</span>

<span class="sd">        Args:</span>
<span class="sd">            target:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">            source:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">            id_type:</span>
<span class="sd">                The identifier type to use.</span>
<span class="sd">            only_swissprot:</span>
<span class="sd">                Use only SwissProt IDs.</span>
<span class="sd">            oma</span>
<span class="sd">                Use orthology information from the Orthologous Matrix (OMA).</span>
<span class="sd">                Currently this is the recommended source for orthology data.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use orthology information from NCBI HomoloGene.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use orthology information from Ensembl.</span>
<span class="sd">            oma_rel_type:</span>
<span class="sd">                Restrict relations to certain types.</span>
<span class="sd">            oma_score:</span>
<span class="sd">                Lower threshold for similarity metric.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only the high confidence orthology relations from Ensembl.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                Ensembl orthology relation types to use. Possible values are</span>
<span class="sd">                `one2one`, `one2many` and `many2many`. By default only</span>
<span class="sd">                `one2one` is used.</span>
<span class="sd">            full_records:</span>
<span class="sd">                Include not only the identifiers, but also some properties of</span>
<span class="sd">                the orthology relationships.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A dict with identifiers of the source organism as keys, and</span>
<span class="sd">            sets of their orthologs as values.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="n">param</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_translation_param</span><span class="p">(</span><span class="nb">locals</span><span class="p">())</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">resource</span><span class="p">,</span> <span class="n">keys</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">RESOURCE_PARAM</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">param</span><span class="p">[</span><span class="n">resource</span><span class="p">]:</span>

                <span class="k">continue</span>

            <span class="n">table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">which_table</span><span class="p">(</span>
                <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
                <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
                <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
                <span class="n">resource</span> <span class="o">=</span> <span class="n">resource</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">dct</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">asdict</span><span class="p">(</span>
                <span class="n">full_records</span> <span class="o">=</span> <span class="n">full_records</span><span class="p">,</span>
                <span class="o">**</span><span class="p">{</span>
                    <span class="n">p</span><span class="p">:</span> <span class="n">v</span>
                    <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">param</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">keys</span>
                <span class="p">}</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">dct</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

                <span class="n">result</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">o</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>



<div class="viewcode-block" id="OrthologyManager.get_df">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OrthologyManager.html#pypath.utils.orthology.OrthologyManager.get_df">[docs]</a>
    <span class="k">def</span> <span class="nf">get_df</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">oma</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">oma_rel_type</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">oma_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a data frame for one source organism and ID type.</span>

<span class="sd">        Args:</span>
<span class="sd">            target:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">            source:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">            id_type:</span>
<span class="sd">                The identifier type to use.</span>
<span class="sd">            only_swissprot:</span>
<span class="sd">                Use only SwissProt IDs.</span>
<span class="sd">            oma</span>
<span class="sd">                Use orthology information from the Orthologous Matrix (OMA).</span>
<span class="sd">                Currently this is the recommended source for orthology data.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use orthology information from NCBI HomoloGene.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use orthology information from Ensembl.</span>
<span class="sd">            oma_rel_type:</span>
<span class="sd">                Restrict relations to certain types.</span>
<span class="sd">            oma_score:</span>
<span class="sd">                Lower threshold for similarity metric.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only the high confidence orthology relations from Ensembl.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                Ensembl orthology relation types to use. Possible values are</span>
<span class="sd">                `one2one`, `one2many` and `many2many`. By default only</span>
<span class="sd">                `one2one` is used.</span>
<span class="sd">            full_records:</span>
<span class="sd">                Include not only the identifiers, but also some properties of</span>
<span class="sd">                the orthology relationships.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Ignored.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A data frame with pairs of orthologous identifiers,</span>
<span class="sd">            in two columns: &quot;source&quot; and &quot;target&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="n">param</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_translation_param</span><span class="p">(</span><span class="nb">locals</span><span class="p">())</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">resource</span><span class="p">,</span> <span class="n">keys</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">RESOURCE_PARAM</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">param</span><span class="p">[</span><span class="n">resource</span><span class="p">]:</span>

                <span class="k">continue</span>

            <span class="n">table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">which_table</span><span class="p">(</span>
                <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
                <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
                <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
                <span class="n">resource</span> <span class="o">=</span> <span class="n">resource</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">table</span><span class="o">.</span><span class="n">df</span><span class="p">(</span>
                    <span class="n">full_records</span> <span class="o">=</span> <span class="n">full_records</span><span class="p">,</span>
                    <span class="o">**</span><span class="p">{</span>
                        <span class="n">p</span><span class="p">:</span> <span class="n">v</span>
                        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">param</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                        <span class="k">if</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">keys</span>
                    <span class="p">}</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>



<div class="viewcode-block" id="OrthologyManager.translate_df">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OrthologyManager.html#pypath.utils.orthology.OrthologyManager.translate_df">[docs]</a>
    <span class="k">def</span> <span class="nf">translate_df</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">cols</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">oma</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">oma_rel_type</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">oma_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translate columns in a data frame.</span>

<span class="sd">        Args:</span>
<span class="sd">            df:</span>
<span class="sd">                A data frame.</span>
<span class="sd">            cols:</span>
<span class="sd">                One or more columns to be translated. It can be a single</span>
<span class="sd">                column name, an iterable of column names or a dict where</span>
<span class="sd">                keys are column names and values are ID types. Except this</span>
<span class="sd">                last case, identifiers are assumed to be `id_type`.</span>
<span class="sd">            target:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">            source:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">            id_type:</span>
<span class="sd">                The default identifier type to use, will be used for all</span>
<span class="sd">                columns where ID type is not specified.</span>
<span class="sd">            only_swissprot:</span>
<span class="sd">                Use only SwissProt IDs.</span>
<span class="sd">            oma</span>
<span class="sd">                Use orthology information from the Orthologous Matrix (OMA).</span>
<span class="sd">                Currently this is the recommended source for orthology data.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use orthology information from NCBI HomoloGene.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use orthology information from Ensembl.</span>
<span class="sd">            oma_rel_type:</span>
<span class="sd">                Restrict relations to certain types.</span>
<span class="sd">            oma_score:</span>
<span class="sd">                Lower threshold for similarity metric.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only the high confidence orthology relations from Ensembl.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                Ensembl orthology relation types to use. Possible values are</span>
<span class="sd">                `one2one`, `one2many` and `many2many`. By default only</span>
<span class="sd">                `one2one` is used.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Same as providing a dict to ``cols``, but beware, keys</span>
<span class="sd">                (column names) can not match existing argument names of</span>
<span class="sd">                this function.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A data frame with the same column layout as the input, and the</span>
<span class="sd">            identifiers translated as demanded. Rows that could not be</span>
<span class="sd">            translated are omitted.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cols</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>

            <span class="n">cols</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">col</span><span class="p">,</span> <span class="n">id_type</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">cols</span><span class="p">))</span>

        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span>
        <span class="n">id_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="k">for</span> <span class="n">_id_type</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">cols</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>

            <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;self&#39;</span><span class="p">)</span>
            <span class="n">args</span><span class="p">[</span><span class="s1">&#39;id_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_id_type</span>
            <span class="n">ortho_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_df</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">)</span>

            <span class="n">table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">which_table</span><span class="p">(</span>
                <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
                <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
                <span class="n">id_type</span> <span class="o">=</span> <span class="n">_id_type</span><span class="p">,</span>
                <span class="n">resource</span> <span class="o">=</span> <span class="s1">&#39;oma&#39;</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">df</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">translate_df</span><span class="p">(</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">,</span>
                <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cols</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">_id_type</span><span class="p">],</span>
                <span class="n">ortho_df</span> <span class="o">=</span> <span class="n">ortho_df</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span></div>



    <span class="k">def</span> <span class="nf">_translation_param</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">loc</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>

        <span class="n">param</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">resource</span><span class="p">,</span> <span class="n">keys</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">RESOURCE_PARAM</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="n">enabled</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">first_value</span><span class="p">(</span>
                <span class="n">loc</span><span class="p">[</span><span class="n">resource</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_param</span><span class="p">[</span><span class="n">resource</span><span class="p">],</span>
                <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;orthology_</span><span class="si">{</span><span class="n">resource</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="n">param</span><span class="p">[</span><span class="n">resource</span><span class="p">]</span> <span class="o">=</span> <span class="n">enabled</span>

            <span class="k">if</span> <span class="n">enabled</span><span class="p">:</span>

                <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">keys</span><span class="p">:</span>

                    <span class="n">param</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">first_value</span><span class="p">(</span>
                        <span class="n">loc</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">resource</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">],</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_param</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">resource</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">],</span>
                        <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;orthology_</span><span class="si">{</span><span class="n">resource</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">),</span>
                    <span class="p">)</span>

        <span class="k">return</span> <span class="n">param</span>


    <span class="k">def</span> <span class="nf">_remove_expired</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">last_used</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expiry</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>

            <span class="k">if</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">last_used</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">lifetime</span> <span class="ow">and</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
                    <span class="s1">&#39;Removing orthology table from taxon </span><span class="si">%u</span><span class="s1"> to </span><span class="si">%u</span><span class="s1"> &#39;</span>
                    <span class="s1">&#39;(only SwissProt: </span><span class="si">%s</span><span class="s1">; resource: </span><span class="si">%s</span><span class="s1">; ID type: </span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="n">key</span>
                <span class="p">)</span>

                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">expiry</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>


    <span class="k">def</span> <span class="fm">__del__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">_orthology_cleanup_timeloop</span><span class="p">,</span> <span class="s1">&#39;stop&#39;</span><span class="p">):</span>

            <span class="n">_orthology_cleanup_timeloop</span><span class="o">.</span><span class="n">stop</span><span class="p">()</span></div>



<div class="viewcode-block" id="SequenceContainer">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.SequenceContainer.html#pypath.utils.orthology.SequenceContainer">[docs]</a>
<span class="k">class</span> <span class="nc">SequenceContainer</span><span class="p">(</span><span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="p">):</span>


<div class="viewcode-block" id="SequenceContainer.__init__">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.SequenceContainer.html#pypath.utils.orthology.SequenceContainer.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">preload_seq</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">isoforms</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is an object to store sequences of multiple</span>
<span class="sd">        organisms and select the appropriate one.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;_logger&#39;</span><span class="p">):</span>

            <span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;orthology&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">seq_isoforms</span> <span class="o">=</span> <span class="n">isoforms</span>

        <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">preload_seq</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load_seq</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span></div>



    <span class="k">def</span> <span class="nf">load_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon</span><span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;seq&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">ncbi_tax_id</span>

        <span class="k">if</span> <span class="n">taxon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="n">_se</span><span class="o">.</span><span class="n">swissprot_seq</span><span class="p">(</span>
                <span class="n">organism</span> <span class="o">=</span> <span class="n">taxon</span><span class="p">,</span>
                <span class="n">isoforms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_isoforms</span>
            <span class="p">)</span>


    <span class="k">def</span> <span class="nf">get_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">,</span> <span class="n">taxon</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">taxon</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

            <span class="k">if</span> <span class="n">taxon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">load_seq</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">taxon</span><span class="p">]:</span>

                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">taxon</span><span class="p">][</span><span class="n">protein</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="k">for</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">iteritems</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">:</span>

                    <span class="k">return</span> <span class="n">seq</span><span class="p">[</span><span class="n">protein</span><span class="p">]</span></div>



<div class="viewcode-block" id="Proteomes">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.Proteomes.html#pypath.utils.orthology.Proteomes">[docs]</a>
<span class="k">class</span> <span class="nc">Proteomes</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>


<div class="viewcode-block" id="Proteomes.__init__">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.Proteomes.html#pypath.utils.orthology.Proteomes.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">preload_prot</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;_taxonomy&#39;</span><span class="p">):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_up_taxonomy</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_taxonomy</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="p">(</span><span class="n">preload_prot</span> <span class="ow">or</span> <span class="p">()):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load_proteome</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span></div>



    <span class="k">def</span> <span class="nf">load_proteome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

        <span class="n">only_swissprot</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span>
                <span class="k">if</span> <span class="n">only_swissprot</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span>
            <span class="n">only_swissprot</span>
        <span class="p">)</span>

        <span class="n">key</span> <span class="o">=</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">only_swissprot</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span><span class="n">uniprot_db</span><span class="o">.</span><span class="n">all_uniprots</span><span class="p">(</span><span class="o">*</span><span class="n">key</span><span class="p">))</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span><span class="p">[</span><span class="n">key</span><span class="p">]:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span><span class="p">[</span><span class="n">protein</span><span class="p">]</span> <span class="o">=</span> <span class="n">key</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">only_swissprot</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">load_proteome</span><span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">get_taxon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">,</span> <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>

        <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_taxon_trembl</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span>
            <span class="n">only_swissprot</span> <span class="ow">and</span>
            <span class="n">ncbi_tax_id</span> <span class="ow">and</span>
            <span class="ow">not</span> <span class="n">uniprot_db</span><span class="o">.</span><span class="n">is_swissprot</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="n">organism</span> <span class="o">=</span> <span class="n">ncbi_tax_id</span><span class="p">)</span>
        <span class="p">):</span>

            <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">ncbi_tax_id</span>


    <span class="k">def</span> <span class="nf">get_taxon_trembl</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">):</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_up_taxonomy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">has_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">):</span>

        <span class="k">return</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span>


    <span class="k">def</span> <span class="nf">is_swissprot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">):</span>

        <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_taxon</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">))</span>


    <span class="k">def</span> <span class="nf">load_taxonomy</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_up_taxonomy</span> <span class="o">=</span> <span class="n">uniprot_input</span><span class="o">.</span><span class="n">uniprot_taxonomy</span><span class="p">(</span><span class="n">ncbi_tax_ids</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span></div>



<div class="viewcode-block" id="ProteinOrthology">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.ProteinOrthology.html#pypath.utils.orthology.ProteinOrthology">[docs]</a>
<span class="k">class</span> <span class="nc">ProteinOrthology</span><span class="p">(</span><span class="n">Proteomes</span><span class="p">):</span>

    <span class="n">_param</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;id&#39;</span><span class="p">,)</span>


<div class="viewcode-block" id="ProteinOrthology.__init__">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.ProteinOrthology.html#pypath.utils.orthology.ProteinOrthology.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This class translates between homologous UniProt IDs of two organisms</span>
<span class="sd">        based on NCBI HomoloGene and Ensembl data. In case of HomoloGene,</span>
<span class="sd">        the UniProt-UniProt translation table is created by translating the</span>
<span class="sd">        source organism UniProts to RefSeq and Entrez IDs, finding the</span>
<span class="sd">        homologues (orthologues) for these IDs, and then translating them</span>
<span class="sd">        to the target organism UniProt IDs. In case of Ensembl, we obtain</span>
<span class="sd">        data with Ensembl protein identifiers and translate those to UniProt.</span>

<span class="sd">        Args</span>
<span class="sd">            target:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">            source:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">            id_type:</span>
<span class="sd">                The identifier type to use.</span>
<span class="sd">            only_swissprot:</span>
<span class="sd">                Use only SwissProt IDs.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Resource specific parameters.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_resource_l</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">resource</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="n">Proteomes</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_proteome</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_set_param</span><span class="p">(</span><span class="n">kwargs</span><span class="p">,</span> <span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">_param</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load</span><span class="p">()</span></div>



    <span class="k">def</span> <span class="nf">reload</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="n">modname</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__module__</span>
        <span class="n">mod</span> <span class="o">=</span> <span class="nb">__import__</span><span class="p">(</span><span class="n">modname</span><span class="p">,</span> <span class="n">fromlist</span><span class="o">=</span><span class="p">[</span><span class="n">modname</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]])</span>
        <span class="n">imp</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
        <span class="n">new</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mod</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;__class__&#39;</span><span class="p">,</span> <span class="n">new</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

        <span class="k">pass</span>


<div class="viewcode-block" id="ProteinOrthology.translate">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.ProteinOrthology.html#pypath.utils.orthology.ProteinOrthology.translate">[docs]</a>
    <span class="k">def</span> <span class="nf">translate</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
            <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        For one UniProt ID of the source organism returns all orthologues</span>
<span class="sd">        from the target organism.</span>

<span class="sd">        Args:</span>
<span class="sd">            identifier:</span>
<span class="sd">                An identifier corresponding to the ID type and source organism</span>
<span class="sd">                of the instance.</span>
<span class="sd">            full_records:</span>
<span class="sd">                Include not only the identifiers, but also some properties of</span>
<span class="sd">                the orthology relationships.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Resource specific translation parameters.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A set of identifiers of orthologues in the target taxon.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">identifier</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">identifier</span><span class="p">,)</span>
                <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">identifier</span><span class="p">,</span> <span class="s1">&#39;components&#39;</span><span class="p">)</span> <span class="k">else</span>
            <span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">identifier</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">identifier</span><span class="p">))</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">full_records</span><span class="p">:</span>

            <span class="n">result</span> <span class="o">=</span> <span class="p">{</span><span class="n">o</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">result</span><span class="p">}</span>

        <span class="k">return</span> <span class="n">result</span></div>



<div class="viewcode-block" id="ProteinOrthology.asdict">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.ProteinOrthology.html#pypath.utils.orthology.ProteinOrthology.asdict">[docs]</a>
    <span class="k">def</span> <span class="nf">asdict</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">set</span><span class="p">[</span><span class="n">OrthologBase</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a dictionary from the translation table.</span>

<span class="sd">        Args:</span>
<span class="sd">            full_records:</span>
<span class="sd">                Include not only the identifiers, but also some properties of</span>
<span class="sd">                the orthology relationships.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Resource specific filtering criteria.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A dict with identifiers of the source organism as keys, and</span>
<span class="sd">            sets of their orthologs as values.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">proc</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">full_records</span> <span class="k">else</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span>
            <span class="n">s</span><span class="p">:</span> <span class="p">{</span><span class="n">proc</span><span class="p">(</span><span class="n">o</span><span class="p">)</span> <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">orthologs</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">o</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)}</span>
            <span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">orthologs</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span></div>



<div class="viewcode-block" id="ProteinOrthology.df">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.ProteinOrthology.html#pypath.utils.orthology.ProteinOrthology.df">[docs]</a>
    <span class="k">def</span> <span class="nf">df</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Orthologous pairs as data frame.</span>

<span class="sd">        Args:</span>
<span class="sd">            full_records:</span>
<span class="sd">                Include not only the identifiers, but also some properties of</span>
<span class="sd">                the orthology relationships.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Resource specific filtering criteria.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A data frame with pairs of orthologous identifiers,</span>
<span class="sd">            in two columns: &quot;source&quot; and &quot;target&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Creating translation data frame between &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;organisms `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1">` and `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`, &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;ID type `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="si">}</span><span class="s1">`.&#39;</span>
        <span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">asdict</span><span class="p">(</span><span class="n">full_records</span> <span class="o">=</span> <span class="n">full_records</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span>
                <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">],</span>
            <span class="p">)</span><span class="o">.</span>
            <span class="n">explode</span><span class="p">(</span><span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="n">ignore_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span>
            <span class="n">dropna</span><span class="p">()</span><span class="o">.</span>
            <span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">full_records</span><span class="p">:</span>

            <span class="c1"># some beautiful pandas code again</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="n">df</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
                        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">tolist</span><span class="p">()),</span>
                    <span class="p">],</span>
                    <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span>
                <span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;target&#39;</span><span class="p">})</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span></div>



<div class="viewcode-block" id="ProteinOrthology.translate_df">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.ProteinOrthology.html#pypath.utils.orthology.ProteinOrthology.translate_df">[docs]</a>
    <span class="k">def</span> <span class="nf">translate_df</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
            <span class="n">cols</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ortho_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translate columns in a data frame.</span>

<span class="sd">        Args:</span>
<span class="sd">            df:</span>
<span class="sd">                A data frame.</span>
<span class="sd">            cols:</span>
<span class="sd">                One or more columns to be translated. It can be a single</span>
<span class="sd">                column name, an iterable of column names or a dict where</span>
<span class="sd">                keys are column names and values are ID types. Except this</span>
<span class="sd">                last case, identifiers are assumed to be UniProt.</span>
<span class="sd">            ortho_df:</span>
<span class="sd">                Override the translation data frame. If provided, the</span>
<span class="sd">                parameters in `kwargs` won&#39;t have an effect. Must have</span>
<span class="sd">                columns &quot;source&quot; and &quot;target&quot;.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Resource specific translation parameters.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A data frame with the same column layout as the input, and the</span>
<span class="sd">            identifiers translated as demanded. Rows that could not be</span>
<span class="sd">            translated are omitted.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="sa">f</span><span class="s1">&#39;Translating data frame column(s) from &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;organism `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1">` to `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
        <span class="p">)</span>

        <span class="n">ortho_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="k">if</span> <span class="n">ortho_df</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">ortho_df</span><span class="p">)</span><span class="o">.</span>
            <span class="n">rename</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s1">&#39;source&#39;</span><span class="p">:</span> <span class="s1">&#39;pypath_internal_source&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;target&#39;</span><span class="p">:</span> <span class="s1">&#39;pypath_internal_target&#39;</span><span class="p">,</span>
                <span class="p">},</span>
                <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="n">col_order</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">:</span>

            <span class="n">_log</span><span class="p">(</span>
                <span class="sa">f</span><span class="s1">&#39;Translating `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="si">}</span><span class="s1">` IDs of organism `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1">` &#39;</span>
                <span class="sa">f</span><span class="s1">&#39;in column `</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s1">` to organism `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
            <span class="p">)</span>

            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                    <span class="n">ortho_df</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s1">&#39;pypath_internal_source&#39;</span><span class="p">:</span> <span class="n">col</span><span class="p">},</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">),</span>
                    <span class="n">on</span> <span class="o">=</span> <span class="n">col</span><span class="p">,</span>
                    <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;inner&#39;</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span>
                <span class="n">drop</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span>
                <span class="n">rename</span><span class="p">({</span><span class="s1">&#39;pypath_internal_target&#39;</span><span class="p">:</span> <span class="n">col</span><span class="p">},</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span><span class="p">[</span><span class="n">col_order</span><span class="p">]</span></div>



    <span class="k">def</span> <span class="nf">_translation_param</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">loc</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>

        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">loc</span><span class="p">[</span><span class="n">p</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">OrthologyManager</span><span class="o">.</span><span class="n">TRANSLATION_PARAM</span>
        <span class="p">)</span>


    <span class="k">def</span> <span class="nf">_set_param</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">loc</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="o">*</span><span class="n">params</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>

        <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">params</span><span class="p">:</span>

            <span class="n">key</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;orthology_</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_resource_l</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">param</span><span class="si">}</span><span class="s1">&#39;</span>
            <span class="nb">setattr</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">,</span>
                <span class="n">param</span><span class="p">,</span>
                <span class="n">common</span><span class="o">.</span><span class="n">first_value</span><span class="p">(</span><span class="n">loc</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">param</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span> <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">key</span><span class="p">)),</span>
            <span class="p">)</span>


    <span class="k">def</span> <span class="nf">match</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ortholog</span><span class="p">:</span> <span class="n">OrthologBase</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>

        <span class="k">return</span> <span class="kc">True</span>


    <span class="k">def</span> <span class="nf">_from_pickle</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>

        <span class="k">if</span> <span class="p">(</span>
            <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;orthology_cache&#39;</span><span class="p">)</span> <span class="ow">and</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pickle_path</span><span class="p">)</span>
        <span class="p">):</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pickle_path</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fp</span><span class="p">)</span>
                <span class="n">_log</span><span class="p">(</span>
                    <span class="s1">&#39;Orthology table from taxon </span><span class="si">%u</span><span class="s1"> to </span><span class="si">%u</span><span class="s1"> (only SwissProt: </span><span class="si">%s</span><span class="s1">; &#39;</span>
                    <span class="s1">&#39;resource: </span><span class="si">%s</span><span class="s1">; ID type: </span><span class="si">%s</span><span class="s1">) has been loaded from `</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">key</span> <span class="o">+</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pickle_path</span><span class="p">,)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="kc">True</span>

        <span class="k">return</span> <span class="kc">False</span>


    <span class="k">def</span> <span class="nf">_to_pickle</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pickle_path</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>

            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="n">fp</span><span class="p">)</span>
            <span class="n">_log</span><span class="p">(</span>
                <span class="s1">&#39;Orthology table from taxon </span><span class="si">%u</span><span class="s1"> to </span><span class="si">%u</span><span class="s1"> (only SwissProt: </span><span class="si">%s</span><span class="s1">; &#39;</span>
                <span class="s1">&#39;resource: </span><span class="si">%s</span><span class="s1">; ID type: </span><span class="si">%s</span><span class="s1">) has been saved to `</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">key</span> <span class="o">+</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pickle_path</span><span class="p">,)</span>
                <span class="p">)</span>
            <span class="p">)</span>


    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">key</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="n">OrthologyTableKey</span><span class="p">(</span>
            <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
            <span class="n">target</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span><span class="p">,</span>
            <span class="n">resource</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_resource_l</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
        <span class="p">)</span>


    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">pickle_path</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">cache_mod</span><span class="o">.</span><span class="n">get_cachedir</span><span class="p">(),</span>
            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">common</span><span class="o">.</span><span class="n">md5</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">key</span><span class="p">))</span><span class="si">}</span><span class="s1">.pickle&#39;</span><span class="p">,</span>
        <span class="p">)</span>


    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">len</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>


    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s1">&#39;&lt;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">resource</span><span class="si">}</span><span class="s1"> Orthology table from </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1"> to &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="si">}</span><span class="s1"> IDs, </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="si">}</span><span class="s1"> relationships&gt;&#39;</span>
        <span class="p">)</span></div>



<div class="viewcode-block" id="HomologeneOrthology">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.HomologeneOrthology.html#pypath.utils.orthology.HomologeneOrthology">[docs]</a>
<span class="k">class</span> <span class="nc">HomologeneOrthology</span><span class="p">(</span><span class="n">ProteinOrthology</span><span class="p">):</span>

    <span class="n">resource</span> <span class="o">=</span> <span class="s1">&#39;HomoloGene&#39;</span>


<div class="viewcode-block" id="HomologeneOrthology.load">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.HomologeneOrthology.html#pypath.utils.orthology.HomologeneOrthology.load">[docs]</a>
    <span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Load orthology data from NCBI HomoloGene.</span>

<span class="sd">        Builds orthology translation table as dict based on NCBI HomoloGene</span>
<span class="sd">        data. If the `id_type` is supported by HomoloGene (Gene Symbol, RefSeq,</span>
<span class="sd">        Entrez, GI), the data will be simply loaded. For other ID types it</span>
<span class="sd">        translates HomoloGene Gene Symbol, RefSeq and Entrez tables to UniProt</span>
<span class="sd">        and then translates the orthologous UniProt pairs to the desired</span>
<span class="sd">        ID type.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_pickle</span><span class="p">():</span>

            <span class="k">return</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;refseq&#39;</span><span class="p">,</span> <span class="s1">&#39;refseqp&#39;</span><span class="p">,</span> <span class="s1">&#39;entrez&#39;</span><span class="p">,</span> <span class="s1">&#39;gi&#39;</span><span class="p">):</span>

            <span class="n">data</span> <span class="o">=</span> <span class="n">homologene_input</span><span class="o">.</span><span class="n">homologene_dict</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">s</span><span class="p">:</span> <span class="p">{</span><span class="n">HomologeneOrtholog</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">target_ids</span><span class="p">}</span>
                <span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">target_ids</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>
            <span class="k">return</span>

        <span class="n">hg</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="n">homologene_input</span><span class="o">.</span><span class="n">homologene_dict</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                <span class="n">id_type</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">id_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;refseq&#39;</span><span class="p">,</span> <span class="s1">&#39;entrez&#39;</span><span class="p">)</span>
        <span class="p">}</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Loading orthology data from NCBI HomoloGene &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;between organisms `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1">` and `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span><span class="p">)]:</span>

            <span class="n">target_uniprots</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

            <span class="k">for</span> <span class="n">id_type</span><span class="p">,</span> <span class="n">hgdata</span> <span class="ow">in</span> <span class="n">hg</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

                <span class="n">hg_source_ids</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span>
                    <span class="n">u</span><span class="p">,</span>
                    <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                    <span class="n">id_type</span><span class="p">,</span>
                    <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">hg_source_ids</span><span class="p">:</span> <span class="k">continue</span>

                <span class="n">hg_target_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">(</span>
                    <span class="n">hgdata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span>
                    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">hg_source_ids</span><span class="p">)</span>
                <span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">hg_target_ids</span><span class="p">:</span> <span class="k">continue</span>

                <span class="n">target_uniprots</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="n">mapping</span><span class="o">.</span><span class="n">map_names</span><span class="p">(</span>
                        <span class="n">hg_target_ids</span><span class="p">,</span>
                        <span class="n">id_type</span><span class="p">,</span>
                        <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                        <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span> <span class="o">==</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">:</span>

                <span class="n">source_ids</span> <span class="o">=</span> <span class="p">(</span><span class="n">u</span><span class="p">,)</span>
                <span class="n">target_ids</span> <span class="o">=</span> <span class="n">target_uniprots</span>

            <span class="k">else</span><span class="p">:</span>

                <span class="n">source_ids</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span>
                    <span class="n">u</span><span class="p">,</span>
                    <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
                    <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">target_ids</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_names</span><span class="p">(</span>
                    <span class="n">target_uniprots</span><span class="p">,</span>
                    <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
                    <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">source_ids</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="n">HomologeneOrtholog</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">target_ids</span>
                    <span class="p">}</span>
                <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_to_pickle</span><span class="p">()</span></div>
</div>



<div class="viewcode-block" id="EnsemblOrthology">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.EnsemblOrthology.html#pypath.utils.orthology.EnsemblOrthology">[docs]</a>
<span class="k">class</span> <span class="nc">EnsemblOrthology</span><span class="p">(</span><span class="n">ProteinOrthology</span><span class="p">):</span>

    <span class="n">_param</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;hc&#39;</span><span class="p">,</span> <span class="s1">&#39;types&#39;</span><span class="p">)</span>
    <span class="n">resource</span> <span class="o">=</span> <span class="s1">&#39;Ensembl&#39;</span>

<div class="viewcode-block" id="EnsemblOrthology.__init__">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.EnsemblOrthology.html#pypath.utils.orthology.EnsemblOrthology.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">types</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span>
                <span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span>
            <span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Orthology translation with Ensembl data.</span>

<span class="sd">        Args</span>
<span class="sd">            target:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">            source:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">            id_type:</span>
<span class="sd">                The identifier type to use.</span>
<span class="sd">            only_swissprot:</span>
<span class="sd">                Use only SwissProt IDs.</span>
<span class="sd">            hc:</span>
<span class="sd">                Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            types:</span>
<span class="sd">                The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ProteinOrthology</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="nb">locals</span><span class="p">())</span></div>



    <span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="n">target_organism</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
        <span class="n">source_organism</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">)</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Loading orthology data from Ensembl &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;between organisms `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1">` and `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_pickle</span><span class="p">():</span>

            <span class="k">return</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">target_organism</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">source_organism</span><span class="p">:</span>

            <span class="n">_log</span><span class="p">(</span>
                <span class="s1">&#39;No Ensembl orthology data available between &#39;</span>
                <span class="sa">f</span><span class="s1">&#39;organisms `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1">` and `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
            <span class="p">)</span>
            <span class="k">return</span>

        <span class="n">target_prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">target_organism</span><span class="si">}</span><span class="s1">_homolog_&#39;</span>

        <span class="n">attr_target_ensp</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">target_prefix</span><span class="si">}</span><span class="s1">ensembl_peptide&#39;</span>
        <span class="n">attr_conf</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">target_prefix</span><span class="si">}</span><span class="s1">orthology_confidence&#39;</span>
        <span class="n">attr_type</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">target_prefix</span><span class="si">}</span><span class="s1">orthology_type&#39;</span>

        <span class="n">ensembl_data</span> <span class="o">=</span> <span class="n">biomart</span><span class="o">.</span><span class="n">biomart_homology</span><span class="p">(</span>
            <span class="n">source_organism</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
            <span class="n">target_organism</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">_id_types</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;target&#39;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s1">&#39;genesymbol&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">target_prefix</span><span class="si">}</span><span class="s1">associated_gene_name&#39;</span><span class="p">,</span>
                <span class="s1">&#39;ensp&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">target_prefix</span><span class="si">}</span><span class="s1">ensembl_peptide&#39;</span><span class="p">,</span>
                <span class="s1">&#39;ensg&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">target_prefix</span><span class="si">}</span><span class="s1">ensembl_gene&#39;</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="s1">&#39;source&#39;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s1">&#39;genesymbol&#39;</span><span class="p">:</span> <span class="s1">&#39;external_gene_name&#39;</span><span class="p">,</span>
                <span class="s1">&#39;ensp&#39;</span><span class="p">:</span> <span class="s1">&#39;ensembl_peptide_id&#39;</span><span class="p">,</span>
                <span class="s1">&#39;ensg&#39;</span><span class="p">:</span> <span class="s1">&#39;ensembl_gene_id&#39;</span><span class="p">,</span>
            <span class="p">},</span>
        <span class="p">}</span>
        <span class="n">attr_tgt_id</span> <span class="o">=</span> <span class="n">_id_types</span><span class="p">[</span><span class="s1">&#39;target&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">target_prefix</span><span class="si">}</span><span class="s1">ensembl_peptide&#39;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">attr_src_id</span> <span class="o">=</span> <span class="n">_id_types</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
            <span class="s1">&#39;ensembl_peptide_id&#39;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span> <span class="ow">in</span> <span class="n">_id_types</span><span class="p">[</span><span class="s1">&#39;target&#39;</span><span class="p">]:</span>

            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">ensembl_data</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_src_id</span><span class="p">)]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
                    <span class="n">EnsemblOrtholog</span><span class="p">(</span>
                        <span class="nb">id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_tgt_id</span><span class="p">),</span>
                        <span class="n">hc</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_conf</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span>
                        <span class="n">types</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_type</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
                    <span class="p">)</span>
                <span class="p">)</span>

        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">ensembl_data</span><span class="p">:</span>

            <span class="n">ids</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="k">for</span> <span class="n">side</span><span class="p">,</span> <span class="n">attr_id</span> <span class="ow">in</span> <span class="p">(</span>
                    <span class="p">(</span><span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="n">attr_src_id</span><span class="p">),</span>
                    <span class="p">(</span><span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="n">attr_tgt_id</span><span class="p">)</span>
                <span class="p">):</span>

                <span class="n">uniprots</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span>
                    <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_id</span><span class="p">),</span>
                    <span class="s1">&#39;ensp&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                    <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">side</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="n">ids</span><span class="p">[</span><span class="n">side</span><span class="p">]</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_names</span><span class="p">(</span>
                    <span class="n">uniprots</span><span class="p">,</span>
                    <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
                    <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">side</span><span class="p">),</span>
                    <span class="n">uniprot_cleanup</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">ids</span><span class="p">[</span><span class="n">side</span><span class="p">]:</span>

                    <span class="k">continue</span>

            <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="n">EnsemblOrtholog</span><span class="p">(</span>
                            <span class="nb">id</span> <span class="o">=</span> <span class="n">t</span><span class="p">,</span>
                            <span class="n">hc</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_conf</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span>
                            <span class="n">types</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_type</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
                        <span class="p">)</span>
                        <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">[</span><span class="s1">&#39;target&#39;</span><span class="p">]</span>
                    <span class="p">}</span>
                <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_to_pickle</span><span class="p">()</span>


<div class="viewcode-block" id="EnsemblOrthology.match">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.EnsemblOrthology.html#pypath.utils.orthology.EnsemblOrthology.match">[docs]</a>
    <span class="k">def</span> <span class="nf">match</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ortholog</span><span class="p">:</span> <span class="n">OrthologBase</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check an ortholog against filtering criteria.</span>

<span class="sd">        Args:</span>
<span class="sd">            ortholog:</span>
<span class="sd">                An ortholog record.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Override default filtering parameters.</span>

<span class="sd">        Returns:</span>
<span class="sd">            True if the ortholog meets the criteria.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">}</span>
        <span class="n">hc</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;hc&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">hc</span><span class="p">)</span>
        <span class="n">types</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;types&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">types</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="p">(</span><span class="ow">not</span> <span class="n">hc</span> <span class="ow">or</span> <span class="n">ortholog</span><span class="o">.</span><span class="n">hc</span><span class="p">)</span> <span class="ow">and</span>
            <span class="p">(</span><span class="ow">not</span> <span class="n">types</span> <span class="ow">or</span> <span class="n">ortholog</span><span class="o">.</span><span class="n">types</span> <span class="ow">in</span> <span class="n">types</span><span class="p">)</span>
        <span class="p">)</span></div>
</div>



<div class="viewcode-block" id="OmaOrthology">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OmaOrthology.html#pypath.utils.orthology.OmaOrthology">[docs]</a>
<span class="k">class</span> <span class="nc">OmaOrthology</span><span class="p">(</span><span class="n">ProteinOrthology</span><span class="p">):</span>

    <span class="n">_param</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;rel_type&#39;</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">)</span>
    <span class="n">resource</span> <span class="o">=</span> <span class="s1">&#39;OMA&#39;</span>

<div class="viewcode-block" id="OmaOrthology.__init__">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OmaOrthology.html#pypath.utils.orthology.OmaOrthology.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">rel_type</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
             <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Orthology translation with Ensembl data.</span>

<span class="sd">        Args</span>
<span class="sd">            target:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">            source:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">            id_type:</span>
<span class="sd">                The identifier type to use.</span>
<span class="sd">            only_swissprot:</span>
<span class="sd">                Use only SwissProt IDs.</span>
<span class="sd">            rel_type:</span>
<span class="sd">                Restrict relations to certain types.</span>
<span class="sd">            score:</span>
<span class="sd">                Lower threshold for similarity metric.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ProteinOrthology</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">**</span><span class="nb">locals</span><span class="p">())</span></div>



    <span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Loading orthology data from OMA &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;between organisms `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1">` and `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_pickle</span><span class="p">():</span>

            <span class="k">return</span>

        <span class="n">oma_data</span> <span class="o">=</span> <span class="n">oma_input</span><span class="o">.</span><span class="n">oma_orthologs</span><span class="p">(</span>
            <span class="n">organism_a</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
            <span class="n">organism_b</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">oma_data</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">a</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
                <span class="n">OmaOrtholog</span><span class="p">(</span>
                    <span class="nb">id</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">b</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="n">score</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                    <span class="n">rel_type</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">rel_type</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_to_pickle</span><span class="p">()</span>


<div class="viewcode-block" id="OmaOrthology.match">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.OmaOrthology.html#pypath.utils.orthology.OmaOrthology.match">[docs]</a>
    <span class="k">def</span> <span class="nf">match</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ortholog</span><span class="p">:</span> <span class="n">OrthologBase</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check an ortholog against filtering criteria.</span>

<span class="sd">        Args:</span>
<span class="sd">            ortholog:</span>
<span class="sd">                An ortholog record.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Override default filtering parameters.</span>

<span class="sd">        Returns:</span>
<span class="sd">            True if the ortholog meets the criteria.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">}</span>
        <span class="n">score</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;score&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
        <span class="n">rel_type</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;rel_type&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">rel_type</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">score</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">ortholog</span><span class="o">.</span><span class="n">score</span> <span class="o">&gt;=</span> <span class="n">score</span><span class="p">)</span> <span class="ow">and</span>
            <span class="p">(</span><span class="ow">not</span> <span class="n">rel_type</span> <span class="ow">or</span> <span class="n">ortholog</span><span class="o">.</span><span class="n">rel_type</span> <span class="ow">in</span> <span class="n">rel_type</span><span class="p">)</span>
        <span class="p">)</span></div>
</div>



<div class="viewcode-block" id="PtmOrthology">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.PtmOrthology.html#pypath.utils.orthology.PtmOrthology">[docs]</a>
<span class="k">class</span> <span class="nc">PtmOrthology</span><span class="p">(</span><span class="n">Proteomes</span><span class="p">,</span> <span class="n">SequenceContainer</span><span class="p">):</span>


<div class="viewcode-block" id="PtmOrthology.__init__">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.PtmOrthology.html#pypath.utils.orthology.PtmOrthology.__init__">[docs]</a>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">strict</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">orthology_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;_logger&#39;</span><span class="p">):</span>

            <span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;orthology&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>
        <span class="n">SequenceContainer</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">Proteomes</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_seq</span><span class="p">(</span><span class="n">taxon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reptm</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;([A-Z\d]{6,10})_([A-Z])(\d*)&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">strict</span> <span class="o">=</span> <span class="n">strict</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">orthology_args</span> <span class="o">=</span> <span class="n">orthology_args</span> <span class="ow">or</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ptm_orthology</span><span class="p">()</span></div>



    <span class="k">def</span> <span class="nf">reload</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="n">modname</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__module__</span>
        <span class="n">mod</span> <span class="o">=</span> <span class="nb">__import__</span><span class="p">(</span><span class="n">modname</span><span class="p">,</span> <span class="n">fromlist</span><span class="o">=</span><span class="p">[</span><span class="n">modname</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]])</span>
        <span class="n">imp</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
        <span class="n">new</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mod</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;__class__&#39;</span><span class="p">,</span> <span class="n">new</span><span class="p">)</span>


<div class="viewcode-block" id="PtmOrthology.translate_site">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.PtmOrthology.html#pypath.utils.orthology.PtmOrthology.translate_site">[docs]</a>
    <span class="k">def</span> <span class="nf">translate_site</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">protein</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">intera</span><span class="o">.</span><span class="n">Protein</span><span class="p">,</span>
            <span class="n">res</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">offset</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
            <span class="n">isoform</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">typ</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;phosphorylation&#39;</span><span class="p">,</span>
            <span class="n">source_organism</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translates one PTM site.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein:</span>
<span class="sd">                A protein identifier or an intera.Protein object.</span>
<span class="sd">            res:</span>
<span class="sd">                Single letter code of the residue.</span>
<span class="sd">            offset:</span>
<span class="sd">                Sequence offset of the site.</span>
<span class="sd">            isoform:</span>
<span class="sd">                Sequence isoform.</span>
<span class="sd">            typ:</span>
<span class="sd">                Modification type.</span>
<span class="sd">            source_organism:</span>
<span class="sd">                Name or NCBI Taxonomy ID of the source organism.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list of tuples with the identifier, isoform, residue, offset,</span>
<span class="sd">            taxon and modification type of the orthologous PTM sites.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">result</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_source</span><span class="p">(</span><span class="n">source_organism</span><span class="p">)</span>
        <span class="n">protein_id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="s1">&#39;identifier&#39;</span><span class="p">,</span> <span class="n">protein</span><span class="p">)</span>
        <span class="n">sourceptm</span> <span class="o">=</span> <span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">isoform</span><span class="p">,</span> <span class="n">res</span><span class="p">,</span> <span class="n">offset</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="n">typ</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_taxon</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">:</span>
            <span class="n">result</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">sourceptm</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">result</span>

        <span class="k">if</span> <span class="n">sourceptm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span><span class="p">:</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span><span class="p">[</span><span class="n">sourceptm</span><span class="p">]:</span>

                <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span><span class="p">[</span><span class="n">sourceptm</span><span class="p">]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">strict</span><span class="p">:</span>

            <span class="n">tsubs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manager</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
                <span class="n">identifiers</span> <span class="o">=</span> <span class="n">protein_id</span><span class="p">,</span>
                <span class="n">target</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
                <span class="n">only_swissprot</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span><span class="p">,</span>
                <span class="n">id_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
                <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">orthology_args</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">tsub</span> <span class="ow">in</span> <span class="n">tsubs</span><span class="p">:</span>

                <span class="n">se</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="n">tsub</span><span class="p">,</span> <span class="n">taxon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">se</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="k">for</span> <span class="n">toffset</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="n">offset</span><span class="p">,</span> <span class="n">offset</span> <span class="o">+</span> <span class="mi">3</span><span class="p">):</span>

                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">se</span><span class="o">.</span><span class="n">isoforms</span><span class="p">():</span>

                        <span class="n">tres</span> <span class="o">=</span> <span class="n">se</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">toffset</span><span class="p">,</span> <span class="n">isoform</span> <span class="o">=</span> <span class="n">i</span><span class="p">)</span>

                        <span class="k">if</span> <span class="n">tres</span> <span class="o">==</span> <span class="n">res</span><span class="p">:</span>

                            <span class="n">result</span><span class="o">.</span><span class="n">add</span><span class="p">((</span>
                                <span class="n">tsub</span><span class="p">,</span>
                                <span class="n">i</span><span class="p">,</span>
                                <span class="n">tres</span><span class="p">,</span>
                                <span class="n">toffset</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                                <span class="n">typ</span><span class="p">,</span>
                            <span class="p">))</span>

                    <span class="k">if</span> <span class="n">result</span><span class="p">:</span>
                        <span class="k">break</span>

        <span class="k">return</span> <span class="n">result</span></div>



    <span class="k">def</span> <span class="nf">translate_domain</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">domain</span><span class="p">:</span> <span class="n">intera</span><span class="o">.</span><span class="n">Domain</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">intera</span><span class="o">.</span><span class="n">Domain</span><span class="p">]:</span>

        <span class="k">return</span> <span class="p">[</span>
            <span class="n">intera</span><span class="o">.</span><span class="n">Domain</span><span class="p">(</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">target_id</span><span class="p">,</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">target_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manager</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
                <span class="n">identifiers</span> <span class="o">=</span> <span class="n">domain</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span>
                <span class="n">target</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                <span class="n">id_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_type</span><span class="p">,</span>
                <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_source</span><span class="p">(</span><span class="n">domain</span><span class="o">.</span><span class="n">ncbi_tax_id</span><span class="p">),</span>
                <span class="n">only_swissprot</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span><span class="p">,</span>
                <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">orthology_args</span>
            <span class="p">)</span>
        <span class="p">]</span>


    <span class="k">def</span> <span class="nf">translate_ptm</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ptm</span><span class="p">:</span> <span class="n">intera</span><span class="o">.</span><span class="n">Ptm</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">intera</span><span class="o">.</span><span class="n">Ptm</span><span class="p">]:</span>

        <span class="n">tptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_site</span><span class="p">(</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">protein</span><span class="p">,</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">number</span><span class="p">,</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">isoform</span><span class="p">,</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">typ</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tptms</span><span class="p">:</span>

            <span class="n">se</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">taxon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>

            <span class="k">if</span> <span class="p">(</span><span class="n">se</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">se</span><span class="o">.</span><span class="n">isof</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">strict</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="n">res</span> <span class="o">=</span> <span class="n">intera</span><span class="o">.</span><span class="n">Residue</span><span class="p">(</span>
                <span class="n">number</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">isoform</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">region</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">se</span><span class="o">.</span><span class="n">get_region</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">isoform</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">se</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">se</span><span class="o">.</span><span class="n">isof</span>
                <span class="k">else</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">mot</span> <span class="o">=</span> <span class="n">intera</span><span class="o">.</span><span class="n">Motif</span><span class="p">(</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">start</span><span class="p">,</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">end</span><span class="p">,</span>
                <span class="n">instance</span> <span class="o">=</span> <span class="n">region</span><span class="p">,</span>
                <span class="n">isoform</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">ptm</span> <span class="o">=</span> <span class="n">intera</span><span class="o">.</span><span class="n">Ptm</span><span class="p">(</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">motif</span> <span class="o">=</span> <span class="n">mot</span><span class="p">,</span>
                <span class="n">residue</span> <span class="o">=</span> <span class="n">res</span><span class="p">,</span>
                <span class="n">typ</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">5</span><span class="p">],</span>
                <span class="n">isoform</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">evidences</span> <span class="o">=</span> <span class="n">ptm</span><span class="o">.</span><span class="n">evidences</span><span class="p">,</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ptm</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span>


    <span class="k">def</span> <span class="nf">translate_domain_motif</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">dmotif</span><span class="p">:</span> <span class="n">intera</span><span class="o">.</span><span class="n">DomainMotif</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">intera</span><span class="o">.</span><span class="n">DomainMotif</span><span class="p">]:</span>

        <span class="n">ds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_domain</span><span class="p">(</span><span class="n">dmotif</span><span class="o">.</span><span class="n">domain</span><span class="p">)</span>
        <span class="n">ps</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_ptm</span><span class="p">(</span><span class="n">dmotif</span><span class="o">.</span><span class="n">ptm</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">[</span>
            <span class="n">intera</span><span class="o">.</span><span class="n">DomainMotif</span><span class="p">(</span>
                <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">evidences</span> <span class="o">=</span> <span class="n">dmotif</span><span class="o">.</span><span class="n">evidences</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">ps</span><span class="p">)</span>
        <span class="p">]</span>


    <span class="k">def</span> <span class="nf">translate_residue</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">residue</span><span class="p">:</span> <span class="n">intera</span><span class="o">.</span><span class="n">Residue</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">intera</span><span class="o">.</span><span class="n">Residue</span><span class="p">]:</span>

        <span class="k">return</span> <span class="p">[</span>
            <span class="n">intera</span><span class="o">.</span><span class="n">Residue</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">isoform</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_site</span><span class="p">(</span>
                <span class="n">residue</span><span class="o">.</span><span class="n">protein</span><span class="p">,</span>
                <span class="n">residue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="n">residue</span><span class="o">.</span><span class="n">number</span><span class="p">,</span>
                <span class="n">residue</span><span class="o">.</span><span class="n">isoform</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">]</span>


<div class="viewcode-block" id="PtmOrthology.translate">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.PtmOrthology.html#pypath.utils.orthology.PtmOrthology.translate">[docs]</a>
    <span class="k">def</span> <span class="nf">translate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">return_strings</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translates anything: string notation, intera objects, tuples.</span>

<span class="sd">        - one PTM provided as tuple of (UniProt, amino acid, offest)</span>
<span class="sd">        - one PTM provided as string (e.g. `P00533_S231`)</span>
<span class="sd">        - instance from pypath.intera: DomainMotif, Domain or Ptm</span>

<span class="sd">        Additional arguments can be isoform and typ (modification type).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">tuple</span><span class="p">:</span>

            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_site</span><span class="p">(</span><span class="o">*</span><span class="n">x</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="ow">in</span> <span class="n">_const</span><span class="o">.</span><span class="n">CHAR_TYPES</span><span class="p">:</span>

            <span class="n">ptm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reptm</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">ptm</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

                <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_site</span><span class="p">(</span>
                    <span class="n">ptm</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                    <span class="n">ptm</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                    <span class="nb">int</span><span class="p">(</span><span class="n">ptm</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span>
                    <span class="o">**</span><span class="n">kwargs</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="n">return_strings</span><span class="p">:</span>

            <span class="n">result</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_</span><span class="si">%s%u</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">result</span><span class="p">]</span>

        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">intera</span><span class="o">.</span><span class="n">Ptm</span><span class="p">):</span>

            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_ptm</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">intera</span><span class="o">.</span><span class="n">Domain</span><span class="p">):</span>

            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_domain</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">intera</span><span class="o">.</span><span class="n">DomainMotif</span><span class="p">):</span>

            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_domain_motif</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span></div>



<div class="viewcode-block" id="PtmOrthology.ptm_orthology">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.PtmOrthology.html#pypath.utils.orthology.PtmOrthology.ptm_orthology">[docs]</a>
    <span class="k">def</span> <span class="nf">ptm_orthology</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Load PTM orthology data from PhosphoSite.</span>

<span class="sd">        Creates an orthology translation dict of phosphosites</span>
<span class="sd">        based on phosphorylation sites table from PhosphoSitePlus.</span>
<span class="sd">        In the result all PTMs represented by a tuple of the following</span>
<span class="sd">        6 elements: UniProt ID, isoform (int), residue one letter code,</span>
<span class="sd">        residue number (int), NCBI Taxonomy ID (int), modification type.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="n">nondigit</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;[^\d]+&#39;</span><span class="p">)</span>

        <span class="n">unknown_taxa</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">typ</span> <span class="ow">in</span> <span class="n">common</span><span class="o">.</span><span class="n">psite_mod_types</span><span class="p">:</span>

            <span class="n">groups</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="n">url</span> <span class="o">=</span> <span class="n">urls</span><span class="o">.</span><span class="n">urls</span><span class="p">[</span><span class="s1">&#39;psite_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">typ</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;url&#39;</span><span class="p">]</span>
            <span class="n">c</span> <span class="o">=</span> <span class="n">curl</span><span class="o">.</span><span class="n">Curl</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">large</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

            <span class="n">data</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">result</span>

            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="mi">4</span><span class="p">):</span>
                <span class="n">null</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>

                <span class="n">r</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">:</span>

                    <span class="k">continue</span>

                <span class="n">uniprot</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                <span class="n">isoform</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="mi">1</span>
                        <span class="k">if</span> <span class="s1">&#39;-&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">uniprot</span> <span class="k">else</span>
                    <span class="nb">int</span><span class="p">(</span><span class="n">uniprot</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
                <span class="p">)</span>
                <span class="n">uniprot</span> <span class="o">=</span> <span class="n">uniprot</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">aa</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">4</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">num</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">nondigit</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">r</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>

                <span class="k">if</span> <span class="n">r</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">taxa</span><span class="p">:</span>

                    <span class="n">unknown_taxa</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
                    <span class="k">continue</span>

                <span class="n">tax</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">taxa</span><span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="mi">6</span><span class="p">]]</span>
                <span class="n">group</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">5</span><span class="p">])</span>

                <span class="n">this_site</span> <span class="o">=</span> <span class="p">(</span><span class="n">uniprot</span><span class="p">,</span> <span class="n">isoform</span><span class="p">,</span> <span class="n">aa</span><span class="p">,</span> <span class="n">num</span><span class="p">,</span> <span class="n">tax</span><span class="p">,</span> <span class="n">typ</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

                <span class="k">if</span> <span class="n">group</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
                    <span class="n">groups</span><span class="p">[</span><span class="n">group</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([])</span>

                <span class="n">groups</span><span class="p">[</span><span class="n">group</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">this_site</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">group</span><span class="p">,</span> <span class="n">sites</span> <span class="ow">in</span> <span class="n">iteritems</span><span class="p">(</span><span class="n">groups</span><span class="p">):</span>

                <span class="k">for</span> <span class="n">site1</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">:</span>

                    <span class="k">for</span> <span class="n">site2</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">:</span>

                        <span class="k">if</span> <span class="n">site1</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="n">site2</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span>

                            <span class="k">continue</span>

                        <span class="k">if</span> <span class="n">site1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span><span class="p">:</span>

                            <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span><span class="p">[</span><span class="n">site1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

                        <span class="k">if</span> <span class="n">site2</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span><span class="p">[</span><span class="n">site1</span><span class="p">]:</span>

                            <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span><span class="p">[</span><span class="n">site1</span><span class="p">][</span><span class="n">site2</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([])</span>

                        <span class="bp">self</span><span class="o">.</span><span class="n">ptmortho</span><span class="p">[</span><span class="n">site1</span><span class="p">][</span><span class="n">site2</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">site2</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">unknown_taxa</span><span class="p">):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
                <span class="s1">&#39;Unknown taxa encountered: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span>
                    <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">unknown_taxa</span><span class="p">))</span>
                <span class="p">)</span>
            <span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">_get_source</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the NCBI Taxonomy ID of the source taxon.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">source</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">ncbi_tax_id</span><span class="p">:</span>

            <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s1">&#39;No source taxon provided (argument: `</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s1">`, &#39;</span>
                <span class="sa">f</span><span class="s1">&#39;instance: `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">source</span><span class="si">}</span><span class="s1">`)&#39;</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">ncbi_tax_id</span>


    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;ptmortho&#39;</span><span class="p">,</span> <span class="p">()))</span>


    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;&lt;PTM Orthology: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="si">}</span><span class="s1"> sites&gt;&#39;</span></div>



<div class="viewcode-block" id="init">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.init.html#pypath.utils.orthology.init">[docs]</a>
<span class="k">def</span> <span class="nf">init</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialize the orthology manager.</span>

<span class="sd">    Creates an instance of the orthology manager. Stores it in the module</span>
<span class="sd">    namespace.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">globals</span><span class="p">()[</span><span class="s1">&#39;manager&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">OrthologyManager</span><span class="p">()</span></div>



<div class="viewcode-block" id="get_manager">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.get_manager.html#pypath.utils.orthology.get_manager">[docs]</a>
<span class="k">def</span> <span class="nf">get_manager</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Access the orthology manager.</span>

<span class="sd">    Returns the orthology manager, an object which loads and unloads the</span>
<span class="sd">    orthology lookup tables as necessary, and provides the interface for</span>
<span class="sd">    querying the orthology data. Normally an instance of the manager</span>
<span class="sd">    belongs to the module, and if it does not exist yet, will be created</span>
<span class="sd">    automatically.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="s1">&#39;manager&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">globals</span><span class="p">():</span>

        <span class="n">init</span><span class="p">()</span>

    <span class="k">return</span> <span class="nb">globals</span><span class="p">()[</span><span class="s1">&#39;manager&#39;</span><span class="p">]</span></div>



<div class="viewcode-block" id="translate">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.translate.html#pypath.utils.orthology.translate">[docs]</a>
<span class="k">def</span> <span class="nf">translate</span><span class="p">(</span>
        <span class="n">identifiers</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">oma</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">oma_rel_type</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">oma_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">ensembl_types</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Translate one or more identifiers by orthologous gene pairs.</span>

<span class="sd">    Args:</span>
<span class="sd">        identifiers:</span>
<span class="sd">            One or more identifers of the source organism, of ID type</span>
<span class="sd">            `id_type`.</span>
<span class="sd">        target:</span>
<span class="sd">            Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">        source:</span>
<span class="sd">            Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">        id_type:</span>
<span class="sd">            The identifier type to use.</span>
<span class="sd">        only_swissprot:</span>
<span class="sd">            Use only SwissProt IDs.</span>
<span class="sd">        oma</span>
<span class="sd">            Use orthology information from the Orthologous Matrix (OMA).</span>
<span class="sd">            Currently this is the recommended source for orthology data.</span>
<span class="sd">        homologene:</span>
<span class="sd">            Use orthology information from NCBI HomoloGene.</span>
<span class="sd">        ensembl:</span>
<span class="sd">            Use orthology information from Ensembl.</span>
<span class="sd">        oma_rel_type:</span>
<span class="sd">            Restrict relations to certain types.</span>
<span class="sd">        oma_score:</span>
<span class="sd">            Lower threshold for similarity metric.</span>
<span class="sd">        ensembl_hc:</span>
<span class="sd">            Use only the high confidence orthology relations from Ensembl.</span>
<span class="sd">        ensembl_types:</span>
<span class="sd">            Ensembl orthology relation types to use. Possible values are</span>
<span class="sd">            `one2one`, `one2many` and `many2many`. By default only</span>
<span class="sd">            `one2one` is used.</span>
<span class="sd">        full_records:</span>
<span class="sd">            Include not only the identifiers, but also some properties of</span>
<span class="sd">            the orthology relationships.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Set of identifiers of orthologous genes or proteins in the</span>
<span class="sd">        target taxon.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>

    <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;manager&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">manager</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">)</span></div>



<div class="viewcode-block" id="get_dict">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.get_dict.html#pypath.utils.orthology.get_dict">[docs]</a>
<span class="k">def</span> <span class="nf">get_dict</span><span class="p">(</span>
        <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">oma</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">oma_rel_type</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">oma_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">ensembl_types</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">set</span><span class="p">[</span><span class="n">OrthologBase</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a dictionary for one source organism and ID type.</span>

<span class="sd">    Args:</span>
<span class="sd">        target:</span>
<span class="sd">            Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">        source:</span>
<span class="sd">            Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">        id_type:</span>
<span class="sd">            The identifier type to use.</span>
<span class="sd">        only_swissprot:</span>
<span class="sd">            Use only SwissProt IDs.</span>
<span class="sd">        oma</span>
<span class="sd">            Use orthology information from the Orthologous Matrix (OMA).</span>
<span class="sd">            Currently this is the recommended source for orthology data.</span>
<span class="sd">        homologene:</span>
<span class="sd">            Use orthology information from NCBI HomoloGene.</span>
<span class="sd">        ensembl:</span>
<span class="sd">            Use orthology information from Ensembl.</span>
<span class="sd">        oma_rel_type:</span>
<span class="sd">            Restrict relations to certain types.</span>
<span class="sd">        oma_score:</span>
<span class="sd">            Lower threshold for similarity metric.</span>
<span class="sd">        ensembl_hc:</span>
<span class="sd">            Use only the high confidence orthology relations from Ensembl.</span>
<span class="sd">        ensembl_types:</span>
<span class="sd">            Ensembl orthology relation types to use. Possible values are</span>
<span class="sd">            `one2one`, `one2many` and `many2many`. By default only</span>
<span class="sd">            `one2one` is used.</span>
<span class="sd">        full_records:</span>
<span class="sd">            Include not only the identifiers, but also some properties of</span>
<span class="sd">            the orthology relationships.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dict with identifiers of the source organism as keys, and</span>
<span class="sd">        sets of their orthologs as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>

    <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;manager&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">manager</span><span class="o">.</span><span class="n">get_dict</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">)</span></div>



<div class="viewcode-block" id="get_df">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.get_df.html#pypath.utils.orthology.get_df">[docs]</a>
<span class="k">def</span> <span class="nf">get_df</span><span class="p">(</span>
        <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">oma</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">oma_rel_type</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">oma_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">ensembl_types</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">full_records</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a data frame for one source organism and ID type.</span>

<span class="sd">    Args:</span>
<span class="sd">        target:</span>
<span class="sd">            Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">        source:</span>
<span class="sd">            Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">        id_type:</span>
<span class="sd">            The identifier type to use.</span>
<span class="sd">        only_swissprot:</span>
<span class="sd">            Use only SwissProt IDs.</span>
<span class="sd">        oma</span>
<span class="sd">            Use orthology information from the Orthologous Matrix (OMA).</span>
<span class="sd">            Currently this is the recommended source for orthology data.</span>
<span class="sd">        homologene:</span>
<span class="sd">            Use orthology information from NCBI HomoloGene.</span>
<span class="sd">        ensembl:</span>
<span class="sd">            Use orthology information from Ensembl.</span>
<span class="sd">        oma_rel_type:</span>
<span class="sd">            Restrict relations to certain types.</span>
<span class="sd">        oma_score:</span>
<span class="sd">            Lower threshold for similarity metric.</span>
<span class="sd">        ensembl_hc:</span>
<span class="sd">            Use only the high confidence orthology relations from Ensembl.</span>
<span class="sd">        ensembl_types:</span>
<span class="sd">            Ensembl orthology relation types to use. Possible values are</span>
<span class="sd">            `one2one`, `one2many` and `many2many`. By default only</span>
<span class="sd">            `one2one` is used.</span>
<span class="sd">        full_records:</span>
<span class="sd">            Include not only the identifiers, but also some properties of</span>
<span class="sd">            the orthology relationships.</span>
<span class="sd">        kwargs:</span>
<span class="sd">            Ignored.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A data frame with pairs of orthologous identifiers,</span>
<span class="sd">        in two columns: &quot;source&quot; and &quot;target&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>

    <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;manager&#39;</span><span class="p">)</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;kwargs&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">manager</span><span class="o">.</span><span class="n">get_df</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">)</span></div>



<div class="viewcode-block" id="translate_df">
<a class="viewcode-back" href="../../../api/pypath.utils.orthology.translate_df.html#pypath.utils.orthology.translate_df">[docs]</a>
<span class="k">def</span> <span class="nf">translate_df</span><span class="p">(</span>
        <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
        <span class="n">target</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">source</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">cols</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="n">only_swissprot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">oma</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">oma_rel_type</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">set</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;1:1&#39;</span><span class="p">,</span> <span class="s1">&#39;1:n&#39;</span><span class="p">,</span> <span class="s1">&#39;m:1&#39;</span><span class="p">,</span> <span class="s1">&#39;m:n&#39;</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">oma_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">ensembl_types</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Translate columns in a data frame.</span>

<span class="sd">    Args:</span>
<span class="sd">        df:</span>
<span class="sd">            A data frame.</span>
<span class="sd">        target:</span>
<span class="sd">            Name or NCBI Taxonomy ID of the target organism.</span>
<span class="sd">        source:</span>
<span class="sd">            Name or NCBI Taxonomy ID of the source organism.</span>
<span class="sd">        cols:</span>
<span class="sd">            One or more columns to be translated. It can be a single</span>
<span class="sd">            column name, an iterable of column names or a dict where</span>
<span class="sd">            keys are column names and values are ID types. Except this</span>
<span class="sd">            last case, identifiers are assumed to be `id_type`.</span>
<span class="sd">        id_type:</span>
<span class="sd">            The default identifier type to use, will be used for all</span>
<span class="sd">            columns where ID type is not specified.</span>
<span class="sd">        only_swissprot:</span>
<span class="sd">            Use only SwissProt IDs.</span>
<span class="sd">        oma</span>
<span class="sd">            Use orthology information from the Orthologous Matrix (OMA).</span>
<span class="sd">            Currently this is the recommended source for orthology data.</span>
<span class="sd">        homologene:</span>
<span class="sd">            Use orthology information from NCBI HomoloGene.</span>
<span class="sd">        ensembl:</span>
<span class="sd">            Use orthology information from Ensembl.</span>
<span class="sd">        oma_rel_type:</span>
<span class="sd">            Restrict relations to certain types.</span>
<span class="sd">        oma_score:</span>
<span class="sd">            Lower threshold for similarity metric.</span>
<span class="sd">        ensembl_hc:</span>
<span class="sd">            Use only the high confidence orthology relations from Ensembl.</span>
<span class="sd">        ensembl_types:</span>
<span class="sd">            Ensembl orthology relation types to use. Possible values are</span>
<span class="sd">            `one2one`, `one2many` and `many2many`. By default only</span>
<span class="sd">            `one2one` is used.</span>
<span class="sd">        kwargs:</span>
<span class="sd">            Same as providing a dict to ``cols``, but beware, keys</span>
<span class="sd">            (column names) can not match existing argument names of</span>
<span class="sd">            this function.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A data frame with the same column layout as the input, and the</span>
<span class="sd">        identifiers translated as demanded. Rows that could not be</span>
<span class="sd">        translated are omitted.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>

    <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;manager&#39;</span><span class="p">)</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;kwargs&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">manager</span><span class="o">.</span><span class="n">translate_df</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../index.html">Welcome to pypath!</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../webservice.html">Webservice</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../releasehistory.html">Release history</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../reference.html">Reference</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../../../notebooks/manual.html">The pypath book</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Melih Darcan, Erva Ulusoy, Tennur Kilic, Omer Kaan Vural, Ahmet Rifaioglu, Charles Tapley Hoyt.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>