
<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor &#8212; pypath 0.15.8 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/pypathdocs.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api/pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor';</script>
    <link rel="icon" href="../_static/omnipath-logo.svg"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="pypath.core.enz_sub.EnzymeSubstrateProcessor" href="pypath.core.enz_sub.EnzymeSubstrateProcessor.html" />
    <link rel="prev" title="pypath.core.enz_sub.EnzymeSubstrateAggregator" href="pypath.core.enz_sub.EnzymeSubstrateAggregator.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  <div class="navbar-header-items__start">
    
      <div class="navbar-item">
  

<a class="navbar-brand logo" href="../contents.html">
  
  
  
  
    
    
      
    
    
    <img src="../_static/omnipath-logo.svg" class="logo__image only-light" alt="Pypath"/>
    <script>document.write(`<img src="../_static/omnipath-logo.svg" class="logo__image only-dark" alt="Pypath"/>`);</script>
  
  
    <p class="title logo__title">Pypath</p>
  
</a></div>
    
  </div>
  
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          
<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
        </div>
      
      
        <div class="navbar-item">
<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter"></i></span>
            <label class="sr-only">Twitter</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box"></i></span>
            <label class="sr-only">PyPI</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university"></i></span>
            <label class="sr-only">OmniPath</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users"></i></span>
            <label class="sr-only">Saez Lab</label></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">
<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter"></i></span>
            <label class="sr-only">Twitter</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box"></i></span>
            <label class="sr-only">PyPI</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university"></i></span>
            <label class="sr-only">OmniPath</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users"></i></span>
            <label class="sr-only">Saez Lab</label></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">





<nav aria-label="Breadcrumbs">
  <ul class="bd-breadcrumbs" role="navigation" aria-label="Breadcrumb">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../contents.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../reference.html" class="nav-link">Reference</a></li>
    
    
    <li class="breadcrumb-item"><i class="fa-solid fa-ellipsis"></i></li>
    
    
    <li class="breadcrumb-item"><a href="pypath.core.enz_sub.html" class="nav-link">pypath.core.enz_sub</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="pypath-core-enz-sub-enzymesubstratehomologyprocessor">
<h1>pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor<a class="headerlink" href="#pypath-core-enz-sub-enzymesubstratehomologyprocessor" title="Permalink to this heading">§</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pypath.core.enz_sub.</span></span><span class="sig-name descname"><span class="pre">EnzymeSubstrateHomologyProcessor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ncbi_tax_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">map_by_homology_from</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type_enzyme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type_substrate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homology_only_swissprot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_homology_strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/core/enz_sub.html#EnzymeSubstrateHomologyProcessor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor" title="Permalink to this definition">§</a></dt>
<dd><p>Bases: <a class="reference internal" href="pypath.utils.homology.PtmHomology.html#pypath.utils.homology.PtmHomology" title="pypath.utils.homology.PtmHomology"><code class="xref py py-class docutils literal notranslate"><span class="pre">PtmHomology</span></code></a>, <a class="reference internal" href="pypath.core.enz_sub.EnzymeSubstrateProcessor.html#pypath.core.enz_sub.EnzymeSubstrateProcessor" title="pypath.core.enz_sub.EnzymeSubstrateProcessor"><code class="xref py py-class docutils literal notranslate"><span class="pre">EnzymeSubstrateProcessor</span></code></a>, <a class="reference internal" href="pypath.share.session.Logger.html#pypath.share.session.Logger" title="pypath.share.session.Logger"><code class="xref py py-class docutils literal notranslate"><span class="pre">Logger</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ncbi_tax_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">map_by_homology_from</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type_enzyme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type_substrate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homology_only_swissprot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_homology_strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/core/enz_sub.html#EnzymeSubstrateHomologyProcessor.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.__init__" title="Permalink to this definition">§</a></dt>
<dd><p>Unifies a <cite>pypath.core.enz_sub.EnzymeSubstrateProcessor</cite> and
a <cite>pypath.utils.homology.EnzymeSubstrateHomology</cite> object to build
a set of enzyme-substrate interactions from a database and
subsequently translate them by homology to one different organism.
Multiple organism can be chosen as the source of the
enzyme-substrate interactions. For example if you want mouse
interactions, you can translate them from human and from rat.
To get the original mouse interactions themselves, use an
other instance of the <cite>EnzymeSubstrateProcessor</cite>.
To have both the original and the homology translated set,
and also from multiple databases, whatmore all these merged
into a single set, use the <cite>EnzymeSubstrateAggregator</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_method</strong> (<em>str</em>) – Data source for <cite>EnzymeSubstrateProcessor</cite>.</p></li>
<li><p><strong>ncbi_tax_id</strong> (<em>int</em>) – The NCBI Taxonomy ID the interactions
should be translated to.</p></li>
<li><p><strong>homology_only_swissprot</strong> (<em>bool</em>) – Use only SwissProt
(i.e. not Trembl) at homology
translation.</p></li>
<li><p><strong>ptm_homology_strict</strong> (<em>bool</em>) – Use only those homologous PTM pairs
which are in PhosphoSite data, i.e.
do not look for residues with same
offset in protein sequence.</p></li>
</ul>
</dd>
</dl>
<p>See further options at <cite>EnzymeSubstrateProcessor</cite>.</p>
</dd></dl>

<p class="rubric">Methods</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.__init__" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.__init__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__init__</span></code></a>(ncbi_tax_id[, input_param, ...])</p></td>
<td><p>Unifies a <cite>pypath.core.enz_sub.EnzymeSubstrateProcessor</cite> and a <cite>pypath.utils.homology.EnzymeSubstrateHomology</cite> object to build a set of enzyme-substrate interactions from a database and subsequently translate them by homology to one different organism.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.asdict" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.asdict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">asdict</span></code></a>([source, homologene, ensembl, ...])</p></td>
<td><p>Create a dictionary for one source organism and ID type.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.df" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.df"><code class="xref py py-obj docutils literal notranslate"><span class="pre">df</span></code></a>([source, homologene, ensembl, ...])</p></td>
<td><p>Create a data frame for one source organism and ID type.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">ensure_source_taxon</span></code>(source)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_seq</span></code>(protein[, taxon])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_source</span></code>([source])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_taxon</span></code>(protein[, swissprot_only])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_taxon_trembl</span></code>(protein)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_protein</span></code>(protein)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">input_is</span></code>(i[, op])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_swissprot</span></code>(protein)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">load</span></code>([source])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.load_data" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.load_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_data</span></code></a>()</p></td>
<td><p>Loads the data by the defined input method.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_ensembl</span></code>(source)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_enz_sub</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.load_homologene" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.load_homologene"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_homologene</span></code></a>(source)</p></td>
<td><p>Builds orthology translation table as dict from UniProt to Uniprot, obtained from NCBI HomoloGene data.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_proteome</span></code>(taxon[, swissprot_only])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_seq</span></code>(taxon)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.ptm_orthology" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.ptm_orthology"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ptm_orthology</span></code></a>()</p></td>
<td><p>Creates an orthology translation dict of phosphosites based on phosphorylation sites table from PhosphoSitePlus.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">reload</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">reset_ptmprocessor</span></code>([seq, ncbi_tax_id])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_default_source</span></code>([source])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.set_inputargs" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.set_inputargs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_inputargs</span></code></a>(**inputargs)</p></td>
<td><p>Sets the arguments to be provided for the input method.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.set_method" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.set_method"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_method</span></code></a>()</p></td>
<td><p>Selects the input method.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_taxon</span></code>(ncbi_tax_id)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">setup</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate</span></code></a>(x[, return_strings])</p></td>
<td><p>Translates anything:</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate_df" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate_df"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_df</span></code></a>(df[, cols, source, homologene, ...])</p></td>
<td><p>Translate columns in a data frame.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_domain</span></code>(domain)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_domain_motif</span></code>(dmotif)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_ptm</span></code>(ptm)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_residue</span></code>(residue)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate_site" title="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate_site"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_site</span></code></a>(protein, res, offset[, ...])</p></td>
<td><p>Translates one PTM site.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'one2one'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'one2many'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'many2many'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'uniprot'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.asdict" title="Permalink to this definition">§</a></dt>
<dd><p>Create a dictionary for one source organism and ID type.</p>
<dl class="simple">
<dt>Args</dt><dd><dl class="simple">
<dt>source:</dt><dd><p>The source organism.</p>
</dd>
<dt>homologene:</dt><dd><p>Use NCBI HomoloGene data for ortholog lookup.</p>
</dd>
<dt>ensembl:</dt><dd><p>Use Ensembl data for ortholog lookup.</p>
</dd>
<dt>ensembl_hc:</dt><dd><p>Use only high confidence orthology relations
from Ensembl. By default it is True. You can also set it
by the <cite>ensembl_hc</cite> attribute.</p>
</dd>
<dt>ensembl_types:</dt><dd><p>The Ensembl orthology relationship types
to use. Possible values are <cite>one2one</cite>, <cite>one2many</cite> and
<cite>many2many</cite>. By default only <cite>one2one</cite> is used. You can
also set this parameter by the <cite>ensembl_types</cite> attribute.</p>
</dd>
<dt>id_type:</dt><dd><p>Identifier type to translate from and to. If a tuple, the
first element will be used for keys (source organism) and
the second for the values (target organism). By default
UniProt IDs are assumed. For a complete list of ID types see
<code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping.get_mapper().id_types()</span></code> and
<code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping.get_mapper().mapping_tables()</span></code>.</p>
</dd>
</dl>
</dd>
<dt>Return</dt><dd><p>A dict with identifiers of the source organism as keys, and
sets of their orthologs as values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.df">
<span class="sig-name descname"><span class="pre">df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'one2one'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'one2many'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'many2many'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'uniprot'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.df" title="Permalink to this definition">§</a></dt>
<dd><p>Create a data frame for one source organism and ID type.</p>
<dl class="simple">
<dt>Args</dt><dd><dl class="simple">
<dt>source:</dt><dd><p>The source organism.</p>
</dd>
<dt>homologene:</dt><dd><p>Use NCBI HomoloGene data for ortholog lookup.</p>
</dd>
<dt>ensembl:</dt><dd><p>Use Ensembl data for ortholog lookup.</p>
</dd>
<dt>ensembl_hc:</dt><dd><p>Use only high confidence orthology relations
from Ensembl. By default it is True. You can also set it
by the <cite>ensembl_hc</cite> attribute.</p>
</dd>
<dt>ensembl_types:</dt><dd><p>The Ensembl orthology relationship types
to use. Possible values are <cite>one2one</cite>, <cite>one2many</cite> and
<cite>many2many</cite>. By default only <cite>one2one</cite> is used. You can
also set this parameter by the <cite>ensembl_types</cite> attribute.</p>
</dd>
<dt>id_type:</dt><dd><p>Identifier type to translate from and to. If a tuple, the
first element will be used for keys (source organism) and
the second for the values (target organism). By default
UniProt IDs are assumed. For a complete list of ID types see
<code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping.get_mapper().id_types()</span></code> and
<code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping.get_mapper().mapping_tables()</span></code>.</p>
</dd>
</dl>
</dd>
<dt>Return</dt><dd><p>A data frame with pairs of orthologous identifiers,
in two columns: “source” and “target”.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.load_data">
<span class="sig-name descname"><span class="pre">load_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.load_data" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the data by the defined input method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.load_homologene">
<span class="sig-name descname"><span class="pre">load_homologene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.load_homologene" title="Permalink to this definition">§</a></dt>
<dd><p>Builds orthology translation table as dict from UniProt to Uniprot,
obtained from NCBI HomoloGene data. Uses RefSeq and Entrez IDs for
translation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.ptm_orthology">
<span class="sig-name descname"><span class="pre">ptm_orthology</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.ptm_orthology" title="Permalink to this definition">§</a></dt>
<dd><p>Creates an orthology translation dict of phosphosites
based on phosphorylation sites table from PhosphoSitePlus.
In the result all PTMs represented by a tuple of the following
6 elements: UniProt ID, isoform (int), residue one letter code,
residue number (int), NCBI Taxonomy ID (int), modification type.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.set_inputargs">
<span class="sig-name descname"><span class="pre">set_inputargs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">inputargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.set_inputargs" title="Permalink to this definition">§</a></dt>
<dd><p>Sets the arguments to be provided for the input method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.set_method">
<span class="sig-name descname"><span class="pre">set_method</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.set_method" title="Permalink to this definition">§</a></dt>
<dd><p>Selects the input method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_strings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate" title="Permalink to this definition">§</a></dt>
<dd><p>Translates anything:</p>
<ul class="simple">
<li><p>one UniProt ID</p></li>
<li><p>one PTM provided as tuple of (UniProt, amino acid, offest)</p></li>
<li><p>one PTM provided as string (e.g. <cite>P00533_S231</cite>)</p></li>
<li><p>instance from pypath.intera: DomainMotif, Domain or Ptm</p></li>
</ul>
<p>Additional arguments can be isoform and typ (modification type).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate_df">
<span class="sig-name descname"><span class="pre">translate_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'one2one'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'one2many'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'many2many'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate_df" title="Permalink to this definition">§</a></dt>
<dd><p>Translate columns in a data frame.</p>
<dl class="simple">
<dt>Args</dt><dd><dl class="simple">
<dt>df:</dt><dd><p>A data frame.</p>
</dd>
<dt>cols:</dt><dd><p>One or more columns to be translated. It can be a single
column name, an iterable of column names or a dict where
keys are column names and values are ID types. Except this
last case, identifiers are assumed to be UniProt.</p>
</dd>
<dt>source:</dt><dd><p>The source organism.</p>
</dd>
<dt>homologene:</dt><dd><p>Use NCBI HomoloGene data for ortholog lookup.</p>
</dd>
<dt>ensembl:</dt><dd><p>Use Ensembl data for ortholog lookup.</p>
</dd>
<dt>ensembl_hc:</dt><dd><p>Use only high confidence orthology relations
from Ensembl. By default it is True. You can also set it
by the <cite>ensembl_hc</cite> attribute.</p>
</dd>
<dt>ensembl_types:</dt><dd><p>The Ensembl orthology relationship types
to use. Possible values are <cite>one2one</cite>, <cite>one2many</cite> and
<cite>many2many</cite>. By default only <cite>one2one</cite> is used. You can
also set this parameter by the <cite>ensembl_types</cite> attribute.</p>
</dd>
<dt>kwargs:</dt><dd><p>Same as providing a dict to <code class="docutils literal notranslate"><span class="pre">cols</span></code>, but beware, keys
(column names) can not match existing argument names of
this function.</p>
</dd>
</dl>
</dd>
<dt>Return</dt><dd><p>A data frame with the same column layout as the input, and the
identifiers translated as demanded. Rows that could not be
translated are omitted.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate_site">
<span class="sig-name descname"><span class="pre">translate_site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">res</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isoform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">typ</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'phosphorylation'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_taxon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.enz_sub.EnzymeSubstrateHomologyProcessor.translate_site" title="Permalink to this definition">§</a></dt>
<dd><p>Translates one PTM site.</p>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="pypath.core.enz_sub.EnzymeSubstrateAggregator.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">pypath.core.enz_sub.EnzymeSubstrateAggregator</p>
      </div>
    </a>
    <a class="right-next"
       href="pypath.core.enz_sub.EnzymeSubstrateProcessor.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">pypath.core.enz_sub.EnzymeSubstrateProcessor</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../index.html">Welcome to pypath!</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../webservice.html">Webservice</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../releasehistory.html">Release history</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../reference.html">Reference</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../notebooks/manual.html">The pypath book</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">
  <p class="copyright">
    
      © Copyright 2023, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Melih Darcan, Erva Ulusoy, Tennur Kilic, Omer Kaan Vural, Ahmet Rifaioglu, Charles Tapley Hoyt.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">
  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.0.0.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item"><p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.13.3.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>