
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>pypath.core.annot.IcellnetComplex &#8212; pypath 0.16.9 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=77553476" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/css/pypathdocs.css?v=b64fc3cb" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../_static/documentation_options.js?v=fa0329c2"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api/pypath.core.annot.IcellnetComplex';</script>
    <link rel="icon" href="../_static/omnipath-logo.svg"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="pypath.core.annot.Integrins" href="pypath.core.annot.Integrins.html" />
    <link rel="prev" title="pypath.core.annot.Icellnet" href="pypath.core.annot.Icellnet.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../contents.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/omnipath-logo.svg" class="logo__image only-light" alt="Pypath"/>
    <script>document.write(`<img src="../_static/omnipath-logo.svg" class="logo__image only-dark" alt="Pypath"/>`);</script>
  
  
    <p class="title logo__title">Pypath</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Twitter</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">PyPI</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">OmniPath</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Saez Lab</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Twitter</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">PyPI</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">OmniPath</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Saez Lab</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"></div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">





<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../contents.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../reference.html" class="nav-link">Reference</a></li>
    
    
    <li class="breadcrumb-item"><i class="fa-solid fa-ellipsis"></i></li>
    
    
    <li class="breadcrumb-item"><a href="pypath.core.annot.html" class="nav-link">pypath.core.annot</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pypath.core....</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="pypath-core-annot-icellnetcomplex">
<h1>pypath.core.annot.IcellnetComplex<a class="headerlink" href="#pypath-core-annot-icellnetcomplex" title="Link to this heading">§</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pypath.core.annot.</span></span><span class="sig-name descname"><span class="pre">IcellnetComplex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/core/annot.html#IcellnetComplex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.core.annot.IcellnetComplex" title="Link to this definition">§</a></dt>
<dd><p>Bases: <a class="reference internal" href="pypath.core.annot.AnnotationBase.html#pypath.core.annot.AnnotationBase" title="pypath.core.annot.AnnotationBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnotationBase</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/core/annot.html#IcellnetComplex.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.__init__" title="Link to this definition">§</a></dt>
<dd><p>Represents annotations for a set of proteins.
Loads the data from the original resource and provides methods
to query the annotations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – A custom name for the annotation resource.</p></li>
<li><p><strong>ncbi_tax_id</strong> (<em>int</em>) – NCBI Taxonomy identifier.</p></li>
<li><p><strong>input_method</strong> (<em>callable</em><em>,</em><em>str</em>) – Either a callable or the name of a method in any submodules of
the <code class="docutils literal notranslate"><span class="pre">pypath.inputs</span></code>  module. Should return a dict with
UniProt IDs as keys or an object suitable for <code class="docutils literal notranslate"><span class="pre">process_method</span></code>.</p></li>
<li><p><strong>input_args</strong> (<em>dict</em>) – Arguments for the <code class="docutils literal notranslate"><span class="pre">input_method</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<p class="rubric">Methods</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.__init__" title="pypath.core.annot.IcellnetComplex.__init__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__init__</span></code></a>(**kwargs)</p></td>
<td><p>Represents annotations for a set of proteins.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.add_complexes_by_inference" title="pypath.core.annot.IcellnetComplex.add_complexes_by_inference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_complexes_by_inference</span></code></a>([complexes])</p></td>
<td><p>Creates complex annotations by in silico inference and adds them to this annotation set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.all_complexes" title="pypath.core.annot.IcellnetComplex.all_complexes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_complexes</span></code></a>()</p></td>
<td><p>All protein complexes annotated in this resource.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.all_entities" title="pypath.core.annot.IcellnetComplex.all_entities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_entities</span></code></a>([entity_types])</p></td>
<td><p>All entities annotated in this resource.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.all_mirnas" title="pypath.core.annot.IcellnetComplex.all_mirnas"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_mirnas</span></code></a>()</p></td>
<td><p>All miRNAs annotated in this resource.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.all_proteins" title="pypath.core.annot.IcellnetComplex.all_proteins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_proteins</span></code></a>()</p></td>
<td><p>All UniProt IDs annotated in this resource.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.all_refs" title="pypath.core.annot.IcellnetComplex.all_refs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_refs</span></code></a>()</p></td>
<td><p>Some annotations contain references.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.annotate_complex" title="pypath.core.annot.IcellnetComplex.annotate_complex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_complex</span></code></a>(cplex)</p></td>
<td><p>Infers annotations for a single complex.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.browse" title="pypath.core.annot.IcellnetComplex.browse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">browse</span></code></a>([field, start])</p></td>
<td><p>Print gene information as a table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.complex_inference" title="pypath.core.annot.IcellnetComplex.complex_inference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">complex_inference</span></code></a>([complexes])</p></td>
<td><p>Annotates all complexes in <cite>complexes</cite>, by default in the default complex database (existing in the <cite>complex</cite> module or generated on demand according to the module's current settings).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.coverage" title="pypath.core.annot.IcellnetComplex.coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">coverage</span></code></a>(other)</p></td>
<td><p>Calculates the coverage of the annotation i.e. the proportion of entities having at least one record in this annotation resource for an arbitrary set of entities.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.curation_effort" title="pypath.core.annot.IcellnetComplex.curation_effort"><code class="xref py py-obj docutils literal notranslate"><span class="pre">curation_effort</span></code></a>()</p></td>
<td><p>Counts the reference-record pairs.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_dump</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_entity_type</span></code>(key)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.get_names" title="pypath.core.annot.IcellnetComplex.get_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_names</span></code></a>()</p></td>
<td><p>Returns the list of field names in the records.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.get_reference_set" title="pypath.core.annot.IcellnetComplex.get_reference_set"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_reference_set</span></code></a>([proteins, complexes, ...])</p></td>
<td><p>Retrieves the reference set i.e. the set of all entities which potentially have annotation in this resource.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.get_subset" title="pypath.core.annot.IcellnetComplex.get_subset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_subset</span></code></a>([method, entity_type])</p></td>
<td><p>Retrieves a subset by filtering based on <code class="docutils literal notranslate"><span class="pre">kwargs</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.get_subset_bool_array" title="pypath.core.annot.IcellnetComplex.get_subset_bool_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_subset_bool_array</span></code></a>([reference_set])</p></td>
<td><p>Returns a boolean vector with True and False values for each entity in the reference set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.get_values" title="pypath.core.annot.IcellnetComplex.get_values"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_values</span></code></a>(name[, exclude_none])</p></td>
<td><p>Returns the set of all possible values of a field.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_complexes</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_complex</span></code>(key)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_mirna</span></code>(key)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_protein</span></code>(key)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.labels" title="pypath.core.annot.IcellnetComplex.labels"><code class="xref py py-obj docutils literal notranslate"><span class="pre">labels</span></code></a>([method])</p></td>
<td><p>Same as <code class="docutils literal notranslate"><span class="pre">select</span></code> but returns a list of labels (more human readable).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.load" title="pypath.core.annot.IcellnetComplex.load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load</span></code></a>()</p></td>
<td><p>Loads the annotation data by calling the input method.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.load_data" title="pypath.core.annot.IcellnetComplex.load_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_data</span></code></a>()</p></td>
<td><p>Loads the data by calling <code class="docutils literal notranslate"><span class="pre">input_method</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.load_proteins" title="pypath.core.annot.IcellnetComplex.load_proteins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_proteins</span></code></a>()</p></td>
<td><p>Retrieves a set of all UniProt IDs to have a base set of the entire proteome.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.make_df" title="pypath.core.annot.IcellnetComplex.make_df"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_df</span></code></a>([rebuild])</p></td>
<td><p>Compiles a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> from the annotation data.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_complex_records</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_complexes</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.numof_entities" title="pypath.core.annot.IcellnetComplex.numof_entities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_entities</span></code></a>()</p></td>
<td><p>The number of annotated entities in the resource.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_mirna_records</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_mirnas</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_protein_records</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_proteins</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.numof_records" title="pypath.core.annot.IcellnetComplex.numof_records"><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_records</span></code></a>([entity_types])</p></td>
<td><p>The total number of annotation records.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.numof_references" title="pypath.core.annot.IcellnetComplex.numof_references"><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_references</span></code></a>()</p></td>
<td><p>Some annotations contain references.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.process" title="pypath.core.annot.IcellnetComplex.process"><code class="xref py py-obj docutils literal notranslate"><span class="pre">process</span></code></a>()</p></td>
<td><p>Calls the <code class="docutils literal notranslate"><span class="pre">_process_method</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">proportion</span></code>(other)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.reload" title="pypath.core.annot.IcellnetComplex.reload"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reload</span></code></a>()</p></td>
<td><p>Reloads the object from the module level.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_to_pickle</span></code>(pickle_file)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.select" title="pypath.core.annot.IcellnetComplex.select"><code class="xref py py-obj docutils literal notranslate"><span class="pre">select</span></code></a>([method, entity_type])</p></td>
<td><p>Retrieves a subset by filtering based on <code class="docutils literal notranslate"><span class="pre">kwargs</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.set_method" title="pypath.core.annot.IcellnetComplex.set_method"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_method</span></code></a>()</p></td>
<td><p>Sets the data input method by looking up in <code class="docutils literal notranslate"><span class="pre">inputs</span></code> module if necessary.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.set_reference_set" title="pypath.core.annot.IcellnetComplex.set_reference_set"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_reference_set</span></code></a>()</p></td>
<td><p>Assigns the reference set to the :py:attr``reference_set`` attribute.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.show" title="pypath.core.annot.IcellnetComplex.show"><code class="xref py py-obj docutils literal notranslate"><span class="pre">show</span></code></a>([method, table_param])</p></td>
<td><p>Same as <code class="docutils literal notranslate"><span class="pre">select</span></code> but prints a table to the console with basic information from the UniProt datasheets.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.subset_intersection" title="pypath.core.annot.IcellnetComplex.subset_intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subset_intersection</span></code></a>(universe, **kwargs)</p></td>
<td><p>Calculates the proportion of entities in a subset occuring in the set <code class="docutils literal notranslate"><span class="pre">universe</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.to_array" title="pypath.core.annot.IcellnetComplex.to_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_array</span></code></a>([reference_set, use_fields])</p></td>
<td><p>Returns an entity vs feature array.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.to_bool_array" title="pypath.core.annot.IcellnetComplex.to_bool_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_bool_array</span></code></a>(reference_set)</p></td>
<td><p>Returns a presence/absence boolean array for a reference set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.core.annot.IcellnetComplex.to_set" title="pypath.core.annot.IcellnetComplex.to_set"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_set</span></code></a>()</p></td>
<td><p>Returns the entities present in this annotation resource as a set.</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_fields</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">summary</span></code></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.add_complexes_by_inference">
<span class="sig-name descname"><span class="pre">add_complexes_by_inference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">complexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.add_complexes_by_inference" title="Link to this definition">§</a></dt>
<dd><p>Creates complex annotations by in silico inference and adds them
to this annotation set.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.all_complexes">
<span class="sig-name descname"><span class="pre">all_complexes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.all_complexes" title="Link to this definition">§</a></dt>
<dd><p>All protein complexes annotated in this resource.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.all_entities">
<span class="sig-name descname"><span class="pre">all_entities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.all_entities" title="Link to this definition">§</a></dt>
<dd><p>All entities annotated in this resource.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.all_mirnas">
<span class="sig-name descname"><span class="pre">all_mirnas</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.all_mirnas" title="Link to this definition">§</a></dt>
<dd><p>All miRNAs annotated in this resource.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.all_proteins">
<span class="sig-name descname"><span class="pre">all_proteins</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.all_proteins" title="Link to this definition">§</a></dt>
<dd><p>All UniProt IDs annotated in this resource.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.all_refs">
<span class="sig-name descname"><span class="pre">all_refs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.all_refs" title="Link to this definition">§</a></dt>
<dd><p>Some annotations contain references. The field name for references
is always <code class="docutils literal notranslate"><span class="pre">pmid</span></code> (PubMed ID). This method collects the references
across all records. Returns <em>list</em>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.annotate_complex">
<span class="sig-name descname"><span class="pre">annotate_complex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cplex</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.annotate_complex" title="Link to this definition">§</a></dt>
<dd><p>Infers annotations for a single complex.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.browse">
<span class="sig-name descname"><span class="pre">browse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">field</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.browse" title="Link to this definition">§</a></dt>
<dd><p>Print gene information as a table.</p>
<p>Presents information about annotation categories as ascii tables
printed in the terminal. If one category provided, prints one table.
If multiple categories provided, prints a table for each of them one
by one proceeding to the next one once you hit return. If no categories
provided goes through all levels of the primary category.</p>
<dl class="simple">
<dt>Args</dt><dd><dl class="simple">
<dt>field:</dt><dd><p>The field to browse categories by.</p>
<ul class="simple">
<li><p>If None the primary field will be selected.
If this annotation resource doesn’t have fields, all proteins
will be presented as one single category.</p></li>
<li><p>If a string it will be considered a field name and it will
browse through all levels of this field.</p></li>
<li><p>If a <code class="docutils literal notranslate"><span class="pre">list</span></code>, set or tuple, it will be considered either a
<code class="docutils literal notranslate"><span class="pre">list</span></code> of field names or a list of values from the primary
field. In the former case all combinations of the values of
the fields will be presented, in the latter case the browsing
will be limited to the levels of the primary field contained
in <code class="docutils literal notranslate"><span class="pre">field</span></code>.</p></li>
<li><p>If a <code class="docutils literal notranslate"><span class="pre">dict</span></code>, keys are supposed to be field names and values
as list of levels. If any of the values are None, all levels
from that field will be used.</p></li>
</ul>
</dd>
<dt>start:</dt><dd><p>Start browsing from this category. E.g. if there are 500
categories and start is 250 it will skip everything before the
250th.</p>
</dd>
<dt>kwargs:</dt><dd><p>Passed to <code class="docutils literal notranslate"><span class="pre">pypath.utils.uniprot.info</span></code>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.complex_inference">
<span class="sig-name descname"><span class="pre">complex_inference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">complexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.complex_inference" title="Link to this definition">§</a></dt>
<dd><p>Annotates all complexes in <cite>complexes</cite>, by default in the default
complex database (existing in the <cite>complex</cite> module or generated
on demand according to the module’s current settings).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>Dict with complexes as keys and sets of annotations as values.</em></p></li>
<li><p><em>Complexes with no valid information in this annotation resource</em></p></li>
<li><p><em>won’t be in the dict.</em></p></li>
</ul>
</p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>complexes</strong> (<em>iterable</em>) – Iterable yielding complexes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.coverage">
<span class="sig-name descname"><span class="pre">coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.coverage" title="Link to this definition">§</a></dt>
<dd><p>Calculates the coverage of the annotation i.e. the proportion of
entities having at least one record in this annotation resource
for an arbitrary set of entities.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.curation_effort">
<span class="sig-name descname"><span class="pre">curation_effort</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.curation_effort" title="Link to this definition">§</a></dt>
<dd><p>Counts the reference-record pairs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.get_names">
<span class="sig-name descname"><span class="pre">get_names</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.get_names" title="Link to this definition">§</a></dt>
<dd><p>Returns the list of field names in the records. The annotation
consists of uniform records and each entity might be annotated
with one or more records. Each record is a tuple of fields, for
example <code class="docutils literal notranslate"><span class="pre">('cell_type',</span> <span class="pre">'expression_level',</span> <span class="pre">'score')</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.get_reference_set">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_reference_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">proteins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">complexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_complexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbi_tax_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9606</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">swissprot_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.get_reference_set" title="Link to this definition">§</a></dt>
<dd><p>Retrieves the reference set i.e. the set of all entities which
potentially have annotation in this resource. Typically this is the
proteome of the organism from UniProt optionally with all the protein
complexes from the complex database.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.get_subset">
<span class="sig-name descname"><span class="pre">get_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.get_subset" title="Link to this definition">§</a></dt>
<dd><p>Retrieves a subset by filtering based on <code class="docutils literal notranslate"><span class="pre">kwargs</span></code>.
Each argument should be a name and a value or set of values.
Elements having the provided values in the annotation will be
returned.
Returns a set of UniProt IDs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.get_subset_bool_array">
<span class="sig-name descname"><span class="pre">get_subset_bool_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.get_subset_bool_array" title="Link to this definition">§</a></dt>
<dd><p>Returns a boolean vector with True and False values for each entity
in the reference set. The values represent presence absence data
in the simplest case, but by providing <code class="docutils literal notranslate"><span class="pre">kwargs</span></code> any kind of matching
and filtering is possible. <code class="docutils literal notranslate"><span class="pre">kwargs</span></code> are passed to the <code class="docutils literal notranslate"><span class="pre">select</span></code>
method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.get_values">
<span class="sig-name descname"><span class="pre">get_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_none</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.get_values" title="Link to this definition">§</a></dt>
<dd><p>Returns the set of all possible values of a field. E.g. if the
records of this annotation have a field <code class="docutils literal notranslate"><span class="pre">cell_type</span></code> then calling
this method can tell you that across all records the values of
this field might be <code class="docutils literal notranslate"><span class="pre">{'macrophage',</span> <span class="pre">'epithelial_cell',</span> <span class="pre">...}</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.labels">
<span class="sig-name descname"><span class="pre">labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.labels" title="Link to this definition">§</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">select</span></code> but returns a list of labels (more human readable).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.load">
<span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.load" title="Link to this definition">§</a></dt>
<dd><p>Loads the annotation data by calling the input method.
Infers annotations for complexes in the complex database if
py:attr:<code class="docutils literal notranslate"><span class="pre">infer_complexes</span></code> is True.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.load_data">
<span class="sig-name descname"><span class="pre">load_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.load_data" title="Link to this definition">§</a></dt>
<dd><p>Loads the data by calling <code class="docutils literal notranslate"><span class="pre">input_method</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.load_proteins">
<span class="sig-name descname"><span class="pre">load_proteins</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.load_proteins" title="Link to this definition">§</a></dt>
<dd><p>Retrieves a set of all UniProt IDs to have a base set of the entire
proteome.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.make_df">
<span class="sig-name descname"><span class="pre">make_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rebuild</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.make_df" title="Link to this definition">§</a></dt>
<dd><p>Compiles a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> from the annotation data.
The data frame will be assigned to :py:attr``df``.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.numof_entities">
<span class="sig-name descname"><span class="pre">numof_entities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.numof_entities" title="Link to this definition">§</a></dt>
<dd><p>The number of annotated entities in the resource.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.numof_records">
<span class="sig-name descname"><span class="pre">numof_records</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.numof_records" title="Link to this definition">§</a></dt>
<dd><p>The total number of annotation records.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.numof_references">
<span class="sig-name descname"><span class="pre">numof_references</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.numof_references" title="Link to this definition">§</a></dt>
<dd><p>Some annotations contain references. The field name for references
is always <code class="docutils literal notranslate"><span class="pre">pmid</span></code> (PubMed ID). This method collects and counts the
references across all records.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.process">
<span class="sig-name descname"><span class="pre">process</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.process" title="Link to this definition">§</a></dt>
<dd><p>Calls the <code class="docutils literal notranslate"><span class="pre">_process_method</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.reload">
<span class="sig-name descname"><span class="pre">reload</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.reload" title="Link to this definition">§</a></dt>
<dd><p>Reloads the object from the module level.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.select">
<span class="sig-name descname"><span class="pre">select</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.select" title="Link to this definition">§</a></dt>
<dd><p>Retrieves a subset by filtering based on <code class="docutils literal notranslate"><span class="pre">kwargs</span></code>.
Each argument should be a name and a value or set of values.
Elements having the provided values in the annotation will be
returned.
Returns a set of UniProt IDs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.set_method">
<span class="sig-name descname"><span class="pre">set_method</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.set_method" title="Link to this definition">§</a></dt>
<dd><p>Sets the data input method by looking up in <code class="docutils literal notranslate"><span class="pre">inputs</span></code> module if
necessary.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.set_reference_set">
<span class="sig-name descname"><span class="pre">set_reference_set</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.set_reference_set" title="Link to this definition">§</a></dt>
<dd><p>Assigns the reference set to the :py:attr``reference_set`` attribute.
The reference set is the set of all entities which
potentially have annotation in this resource. Typically this is the
proteome of the organism from UniProt optionally with all the protein
complexes from the complex database.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.show">
<span class="sig-name descname"><span class="pre">show</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.show" title="Link to this definition">§</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">select</span></code> but prints a table to the console with basic
information from the UniProt datasheets.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.subset_intersection">
<span class="sig-name descname"><span class="pre">subset_intersection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">universe</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.subset_intersection" title="Link to this definition">§</a></dt>
<dd><p>Calculates the proportion of entities in a subset occuring in the
set <code class="docutils literal notranslate"><span class="pre">universe</span></code>. The subset is selected by passing <code class="docutils literal notranslate"><span class="pre">kwargs</span></code> to the
<code class="docutils literal notranslate"><span class="pre">select</span></code> method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.to_array">
<span class="sig-name descname"><span class="pre">to_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_fields</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.to_array" title="Link to this definition">§</a></dt>
<dd><p>Returns an entity vs feature array. In case of more complex
annotations this might be huge.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.to_bool_array">
<span class="sig-name descname"><span class="pre">to_bool_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_set</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.to_bool_array" title="Link to this definition">§</a></dt>
<dd><p>Returns a presence/absence boolean array for a reference set.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.core.annot.IcellnetComplex.to_set">
<span class="sig-name descname"><span class="pre">to_set</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.core.annot.IcellnetComplex.to_set" title="Link to this definition">§</a></dt>
<dd><p>Returns the entities present in this annotation resource as a set.</p>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="pypath.core.annot.Icellnet.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">pypath.core.annot.Icellnet</p>
      </div>
    </a>
    <a class="right-next"
       href="pypath.core.annot.Integrins.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">pypath.core.annot.Integrins</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../index.html">Welcome to pypath!</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../webservice.html">Webservice</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../releasehistory.html">Release history</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../reference.html">Reference</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../notebooks/manual.html">The pypath book</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Melih Darcan, Erva Ulusoy, Tennur Kilic, Omer Kaan Vural, Ahmet Rifaioglu, Charles Tapley Hoyt.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>