

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>pypath.legacy.main.PyPath &#8212; pypath 0.14.50 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/pypathdocs.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api/pypath.legacy.main.PyPath';</script>
    <link rel="icon" href="../_static/omnipath-logo.svg"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="pypath.omnipath" href="pypath.omnipath.html" />
    <link rel="prev" title="pypath.legacy.main.NetworkStatsRecord" href="pypath.legacy.main.NetworkStatsRecord.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  <div class="navbar-header-items__start">
    
      <div class="navbar-item">
  

<a class="navbar-brand logo" href="../contents.html">
  
  
  
  
    
    
      
    
    
    <img src="../_static/omnipath-logo.svg" class="logo__image only-light" alt="Pypath"/>
    <script>document.write(`<img src="../_static/omnipath-logo.svg" class="logo__image only-dark" alt="Pypath"/>`);</script>
  
  
    <p class="title logo__title">Pypath</p>
  
</a></div>
    
  </div>
  
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          
<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
        </div>
      
      
        <div class="navbar-item">
<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter"></i></span>
            <label class="sr-only">Twitter</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box"></i></span>
            <label class="sr-only">PyPI</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university"></i></span>
            <label class="sr-only">OmniPath</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users"></i></span>
            <label class="sr-only">Saez Lab</label></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">
<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../notebooks/manual.html">
                        The pypath book
                      </a>
                    </li>
                
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter"></i></span>
            <label class="sr-only">Twitter</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-box"></i></span>
            <label class="sr-only">PyPI</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-university"></i></span>
            <label class="sr-only">OmniPath</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-solid fa-users"></i></span>
            <label class="sr-only">Saez Lab</label></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">





<nav aria-label="Breadcrumbs">
  <ul class="bd-breadcrumbs" role="navigation" aria-label="Breadcrumb">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../contents.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../reference.html" class="nav-link">Reference</a></li>
    
    
    <li class="breadcrumb-item"><i class="fa-solid fa-ellipsis"></i></li>
    
    
    <li class="breadcrumb-item"><a href="pypath.legacy.main.html" class="nav-link">pypath.legacy.main</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">pypath.legacy.main.PyPath</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="pypath-legacy-main-pypath">
<h1>pypath.legacy.main.PyPath<a class="headerlink" href="#pypath-legacy-main-pypath" title="Permalink to this heading">§</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pypath.legacy.main.</span></span><span class="sig-name descname"><span class="pre">PyPath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ncbi_tax_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'unnamed'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'results'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loglevel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loops</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath" title="Permalink to this definition">§</a></dt>
<dd><p>Bases: <a class="reference internal" href="pypath.share.session.Logger.html#pypath.share.session.Logger" title="pypath.share.session.Logger"><code class="xref py py-class docutils literal notranslate"><span class="pre">Logger</span></code></a></p>
<p>This is the a <em>legacy</em> object representing a molecular interaction
network. At some point it will be removed, we don’t recommend to rely
on it when you build your applications.
The <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.network.Network</span></code> object offers a much clearer and
more versatile API. As of end of 2019 not all functionalities have been
migrated to the new API. For this reason we offer an intermediate
solution: in this <cite>igraph</cite> based object the <cite>attrs</cite> edge attribute
holds instances of <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.interaction.Interaction</span></code> objects,
the same type of object we use to represent interactions in the new
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.network.Network</span></code>.
At the same time we will keep supporting <cite>igraph</cite> with a method for
converting <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.network.Network</span></code> to a
<code class="xref py py-class docutils literal notranslate"><span class="pre">igraph.Graph</span></code> object, however this won’t provide all the
methods available here but will serve only the purpose to make it
possible to use the graph theory methods from the <cite>igraph</cite> library
on networks built with <cite>pypath</cite>.</p>
<p>An object representing a molecular interaction network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ncbi_tax_id</strong> (<em>int</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">9606</span></code> (Homo sapiens) by default. NCBI Taxonomic
identifier of the organism from which the data will be
downloaded.</p></li>
<li><p><strong>default_name_type</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">{'protein':</span> <span class="pre">'uniprot',</span> <span class="pre">'mirna':</span> <span class="pre">'mirbase',</span> <span class="pre">'drug':</span>
<span class="pre">'chembl',</span> <span class="pre">'lncrna':</span> <span class="pre">'lncrna-genesymbol'}</span></code> by default. Contains
the default identifier types to which the downloaded data will
be converted. If others are used, user may need to provide the
format definitions for the conversion tables.</p></li>
<li><p><strong>copy</strong> (<em>pypath.main.PyPath</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Other
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.main.PyPath</span></code> instance from which the data will
be copied.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'unnamed'</span></code> by default. Session or project name
(custom).</p></li>
<li><p><strong>outdir</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'results'</span></code> by default. Output directory where to
store all output files.</p></li>
<li><p><strong>loglevel</strong> (<em>int</em>) – Optional, 0 by default. Sets the level of the logger.
The higher the level the more messages will be written to the log.</p></li>
<li><p><strong>loops</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Determines if self-loop edges
are allowed in the graph.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>adjlist</strong> (<em>list</em>) – List of [set] containing the adjacency of each node. See
<a class="reference internal" href="#pypath.legacy.main.PyPath.update_adjlist" title="pypath.legacy.main.PyPath.update_adjlist"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PyPath.update_adjlist()</span></code></a> method for more information.</p></li>
<li><p><strong>chembl</strong> (<em>pypath.chembl.Chembl</em>) – Contains the ChEMBL data. See <code class="xref py py-mod docutils literal notranslate"><span class="pre">pypath.chembl</span></code> module
documentation for more information.</p></li>
<li><p><strong>chembl_mysql</strong> (<em>tuple</em>) – DEPRECATED Contains the MySQL parameters used by the
<code class="xref py py-mod docutils literal notranslate"><span class="pre">pypath.mapping</span></code> module to load the ChEMBL ID conversion
tables.</p></li>
<li><p><strong>data</strong> (<em>dict</em>) – Stores the loaded interaction and attribute table. See
<code class="xref py py-meth docutils literal notranslate"><span class="pre">PyPath.read_data_file()</span></code> method for more information.</p></li>
<li><p><strong>db_dict</strong> (<em>dict</em>) – Dictionary of dictionaries. Outer-level keys are <code class="docutils literal notranslate"><span class="pre">'nodes'</span></code> and
<code class="docutils literal notranslate"><span class="pre">'edges'</span></code>, corresponding values are [dict] whose keys are the
database sources with values of type [set] containing the
edge/node indexes for which that database provided some
information.</p></li>
<li><p><strong>dgraph</strong> (<em>igraph.Graph</em>) – Directed network graph object.</p></li>
<li><p><strong>disclaimer</strong> (<em>str</em>) – Disclaimer text.</p></li>
<li><p><strong>dlabDct</strong> (<em>dict</em>) – Maps the directed graph node labels [str] (keys) to their
indices [int] (values).</p></li>
<li><p><strong>dnodDct</strong> (<em>dict</em>) – Maps the directed graph node names [str] (keys) to their indices
[int] (values).</p></li>
<li><p><strong>dnodInd</strong> (<em>set</em>) – Stores the directed graph node names [str].</p></li>
<li><p><strong>dnodLab</strong> (<em>dict</em>) – Maps the directed graph node indices [int] (keys) to their
labels [str] (values).</p></li>
<li><p><strong>dnodNam</strong> (<em>dict</em>) – Maps the directed graph node indices [int] (keys) to their names
[str] (values).</p></li>
<li><p><strong>edgeAttrs</strong> (<em>dict</em>) – Stores the edge attribute names [str] as keys and their
corresponding types (e.g.: <code class="docutils literal notranslate"><span class="pre">set</span></code>, <code class="docutils literal notranslate"><span class="pre">list</span></code>, <code class="docutils literal notranslate"><span class="pre">str</span></code>, …) as
values.</p></li>
<li><p><strong>exp</strong> (<em>pandas.DataFrame</em>) – Stores the expression data for the nodes (if loaded).</p></li>
<li><p><strong>exp_prod</strong> (<em>pandas.DataFrame</em>) – Stores the edge expression data (as the product of the
normalized expression between the pair of nodes by default). For
more details see <code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.edges_expression()</span></code>.</p></li>
<li><p><strong>exp_samples</strong> (<em>set</em>) – Contains a list of tissues as downloaded by ProteomicsDB. See
<a class="reference internal" href="#pypath.legacy.main.PyPath.get_proteomicsdb" title="pypath.legacy.main.PyPath.get_proteomicsdb"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PyPath.get_proteomicsdb()</span></code></a> for more information.</p></li>
<li><p><strong>failed_edges</strong> (<em>list</em>) – List of lists containing information about the failed edges.
Each failed edge sublist contains (in this order): [tuple] with
the node IDs, [str] names of nodes A and B, [int] IDs of nodes
A and B and [int] IDs of the edges in both directions.</p></li>
<li><p><strong>go</strong> (<em>dict</em>) – Contains the organism(s) NCBI taxonomy ID as key [int] and
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.go.GOAnnotation</span></code> object as value, which
contains the GO annotations for the nodes in the graph. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.go.GOAnnotation</span></code> for more information.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Undirected network graph object.</p></li>
<li><p><strong>gsea</strong> (<em>pypath.gsea.GSEA</em>) – Contains the loaded gene-sets from MSigDB. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.gsea.GSEA</span></code> for more information.</p></li>
<li><p><strong>has_cats</strong> (<em>set</em>) – Contains the categories (e.g.: resource types) [str] loaded in
the current network. Possible categories are: <code class="docutils literal notranslate"><span class="pre">'m'</span></code> for
PTM/enzyme-substrate resources, <code class="docutils literal notranslate"><span class="pre">'p'</span></code> for pathway/activity
flow resources, <code class="docutils literal notranslate"><span class="pre">'i'</span></code> for undirected/PPI resources, <code class="docutils literal notranslate"><span class="pre">'r'</span></code>
for process description/reaction resources and <code class="docutils literal notranslate"><span class="pre">'t'</span></code> for
transcription resources.</p></li>
<li><p><strong>htp</strong> (<em>dict</em>) – Contains information about high-throughput data of the network
for different thresholds [int] (keys). Values are [dict]
containing the number of references (<code class="docutils literal notranslate"><span class="pre">'rnum'</span></code>) [int], number
of edges (<code class="docutils literal notranslate"><span class="pre">'enum'</span></code>) [int], number of sources (<code class="docutils literal notranslate"><span class="pre">'snum'</span></code>)
[int] and list of PMIDs of the most common references above the
given threshold (<code class="docutils literal notranslate"><span class="pre">'htrefs'</span></code>) [set].</p></li>
<li><p><strong>labDct</strong> (<em>dict</em>) – Maps the undirected graph node labels [str] (keys) to their
indices [int] (values).</p></li>
<li><p><strong>lists</strong> (<em>dict</em>) – Contains specific lists of nodes (values) for different
categories [str] (keys). These can to be loaded from a file or
a resource. Some methods include
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.receptor_list()</span></code> (<code class="docutils literal notranslate"><span class="pre">'rec'</span></code>),
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.druggability_list()</span></code> (<code class="docutils literal notranslate"><span class="pre">'dgb'</span></code>),
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.kinases_list()</span></code> (<code class="docutils literal notranslate"><span class="pre">'kin'</span></code>),
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.tfs_list()</span></code> (<code class="docutils literal notranslate"><span class="pre">'tf'</span></code>),
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.disease_genes_list()</span></code> (<code class="docutils literal notranslate"><span class="pre">'dis'</span></code>),
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.signaling_proteins_list()</span></code>
(<code class="docutils literal notranslate"><span class="pre">'sig'</span></code>), <code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.proteome_list()</span></code>
(<code class="docutils literal notranslate"><span class="pre">'proteome'</span></code>) and
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.cancer_drivers_list()</span></code> (<code class="docutils literal notranslate"><span class="pre">'cdv'</span></code>).</p></li>
<li><p><strong>loglevel</strong> (<em>str</em>) – The level of the logger.</p></li>
<li><p><strong>loops</strong> (<em>bool</em>) – Whether if self-loop edges are allowed in the graph.</p></li>
<li><p><strong>mapper</strong> (<em>pypath.mapping.Mapper</em>) – <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.mapper.Mapper</span></code> object for ID conversion and
other ID-related operations across resources.</p></li>
<li><p><strong>mutation_samples</strong> (<em>list</em>) – DEPRECATED</p></li>
<li><p><strong>mysql_conf</strong> (<em>tuple</em>) – DEPRECATED Contains the MySQL parameters used by the
<code class="xref py py-mod docutils literal notranslate"><span class="pre">pypath.mapping</span></code> module to load the ID conversion
tables.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Session or project name (custom).</p></li>
<li><p><strong>ncbi_tax_id</strong> (<em>int</em>) – NCBI Taxonomic identifier of the organism from which the data
will be downloaded.</p></li>
<li><p><strong>negatives</strong> (<em>dict</em>) – Contains a list of negative interactions according to a given
source (e.g.: Negatome database). See
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.apply_negative()</span></code> for more
information.</p></li>
<li><p><strong>nodDct</strong> (<em>dict</em>) – Maps the undirected graph node names [str] (keys) to their
indices [int] (values).</p></li>
<li><p><strong>nodInd</strong> (<em>set</em>) – Stores the undirected graph node names [str].</p></li>
<li><p><strong>nodLab</strong> (<em>dict</em>) – Maps the undirected graph node indices [int] (keys) to their
labels [str] (values).</p></li>
<li><p><strong>nodNam</strong> (<em>dict</em>) – Maps the directed graph node indices [int] (keys) to their names
[str] (values).</p></li>
<li><p><strong>outdir</strong> (<em>str</em>) – Output directory where to store all output files.</p></li>
<li><p><strong>palette</strong> (<em>list</em>) – Contains a list of hexadecimal [str] of colors. Used for
plotting purposes.</p></li>
<li><p><strong>pathway_types</strong> (<em>list</em>) – Contains the names of all the loaded pathway resources [str].</p></li>
<li><p><strong>pathways</strong> (<em>dict</em>) – Contains the list of pathways (values) for each resource (keys)
loaded in the network.</p></li>
<li><p><strong>plots</strong> (<em>dict</em>) – DEPRECATED (?)</p></li>
<li><p><strong>proteomicsdb</strong> (<em>pypath.proteomicsdb.ProteomicsDB</em>) – Contains a <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.proteomicsdb.ProteomicsDB</span></code>
instance, see the class documentation for more information.</p></li>
<li><p><strong>raw_data</strong> (<em>list</em>) – Contains a list of loaded edges [dict] from a data file. See
<code class="xref py py-meth docutils literal notranslate"><span class="pre">PyPath.read_data_file()</span></code> for more information.</p></li>
<li><p><strong>seq</strong> (<em>dict</em>) – (?)</p></li>
<li><p><strong>session</strong> (<em>str</em>) – Session ID, a five random alphanumeric characters.</p></li>
<li><p><strong>session_name</strong> (<em>str</em>) – Session name and ID (e.g. <code class="docutils literal notranslate"><span class="pre">'unnamed-abc12'</span></code>).</p></li>
<li><p><strong>sourceNetEdges</strong> (<em>igraph.Graph</em>) – (?)</p></li>
<li><p><strong>sourceNetNodes</strong> (<em>igraph.Graph</em>) – (?)</p></li>
<li><p><strong>sources</strong> (<em>list</em>) – List contianing the names of the loaded resources [str].</p></li>
<li><p><strong>u_pfam</strong> (<em>dict</em>) – Dictionary of dictionaries, contains the mapping of UniProt IDs
to their respective protein families and other information.</p></li>
<li><p><strong>uniprot_mapped</strong> (<em>list</em>) – DEPRECATED (?)</p></li>
<li><p><strong>unmapped</strong> (<em>list</em>) – Contains the names of unmapped items [str]. See
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.map_item()</span></code> for more information.</p></li>
<li><p><strong>vertexAttrs</strong> (<em>dict</em>) – Stores the node attribute names [str] as keys and their
corresponding types (e.g.: <code class="docutils literal notranslate"><span class="pre">set</span></code>, <code class="docutils literal notranslate"><span class="pre">list</span></code>, <code class="docutils literal notranslate"><span class="pre">str</span></code>, …) as
values.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ncbi_tax_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'unnamed'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'results'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loglevel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loops</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.__init__" title="Permalink to this definition">§</a></dt>
<dd><p>Initializes the network object.</p>
<p><strong>NOTE:</strong> Only the instance is created, no data is donwloaded
until the corresponding function is called (e.g.:
<a class="reference internal" href="#pypath.legacy.main.PyPath.init_network" title="pypath.legacy.main.PyPath.init_network"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PyPath.init_network()</span></code></a>).</p>
</dd></dl>

<p class="rubric">Methods</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.__init__" title="pypath.legacy.main.PyPath.__init__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__init__</span></code></a>([ncbi_tax_id, copy, name, ...])</p></td>
<td><p>Initializes the network object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.acsn_effects" title="pypath.legacy.main.PyPath.acsn_effects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">acsn_effects</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.add_genesets" title="pypath.legacy.main.PyPath.add_genesets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_genesets</span></code></a>(genesets)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.add_grouped_eattr" title="pypath.legacy.main.PyPath.add_grouped_eattr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_grouped_eattr</span></code></a>(edge, attr, group, value)</p></td>
<td><p>Merges (or creates) a given edge attribute as [dict] of [list] values.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.add_grouped_set_eattr" title="pypath.legacy.main.PyPath.add_grouped_set_eattr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_grouped_set_eattr</span></code></a>(edge, attr, group, value)</p></td>
<td><p>Merges (or creates) a given edge attribute as [dict] of [set] values.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.add_list_eattr" title="pypath.legacy.main.PyPath.add_list_eattr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_list_eattr</span></code></a>(edge, attr, value)</p></td>
<td><p>Merges (or creates) a given edge attribute as [list].</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.add_set_eattr" title="pypath.legacy.main.PyPath.add_set_eattr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_set_eattr</span></code></a>(edge, attr, value)</p></td>
<td><p>Merges (or creates) a given edge attribute as [set].</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">affected_by</span></code>(identifier)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.affects" title="pypath.legacy.main.PyPath.affects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">affects</span></code></a>(identifier)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.all_between" title="pypath.legacy.main.PyPath.all_between"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_between</span></code></a>(id_a, id_b)</p></td>
<td><p>Checks for any edges (in any direction) between the provided nodes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.all_neighbours" title="pypath.legacy.main.PyPath.all_neighbours"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_neighbours</span></code></a>([indices])</p></td>
<td><p>Looks for the first neighbours of all the nodes and creates an attribute (<code class="docutils literal notranslate"><span class="pre">'neighbours'</span></code>) on each one of them containing a list of their UniProt IDs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.apply_list" title="pypath.legacy.main.PyPath.apply_list"><code class="xref py py-obj docutils literal notranslate"><span class="pre">apply_list</span></code></a>(name[, node_or_edge])</p></td>
<td><p>Creates vertex or edge attribute based on a list.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.apply_negative" title="pypath.legacy.main.PyPath.apply_negative"><code class="xref py py-obj docutils literal notranslate"><span class="pre">apply_negative</span></code></a>(settings)</p></td>
<td><p>Loads a negative interaction source (e.g.: Negatome) into the current network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.basic_stats" title="pypath.legacy.main.PyPath.basic_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">basic_stats</span></code></a>([latex, caption, latex_hdr, ...])</p></td>
<td><p>Returns basic numbers about the network resources, e.g.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.basic_stats_intergroup" title="pypath.legacy.main.PyPath.basic_stats_intergroup"><code class="xref py py-obj docutils literal notranslate"><span class="pre">basic_stats_intergroup</span></code></a>(groupA, groupB[, header])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.cancer_drivers_list" title="pypath.legacy.main.PyPath.cancer_drivers_list"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cancer_drivers_list</span></code></a>([intogen_file])</p></td>
<td><p>Loads the list of cancer drivers.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.cancer_gene_census_list" title="pypath.legacy.main.PyPath.cancer_gene_census_list"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cancer_gene_census_list</span></code></a>()</p></td>
<td><p>Loads the list of cancer driver proteins from the COSMIC Cancer Gene Census.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.clean_graph" title="pypath.legacy.main.PyPath.clean_graph"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clean_graph</span></code></a>([organisms_allowed])</p></td>
<td><p>Removes multiple edges, unknown molecules and those from wrong taxon.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.collapse_by_name" title="pypath.legacy.main.PyPath.collapse_by_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">collapse_by_name</span></code></a>([graph])</p></td>
<td><p>Collapses nodes with the same name by copying and merging all edges and attributes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.collect" title="pypath.legacy.main.PyPath.collect"><code class="xref py py-obj docutils literal notranslate"><span class="pre">collect</span></code></a>(method, **kwargs)</p></td>
<td><p>Collects various entities over the network according to <code class="docutils literal notranslate"><span class="pre">method</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.combine_attr" title="pypath.legacy.main.PyPath.combine_attr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">combine_attr</span></code></a>(lst[, num_method])</p></td>
<td><p>Combines multiple attributes into one.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.communities" title="pypath.legacy.main.PyPath.communities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">communities</span></code></a>(method, **kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.complex_comembership_network" title="pypath.legacy.main.PyPath.complex_comembership_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">complex_comembership_network</span></code></a>([graph, resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complex_entities</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complex_entities_by_category</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complex_entities_by_resource</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.complexes" title="pypath.legacy.main.PyPath.complexes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">complexes</span></code></a>([methods])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.complexes_in_network" title="pypath.legacy.main.PyPath.complexes_in_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">complexes_in_network</span></code></a>([csource, graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.compounds_from_chembl" title="pypath.legacy.main.PyPath.compounds_from_chembl"><code class="xref py py-obj docutils literal notranslate"><span class="pre">compounds_from_chembl</span></code></a>([chembl_mysql, nodes, ...])</p></td>
<td><p>Loads compound data from ChEMBL to the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.consistency" title="pypath.legacy.main.PyPath.consistency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">consistency</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.copy" title="pypath.legacy.main.PyPath.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>(other)</p></td>
<td><p>Copies another <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.main.PyPath</span></code> instance into the current one.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.copy_edges" title="pypath.legacy.main.PyPath.copy_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_edges</span></code></a>(sources, target[, move, graph])</p></td>
<td><p>Copies edges from <em>sources</em> node(s) to another one (<em>target</em>), keeping attributes and directions.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.count_sol" title="pypath.legacy.main.PyPath.count_sol"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_sol</span></code></a>()</p></td>
<td><p>Counts the number of nodes with zero degree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.counts" title="pypath.legacy.main.PyPath.counts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts</span></code></a>(collection_method[, add_total, ...])</p></td>
<td><p>Collects various entities over the network according to <code class="docutils literal notranslate"><span class="pre">method</span></code> and counts them in total and by resources.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.coverage" title="pypath.legacy.main.PyPath.coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">coverage</span></code></a>(lst)</p></td>
<td><p>Computes the coverage (range [0, 1]) of a list of nodes against the current (undirected) network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.curation_effort" title="pypath.legacy.main.PyPath.curation_effort"><code class="xref py py-obj docutils literal notranslate"><span class="pre">curation_effort</span></code></a>([resources])</p></td>
<td><p>Returns a <em>set</em> of reference-interactions pairs.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">curation_effort_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.curation_effort_by_resource" title="pypath.legacy.main.PyPath.curation_effort_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">curation_effort_by_resource</span></code></a>([resources])</p></td>
<td><p>A <em>dict</em> with resources as keys and <a href="#id1"><span class="problematic" id="id2">*</span></a>set*s of curation items (interaction-reference pairs) as values.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">curation_effort_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.curation_stats" title="pypath.legacy.main.PyPath.curation_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">curation_stats</span></code></a>([by_category])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.curation_tab" title="pypath.legacy.main.PyPath.curation_tab"><code class="xref py py-obj docutils literal notranslate"><span class="pre">curation_tab</span></code></a>([fname, by_category, use_cats, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.curators_work" title="pypath.legacy.main.PyPath.curators_work"><code class="xref py py-obj docutils literal notranslate"><span class="pre">curators_work</span></code></a>()</p></td>
<td><p>Computes and prints an estimation of how many years of curation took to achieve the amount of information on the network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.databases_similarity" title="pypath.legacy.main.PyPath.databases_similarity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">databases_similarity</span></code></a>([index])</p></td>
<td><p>Computes the similarity across databases according to a given index metric.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.degree_dist" title="pypath.legacy.main.PyPath.degree_dist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degree_dist</span></code></a>(prefix[, g, group])</p></td>
<td><p>Computes the degree distribution over all nodes of the network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.degree_dists" title="pypath.legacy.main.PyPath.degree_dists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degree_dists</span></code></a>()</p></td>
<td><p>Computes the degree distribution for all the different network sources.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.delete_by_organism" title="pypath.legacy.main.PyPath.delete_by_organism"><code class="xref py py-obj docutils literal notranslate"><span class="pre">delete_by_organism</span></code></a>([organisms_allowed])</p></td>
<td><p>Removes the proteins of all organisms which are not given in <em>tax</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.delete_by_source" title="pypath.legacy.main.PyPath.delete_by_source"><code class="xref py py-obj docutils literal notranslate"><span class="pre">delete_by_source</span></code></a>(source[, vertexAttrsToDel, ...])</p></td>
<td><p>Deletes nodes and edges from the network according to a provided source name.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.delete_unknown" title="pypath.legacy.main.PyPath.delete_unknown"><code class="xref py py-obj docutils literal notranslate"><span class="pre">delete_unknown</span></code></a>([organisms_allowed, ...])</p></td>
<td><p>Removes those items which are not in the list of all default IDs of the organisms.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.delete_unmapped" title="pypath.legacy.main.PyPath.delete_unmapped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">delete_unmapped</span></code></a>()</p></td>
<td><p>Checks the network for any existing unmapped node and removes it.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dgenesymbol" title="pypath.legacy.main.PyPath.dgenesymbol"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dgenesymbol</span></code></a>(genesymbol)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol can be found in the default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dgenesymbols" title="pypath.legacy.main.PyPath.dgenesymbols"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dgenesymbols</span></code></a>(genesymbols)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dgs" title="pypath.legacy.main.PyPath.dgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dgs</span></code></a>(genesymbol)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol can be found in the default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dgss" title="pypath.legacy.main.PyPath.dgss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dgss</span></code></a>(genesymbols)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dneighbors" title="pypath.legacy.main.PyPath.dneighbors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dneighbors</span></code></a>(identifier[, mode])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dp" title="pypath.legacy.main.PyPath.dp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dp</span></code></a>(identifier)</p></td>
<td><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dproteins" title="pypath.legacy.main.PyPath.dproteins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dproteins</span></code></a>(identifiers)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dps" title="pypath.legacy.main.PyPath.dps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dps</span></code></a>(identifiers)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.duniprot" title="pypath.legacy.main.PyPath.duniprot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">duniprot</span></code></a>(uniprot)</p></td>
<td><p>Same as <a href="#id3"><span class="problematic" id="id4">``</span></a>PyPath.uniprot(), just for directed graph.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.duniprots" title="pypath.legacy.main.PyPath.duniprots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">duniprots</span></code></a>(uniprots)</p></td>
<td><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object for a list of UniProt IDs omitting those could not be found in the default directed graph.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dup" title="pypath.legacy.main.PyPath.dup"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dup</span></code></a>(uniprot)</p></td>
<td><p>Same as <a href="#id5"><span class="problematic" id="id6">``</span></a>PyPath.uniprot(), just for directed graph.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dups" title="pypath.legacy.main.PyPath.dups"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dups</span></code></a>(uniprots)</p></td>
<td><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object for a list of UniProt IDs omitting those could not be found in the default directed graph.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dv" title="pypath.legacy.main.PyPath.dv"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dv</span></code></a>(identifier)</p></td>
<td><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.dvs" title="pypath.legacy.main.PyPath.dvs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dvs</span></code></a>(identifiers)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edge_exists" title="pypath.legacy.main.PyPath.edge_exists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edge_exists</span></code></a>(id_a, id_b)</p></td>
<td><p>Returns a tuple of vertex indices if edge doesn't exist, otherwise, the edge ID.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edge_loc" title="pypath.legacy.main.PyPath.edge_loc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edge_loc</span></code></a>([graph, topn])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edge_names" title="pypath.legacy.main.PyPath.edge_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edge_names</span></code></a>(e)</p></td>
<td><p>Returns the node names of a given edge.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edges_3d" title="pypath.legacy.main.PyPath.edges_3d"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edges_3d</span></code></a>([methods])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edges_between" title="pypath.legacy.main.PyPath.edges_between"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edges_between</span></code></a>(group1, group2[, directed, strict])</p></td>
<td><p>Selects edges between two groups of vertex IDs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edges_expression" title="pypath.legacy.main.PyPath.edges_expression"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edges_expression</span></code></a>([func])</p></td>
<td><p>Executes function <cite>func</cite> for each pairs of connected proteins in the network, for every expression dataset.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edges_in_complexes" title="pypath.legacy.main.PyPath.edges_in_complexes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edges_in_complexes</span></code></a>([csources, graph])</p></td>
<td><p>Creates edge attributes <code class="docutils literal notranslate"><span class="pre">complexes</span></code> and <code class="docutils literal notranslate"><span class="pre">in_complex</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edges_ptms" title="pypath.legacy.main.PyPath.edges_ptms"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edges_ptms</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.edgeseq_inverse" title="pypath.legacy.main.PyPath.edgeseq_inverse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edgeseq_inverse</span></code></a>(edges)</p></td>
<td><p>Returns the sequence of all edge indexes that are not in the argument <em>edges</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">entities</span></code>([resources, entity_type])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">entities_by_category</span></code>([entity_type])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.entities_by_resource" title="pypath.legacy.main.PyPath.entities_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">entities_by_resource</span></code></a>([resources, entity_type])</p></td>
<td><p>Returns a <em>dict</em> of <a href="#id7"><span class="problematic" id="id8">*</span></a>set*s with resources as keys and sets of entities as values.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.entities_by_resources" title="pypath.legacy.main.PyPath.entities_by_resources"><code class="xref py py-obj docutils literal notranslate"><span class="pre">entities_by_resources</span></code></a>()</p></td>
<td><p>Returns a dict of sets with resources as keys and sets of entity IDs as values.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">entities_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.export_dot" title="pypath.legacy.main.PyPath.export_dot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">export_dot</span></code></a>([nodes, edges, directed, labels, ...])</p></td>
<td><p>Builds a pygraphviz.AGraph() object with filtering the edges and vertices along arbitrary criteria.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.export_edgelist" title="pypath.legacy.main.PyPath.export_edgelist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">export_edgelist</span></code></a>(fname[, graph, names, ...])</p></td>
<td><p>Write edge list to text file with attributes</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.export_graphml" title="pypath.legacy.main.PyPath.export_graphml"><code class="xref py py-obj docutils literal notranslate"><span class="pre">export_graphml</span></code></a>([outfile, graph, name])</p></td>
<td><p>Saves the network in a <code class="docutils literal notranslate"><span class="pre">.graphml</span></code> file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.export_ptms_tab" title="pypath.legacy.main.PyPath.export_ptms_tab"><code class="xref py py-obj docutils literal notranslate"><span class="pre">export_ptms_tab</span></code></a>([outfile])</p></td>
<td><p>Exports a tab file containing the PTM interaction information loaded in the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.export_sif" title="pypath.legacy.main.PyPath.export_sif"><code class="xref py py-obj docutils literal notranslate"><span class="pre">export_sif</span></code></a>([outfile])</p></td>
<td><p>Exports the network interactions in <code class="docutils literal notranslate"><span class="pre">.sif</span></code> format (Simple Interaction Format).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.export_struct_tab" title="pypath.legacy.main.PyPath.export_struct_tab"><code class="xref py py-obj docutils literal notranslate"><span class="pre">export_struct_tab</span></code></a>([outfile])</p></td>
<td><p>Exports a tab file containing the domain interaction information and PTM regulation loaded in the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.export_tab" title="pypath.legacy.main.PyPath.export_tab"><code class="xref py py-obj docutils literal notranslate"><span class="pre">export_tab</span></code></a>([outfile, extra_node_attrs, ...])</p></td>
<td><p>Exports the network in a tabular format.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.find_all_paths" title="pypath.legacy.main.PyPath.find_all_paths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_all_paths</span></code></a>(start, end[, attr, mode, ...])</p></td>
<td><p>Finds all paths up to length <cite>maxlen</cite> between groups of vertices.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.find_all_paths2" title="pypath.legacy.main.PyPath.find_all_paths2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_all_paths2</span></code></a>(graph, start, end[, mode, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.find_complex" title="pypath.legacy.main.PyPath.find_complex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_complex</span></code></a>(search)</p></td>
<td><p>Finds complexes by their non standard names.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.first_neighbours" title="pypath.legacy.main.PyPath.first_neighbours"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_neighbours</span></code></a>(node[, indices])</p></td>
<td><p>Looks for the first neighbours of a given node and returns a list of their UniProt IDs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.fisher_enrichment" title="pypath.legacy.main.PyPath.fisher_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fisher_enrichment</span></code></a>(lst, attr[, ref])</p></td>
<td><p>Computes an enrichment analysis using Fisher's exact test.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.geneset_enrichment" title="pypath.legacy.main.PyPath.geneset_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">geneset_enrichment</span></code></a>(proteins[, all_proteins, ...])</p></td>
<td><p>Does not work at the moment because cfisher module should be replaced with scipy.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.genesymbol" title="pypath.legacy.main.PyPath.genesymbol"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genesymbol</span></code></a>(genesymbol)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol can be found in the default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.genesymbol_labels" title="pypath.legacy.main.PyPath.genesymbol_labels"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genesymbol_labels</span></code></a>([graph, remap_all])</p></td>
<td><p>Creats vertex attribute <code class="docutils literal notranslate"><span class="pre">'label'</span></code> and fills up with the corresponding GeneSymbols of all proteins where the GeneSymbol can be looked up based on the default name of the protein vertex (UniProt ID by default).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.genesymbols" title="pypath.legacy.main.PyPath.genesymbols"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genesymbols</span></code></a>(genesymbols)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_attrs" title="pypath.legacy.main.PyPath.get_attrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_attrs</span></code></a>(line, spec, lnum)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_directed" title="pypath.legacy.main.PyPath.get_directed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_directed</span></code></a>([graph, conv_edges, mutual, ret])</p></td>
<td><p>Converts a copy of <em>graph</em> undirected <em>igraph.Graph</em> object to a directed one.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_dirs_signs" title="pypath.legacy.main.PyPath.get_dirs_signs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_dirs_signs</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_edge" title="pypath.legacy.main.PyPath.get_edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_edge</span></code></a>(source, target[, directed])</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_edges" title="pypath.legacy.main.PyPath.get_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_edges</span></code></a>(sources, targets[, directed])</p></td>
<td><p>Returns a generator with all edges between source and target vertices.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_entity_type</span></code>(entity)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_function" title="pypath.legacy.main.PyPath.get_function"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_function</span></code></a>(fun)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_giant" title="pypath.legacy.main.PyPath.get_giant"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_giant</span></code></a>([replace, graph])</p></td>
<td><p>Returns the giant component of the <em>graph</em>, or replaces the <code class="xref py py-class docutils literal notranslate"><span class="pre">igraph.Graph</span></code> instance with only the giant component if specified.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_go" title="pypath.legacy.main.PyPath.get_go"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_go</span></code></a>([organism])</p></td>
<td><p>Returns the <code class="docutils literal notranslate"><span class="pre">GOAnnotation</span></code> object for the organism requested (or the default one).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_max" title="pypath.legacy.main.PyPath.get_max"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_max</span></code></a>(attrList)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_network" title="pypath.legacy.main.PyPath.get_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_network</span></code></a>(crit[, andor, graph])</p></td>
<td><p>Retrieves a subnetwork according to a set of user-defined attributes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_node" title="pypath.legacy.main.PyPath.get_node"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_node</span></code></a>(identifier)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier is a valid vertex index in the default undirected graph, or a UniProt ID or GeneSymbol which can be found in the default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_node_d" title="pypath.legacy.main.PyPath.get_node_d"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_node_d</span></code></a>(identifier)</p></td>
<td><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_node_pair" title="pypath.legacy.main.PyPath.get_node_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_node_pair</span></code></a>(id_a, id_b[, directed])</p></td>
<td><p>Retrieves the node IDs from a pair of node names.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_nodes" title="pypath.legacy.main.PyPath.get_nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_nodes</span></code></a>(identifiers)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_nodes_d" title="pypath.legacy.main.PyPath.get_nodes_d"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_nodes_d</span></code></a>(identifiers)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_pathways" title="pypath.legacy.main.PyPath.get_pathways"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_pathways</span></code></a>(source)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_proteomicsdb" title="pypath.legacy.main.PyPath.get_proteomicsdb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_proteomicsdb</span></code></a>(user, passwd[, tissues, pickle])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_sub" title="pypath.legacy.main.PyPath.get_sub"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sub</span></code></a>(crit[, andor, graph])</p></td>
<td><p>Selects the nodes from <em>graph</em> (and edges to be removed) according to a set of user-defined attributes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.get_taxon" title="pypath.legacy.main.PyPath.get_taxon"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_taxon</span></code></a>(tax_dict, fields)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.go_annotate_graph" title="pypath.legacy.main.PyPath.go_annotate_graph"><code class="xref py py-obj docutils literal notranslate"><span class="pre">go_annotate_graph</span></code></a>([aspects])</p></td>
<td><p>Annotates protein nodes with GO terms.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.go_enrichment" title="pypath.legacy.main.PyPath.go_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">go_enrichment</span></code></a>([proteins, aspect, alpha, ...])</p></td>
<td><p>Does not work at the moment because cfisher module should be replaced with scipy.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs" title="pypath.legacy.main.PyPath.gs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs</span></code></a>(genesymbol)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol can be found in the default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_affected_by" title="pypath.legacy.main.PyPath.gs_affected_by"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_affected_by</span></code></a>(genesymbol)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_affects" title="pypath.legacy.main.PyPath.gs_affects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_affects</span></code></a>(genesymbol)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_edge" title="pypath.legacy.main.PyPath.gs_edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_edge</span></code></a>(source, target[, directed])</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_in_directed" title="pypath.legacy.main.PyPath.gs_in_directed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_in_directed</span></code></a>(genesymbol)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_in_undirected" title="pypath.legacy.main.PyPath.gs_in_undirected"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_in_undirected</span></code></a>(genesymbol)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_inhibited_by" title="pypath.legacy.main.PyPath.gs_inhibited_by"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_inhibited_by</span></code></a>(genesymbol)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_inhibits" title="pypath.legacy.main.PyPath.gs_inhibits"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_inhibits</span></code></a>(genesymbol)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_neighborhood" title="pypath.legacy.main.PyPath.gs_neighborhood"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_neighborhood</span></code></a>(genesymbols[, order, mode])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_neighbors" title="pypath.legacy.main.PyPath.gs_neighbors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_neighbors</span></code></a>(genesymbol[, mode])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_stimulated_by" title="pypath.legacy.main.PyPath.gs_stimulated_by"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_stimulated_by</span></code></a>(genesymbol)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gs_stimulates" title="pypath.legacy.main.PyPath.gs_stimulates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gs_stimulates</span></code></a>(genesymbol)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.gss" title="pypath.legacy.main.PyPath.gss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gss</span></code></a>(genesymbols)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.guide2pharma" title="pypath.legacy.main.PyPath.guide2pharma"><code class="xref py py-obj docutils literal notranslate"><span class="pre">guide2pharma</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.having_attr" title="pypath.legacy.main.PyPath.having_attr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">having_attr</span></code></a>(attr[, graph, index, edges])</p></td>
<td><p>Checks if edges or nodes of the network have a specific attribute and returns an iterator of the indices (or the edge/node instances) of edges/nodes having such attribute.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.having_eattr" title="pypath.legacy.main.PyPath.having_eattr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">having_eattr</span></code></a>(attr[, graph, index])</p></td>
<td><p>Checks if edges of the network have a specific attribute and returns an iterator of the indices (or the edge instances) of edges having such attribute.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.having_ptm" title="pypath.legacy.main.PyPath.having_ptm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">having_ptm</span></code></a>([index, graph])</p></td>
<td><p>Checks if edges of the network have the <code class="docutils literal notranslate"><span class="pre">'ptm'</span></code> attribute and returns an iterator of the indices (or the edge instances) of edges having such attribute.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.having_vattr" title="pypath.legacy.main.PyPath.having_vattr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">having_vattr</span></code></a>(attr[, graph, index])</p></td>
<td><p>Checks if nodes of the network have a specific attribute and returns an iterator of the indices (or the node instances) of nodes having such attribute.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.homology_translation" title="pypath.legacy.main.PyPath.homology_translation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">homology_translation</span></code></a>(target[, source, ...])</p></td>
<td><p>Translates the current object to another organism by orthology.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.htp_stats" title="pypath.legacy.main.PyPath.htp_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">htp_stats</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.in_complex" title="pypath.legacy.main.PyPath.in_complex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">in_complex</span></code></a>([csources])</p></td>
<td><p>Deprecated, will be removed.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.in_directed" title="pypath.legacy.main.PyPath.in_directed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">in_directed</span></code></a>(vertex)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.in_undirected" title="pypath.legacy.main.PyPath.in_undirected"><code class="xref py py-obj docutils literal notranslate"><span class="pre">in_undirected</span></code></a>(vertex)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.info" title="pypath.legacy.main.PyPath.info"><code class="xref py py-obj docutils literal notranslate"><span class="pre">info</span></code></a>(name)</p></td>
<td><p>Given the name of a resource, prints out the information about that source/database.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.init_complex_attr" title="pypath.legacy.main.PyPath.init_complex_attr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">init_complex_attr</span></code></a>(graph, name)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.init_edge_attr" title="pypath.legacy.main.PyPath.init_edge_attr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">init_edge_attr</span></code></a>(attr)</p></td>
<td><p>Fills all edges attribute <em>attr</em> with its default type (if such attribute value is <code class="docutils literal notranslate"><span class="pre">None</span></code>), creates [list] if in <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.edgeAttrs</span></code> such attribute is registered as [list].</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.init_gsea" title="pypath.legacy.main.PyPath.init_gsea"><code class="xref py py-obj docutils literal notranslate"><span class="pre">init_gsea</span></code></a>(user)</p></td>
<td><p>Initializes a <code class="docutils literal notranslate"><span class="pre">pypath.gsea.GSEA</span></code> object and shows the list of the collections in MSigDB.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.init_network" title="pypath.legacy.main.PyPath.init_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">init_network</span></code></a>([lst, exclude, cache_files, ...])</p></td>
<td><p>Loads the network data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.init_vertex_attr" title="pypath.legacy.main.PyPath.init_vertex_attr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">init_vertex_attr</span></code></a>(attr)</p></td>
<td><p>Fills all vertices attribute <em>attr</em> with its default type (if such attribute value is <code class="docutils literal notranslate"><span class="pre">None</span></code>), creates [list] if in <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.vertexAttrs</span></code> such attribute is registered as [list].</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_all" title="pypath.legacy.main.PyPath.interactions_all"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_all</span></code></a>([resources])</p></td>
<td><p>Returns a <em>set</em> of tuples of node name pairs representing interactions, both directed and undirected.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_all_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_all_by_resource</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_all_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_directed" title="pypath.legacy.main.PyPath.interactions_directed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_directed</span></code></a>([resources])</p></td>
<td><p>Returns a <em>set</em> of tuples of node name pairs with being aware of the directions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_directed_by_category</span></code>([effect])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_directed_by_resource" title="pypath.legacy.main.PyPath.interactions_directed_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_directed_by_resource</span></code></a>([...])</p></td>
<td><p>Returns a <em>dict</em> of <a href="#id9"><span class="problematic" id="id10">*</span></a>set*s of tuples of node name pairs with being aware of the directions. Undirected interactions will be discarded. Pairs of node names represent the directions: first is the source, second is the target.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_directed_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_inhibitory" title="pypath.legacy.main.PyPath.interactions_inhibitory"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_inhibitory</span></code></a>([resources])</p></td>
<td><p>Returns a <em>set</em> of tuples of node name pairs only for inhibitory interactions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_inhibitory_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_inhibitory_by_resource" title="pypath.legacy.main.PyPath.interactions_inhibitory_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_inhibitory_by_resource</span></code></a>([resources])</p></td>
<td><p>Returns a <em>dict</em> of <a href="#id11"><span class="problematic" id="id12">*</span></a>set*s of tuples of node name pairs with being aware of the directions. Undirected interactions will be discarded. Pairs of node names represent the directions: first is the source, second is the target.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_inhibitory_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_mutual" title="pypath.legacy.main.PyPath.interactions_mutual"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_mutual</span></code></a>([resources])</p></td>
<td><p>Returns a <em>set</em> of tuples of node name pairs representing mutual interactions (i.e.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_mutual_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_mutual_by_resource" title="pypath.legacy.main.PyPath.interactions_mutual_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_mutual_by_resource</span></code></a>([resources])</p></td>
<td><p>Returns a <em>dict</em> of <a href="#id13"><span class="problematic" id="id14">*</span></a>set*s of tuples of node name pairs representing mutual interactions (i.e. A--&gt;B and B--&gt;A). Pairs of node names will be sorted alphabetically.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_mutual_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_signed" title="pypath.legacy.main.PyPath.interactions_signed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_signed</span></code></a>([resources])</p></td>
<td><p>Returns a <em>set</em> of tuples of node name pairs only for signed interactions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_signed_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_signed_by_resource" title="pypath.legacy.main.PyPath.interactions_signed_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_signed_by_resource</span></code></a>([resources])</p></td>
<td><p>Returns a <em>dict</em> of <a href="#id15"><span class="problematic" id="id16">*</span></a>set*s of tuples of node name pairs with being aware of the directions. Undirected interactions will be discarded. Pairs of node names represent the directions: first is the source, second is the target.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_signed_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_stimulatory" title="pypath.legacy.main.PyPath.interactions_stimulatory"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_stimulatory</span></code></a>([resources])</p></td>
<td><p>Returns a <em>set</em> of tuples of node name pairs only for stimulatory interactions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_stimulatory_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_stimulatory_by_resource" title="pypath.legacy.main.PyPath.interactions_stimulatory_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_stimulatory_by_resource</span></code></a>([resources])</p></td>
<td><p>Returns a <em>dict</em> of <a href="#id17"><span class="problematic" id="id18">*</span></a>set*s of tuples of node name pairs with being aware of the directions. Undirected interactions will be discarded. Pairs of node names represent the directions: first is the source, second is the target.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_stimulatory_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_undirected" title="pypath.legacy.main.PyPath.interactions_undirected"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_undirected</span></code></a>([resources])</p></td>
<td><p>Returns a <em>set</em> of tuples of node name pairs without being aware of the directions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_undirected_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.interactions_undirected_by_resource" title="pypath.legacy.main.PyPath.interactions_undirected_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_undirected_by_resource</span></code></a>([resources])</p></td>
<td><p>Returns a <em>dict</em> of <a href="#id19"><span class="problematic" id="id20">*</span></a>set*s of tuples of node name pairs without being aware of the directions. Pairs of node names will be sorted alphabetically.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interactions_undirected_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.intergroup_shortest_paths" title="pypath.legacy.main.PyPath.intergroup_shortest_paths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intergroup_shortest_paths</span></code></a>(groupA, groupB[, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.intogen_cancer_drivers_list" title="pypath.legacy.main.PyPath.intogen_cancer_drivers_list"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intogen_cancer_drivers_list</span></code></a>(intogen_file)</p></td>
<td><p>Loads the list of cancer driver proteins from IntOGen data.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_complex</span></code>(entity)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_mirna</span></code>(entity)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_protein</span></code>(entity)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.iter_edges" title="pypath.legacy.main.PyPath.iter_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_edges</span></code></a>([resources])</p></td>
<td><p>Iterates the edges in the graph optionally limited to certain resources.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_entities</span></code>([resources, entity_type])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.iter_interactions" title="pypath.legacy.main.PyPath.iter_interactions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_interactions</span></code></a>([signs, all_undirected, ...])</p></td>
<td><p>Iterates over edges and yields interaction records.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_vertices</span></code>([resources, categories, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.jaccard_edges" title="pypath.legacy.main.PyPath.jaccard_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">jaccard_edges</span></code></a>()</p></td>
<td><p>Computes the Jaccard similarity index between the sets of first neighbours of all node pairs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.jaccard_meta" title="pypath.legacy.main.PyPath.jaccard_meta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">jaccard_meta</span></code></a>(jedges, critical)</p></td>
<td><p>Creates a (undirected) graph from a list of edges filtering by their Jaccard index.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.kegg_directions" title="pypath.legacy.main.PyPath.kegg_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kegg_directions</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.kegg_pathways" title="pypath.legacy.main.PyPath.kegg_pathways"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kegg_pathways</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.kinase_stats" title="pypath.legacy.main.PyPath.kinase_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kinase_stats</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.label" title="pypath.legacy.main.PyPath.label"><code class="xref py py-obj docutils literal notranslate"><span class="pre">label</span></code></a>(label, idx[, what])</p></td>
<td><p>Creates a boolean attribute <code class="docutils literal notranslate"><span class="pre">label</span></code> True for the vertex or edge IDs in the set <code class="docutils literal notranslate"><span class="pre">idx</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.label_by_go" title="pypath.legacy.main.PyPath.label_by_go"><code class="xref py py-obj docutils literal notranslate"><span class="pre">label_by_go</span></code></a>(label, go_terms[, method])</p></td>
<td><p>Assigns a boolean vertex attribute to nodes which tells whether the node is annotated by all or any of the GO terms.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.label_edges" title="pypath.legacy.main.PyPath.label_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">label_edges</span></code></a>(label, edges)</p></td>
<td><p>Creates a boolean edge attribute <code class="docutils literal notranslate"><span class="pre">label</span></code> True for the edge IDs in the set <code class="docutils literal notranslate"><span class="pre">edges</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.label_vertices" title="pypath.legacy.main.PyPath.label_vertices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">label_vertices</span></code></a>(label, vertices)</p></td>
<td><p>Creates a boolean vertex attribute <code class="docutils literal notranslate"><span class="pre">label</span></code> True for the vertex IDs in the set <code class="docutils literal notranslate"><span class="pre">vertices</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.laudanna_directions" title="pypath.legacy.main.PyPath.laudanna_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">laudanna_directions</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.laudanna_effects" title="pypath.legacy.main.PyPath.laudanna_effects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">laudanna_effects</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.license" title="pypath.legacy.main.PyPath.license"><code class="xref py py-obj docutils literal notranslate"><span class="pre">license</span></code></a>(self)</p></td>
<td><p>Prints information about data licences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.list_resources" title="pypath.legacy.main.PyPath.list_resources"><code class="xref py py-obj docutils literal notranslate"><span class="pre">list_resources</span></code></a>()</p></td>
<td><p>Prints the list of resources through the standard output.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_3dcomplexes" title="pypath.legacy.main.PyPath.load_3dcomplexes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_3dcomplexes</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_3did_ddi" title="pypath.legacy.main.PyPath.load_3did_ddi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_3did_ddi</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_3did_ddi2" title="pypath.legacy.main.PyPath.load_3did_ddi2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_3did_ddi2</span></code></a>([ddi, interfaces])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_3did_dmi" title="pypath.legacy.main.PyPath.load_3did_dmi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_3did_dmi</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_3did_interfaces" title="pypath.legacy.main.PyPath.load_3did_interfaces"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_3did_interfaces</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_all_pathways" title="pypath.legacy.main.PyPath.load_all_pathways"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_all_pathways</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_compleat" title="pypath.legacy.main.PyPath.load_compleat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_compleat</span></code></a>([graph])</p></td>
<td><p>Loads complexes from Compleat.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_complexportal" title="pypath.legacy.main.PyPath.load_complexportal"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_complexportal</span></code></a>([graph])</p></td>
<td><p>Loads complexes from ComplexPortal.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_comppi" title="pypath.legacy.main.PyPath.load_comppi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_comppi</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_corum" title="pypath.legacy.main.PyPath.load_corum"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_corum</span></code></a>([graph])</p></td>
<td><p>Loads complexes from CORUM database.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_dbptm" title="pypath.legacy.main.PyPath.load_dbptm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_dbptm</span></code></a>([non_matching, trace])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_ddi" title="pypath.legacy.main.PyPath.load_ddi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_ddi</span></code></a>(ddi)</p></td>
<td><p>ddi is either a list of intera.DomainDomain objects, or a function resulting this list</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_ddis" title="pypath.legacy.main.PyPath.load_ddis"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_ddis</span></code></a>([methods])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_depod_dmi" title="pypath.legacy.main.PyPath.load_depod_dmi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_depod_dmi</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_disgenet" title="pypath.legacy.main.PyPath.load_disgenet"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_disgenet</span></code></a>([dataset, score, umls, full_data])</p></td>
<td><p>Assigns DisGeNet disease-gene associations to the proteins in the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_dmi" title="pypath.legacy.main.PyPath.load_dmi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_dmi</span></code></a>(dmi)</p></td>
<td><p>dmi is either a list of intera.DomainMotif objects, or a function resulting this list</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_dmis" title="pypath.legacy.main.PyPath.load_dmis"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_dmis</span></code></a>([methods])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_domino_dmi" title="pypath.legacy.main.PyPath.load_domino_dmi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_domino_dmi</span></code></a>([organism])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_dorothea" title="pypath.legacy.main.PyPath.load_dorothea"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_dorothea</span></code></a>([levels, only_curated])</p></td>
<td><p>Adds TF-target interactions from TF regulons to the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_elm" title="pypath.legacy.main.PyPath.load_elm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_elm</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_exocarta_attrs" title="pypath.legacy.main.PyPath.load_exocarta_attrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_exocarta_attrs</span></code></a>([load_samples, load_refs])</p></td>
<td><p>Creates vertex attributes from ExoCarta data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_expression" title="pypath.legacy.main.PyPath.load_expression"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_expression</span></code></a>([array])</p></td>
<td><p>Expression data can be loaded into vertex attributes, or into a pandas DataFrame – the latter offers faster ways to process and use these huge matrices.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_from_pickle" title="pypath.legacy.main.PyPath.load_from_pickle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_from_pickle</span></code></a>(pickle_file)</p></td>
<td><p>Shortcut for loading a network from a pickle dump.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_go" title="pypath.legacy.main.PyPath.load_go"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_go</span></code></a>([organism])</p></td>
<td><p>Creates a <code class="docutils literal notranslate"><span class="pre">pypath.go.GOAnnotation</span></code> object for one organism in the dict under <code class="docutils literal notranslate"><span class="pre">go</span></code> attribute.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_havugimana" title="pypath.legacy.main.PyPath.load_havugimana"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_havugimana</span></code></a>([graph])</p></td>
<td><p>Loads complexes from Havugimana 2012.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_hpa" title="pypath.legacy.main.PyPath.load_hpa"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_hpa</span></code></a>([normal, pathology, cancer, ...])</p></td>
<td><p>Loads Human Protein Atlas data into vertex attributes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_hprd_ptms" title="pypath.legacy.main.PyPath.load_hprd_ptms"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_hprd_ptms</span></code></a>([non_matching, trace])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_ielm" title="pypath.legacy.main.PyPath.load_ielm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_ielm</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_interfaces" title="pypath.legacy.main.PyPath.load_interfaces"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_interfaces</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_li2012_ptms" title="pypath.legacy.main.PyPath.load_li2012_ptms"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_li2012_ptms</span></code></a>([non_matching, trace])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_ligand_receptor_network" title="pypath.legacy.main.PyPath.load_ligand_receptor_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_ligand_receptor_network</span></code></a>([...])</p></td>
<td><p>Initializes a ligand-receptor network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_lmpid" title="pypath.legacy.main.PyPath.load_lmpid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_lmpid</span></code></a>(method)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_matrisome_attrs" title="pypath.legacy.main.PyPath.load_matrisome_attrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_matrisome_attrs</span></code></a>([organism])</p></td>
<td><p>Loads vertex attributes from MatrisomeDB 2.0.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_membranome_attrs" title="pypath.legacy.main.PyPath.load_membranome_attrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_membranome_attrs</span></code></a>()</p></td>
<td><p>Loads attributes from Membranome, a database of single-helix transmembrane proteins.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_mimp_dmi" title="pypath.legacy.main.PyPath.load_mimp_dmi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_mimp_dmi</span></code></a>([non_matching, trace])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_mutations" title="pypath.legacy.main.PyPath.load_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_mutations</span></code></a>([attributes, gdsc_datadir, ...])</p></td>
<td><p>Mutations are listed in vertex attributes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_negatives" title="pypath.legacy.main.PyPath.load_negatives"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_negatives</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_old_omnipath" title="pypath.legacy.main.PyPath.load_old_omnipath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_old_omnipath</span></code></a>([kinase_substrate_extra, ...])</p></td>
<td><p>Loads the OmniPath network as it was before August 2016.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_omnipath" title="pypath.legacy.main.PyPath.load_omnipath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_omnipath</span></code></a>([omnipath, ...])</p></td>
<td><p>Loads the OmniPath network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_pathways" title="pypath.legacy.main.PyPath.load_pathways"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_pathways</span></code></a>(source[, graph])</p></td>
<td><p>Generic method to load pathway annotations from a resource.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_pdb" title="pypath.legacy.main.PyPath.load_pdb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_pdb</span></code></a>([graph])</p></td>
<td><p>Loads the 3D structure information from PDB into the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_pepcyber" title="pypath.legacy.main.PyPath.load_pepcyber"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_pepcyber</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_pfam" title="pypath.legacy.main.PyPath.load_pfam"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_pfam</span></code></a>([graph])</p></td>
<td><p>Loads the protein family information from UniProt into the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_pfam2" title="pypath.legacy.main.PyPath.load_pfam2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_pfam2</span></code></a>()</p></td>
<td><p>Loads the protein family information from Pfam into the network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_pfam3" title="pypath.legacy.main.PyPath.load_pfam3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_pfam3</span></code></a>()</p></td>
<td><p>Loads the protein domain information from Pfam into the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_phospho_dmi" title="pypath.legacy.main.PyPath.load_phospho_dmi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_phospho_dmi</span></code></a>(source[, trace, return_raw])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_phosphoelm" title="pypath.legacy.main.PyPath.load_phosphoelm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_phosphoelm</span></code></a>([trace])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_pisa" title="pypath.legacy.main.PyPath.load_pisa"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_pisa</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_pnetworks_dmi" title="pypath.legacy.main.PyPath.load_pnetworks_dmi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_pnetworks_dmi</span></code></a>([trace])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_psite_phos" title="pypath.legacy.main.PyPath.load_psite_phos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_psite_phos</span></code></a>([trace])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_psite_reg" title="pypath.legacy.main.PyPath.load_psite_reg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_psite_reg</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_ptms" title="pypath.legacy.main.PyPath.load_ptms"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_ptms</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_ptms2" title="pypath.legacy.main.PyPath.load_ptms2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_ptms2</span></code></a>([input_methods, ...])</p></td>
<td><p>This is a new method which will replace <cite>load_ptms</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_resource" title="pypath.legacy.main.PyPath.load_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_resource</span></code></a>(settings[, clean, ...])</p></td>
<td><p>Loads the data from a single resource and attaches it to the network</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_resources" title="pypath.legacy.main.PyPath.load_resources"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_resources</span></code></a>([lst, exclude, cache_files, ...])</p></td>
<td><p>Loads multiple resources, and cleans up after.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_signor_ptms" title="pypath.legacy.main.PyPath.load_signor_ptms"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_signor_ptms</span></code></a>([non_matching, trace])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_surfaceome_attrs" title="pypath.legacy.main.PyPath.load_surfaceome_attrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_surfaceome_attrs</span></code></a>()</p></td>
<td><p>Loads vertex attributes from the In Silico Human Surfaceome.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_tfregulons" title="pypath.legacy.main.PyPath.load_tfregulons"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_tfregulons</span></code></a>([levels, only_curated])</p></td>
<td><p>Adds TF-target interactions from TF regulons to the network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.load_vesiclepedia_attrs" title="pypath.legacy.main.PyPath.load_vesiclepedia_attrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_vesiclepedia_attrs</span></code></a>([load_samples, ...])</p></td>
<td><p>Creates vertex attributes from Vesiclepedia data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.lookup_cache" title="pypath.legacy.main.PyPath.lookup_cache"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lookup_cache</span></code></a>(name, cache_files, int_cache, ...)</p></td>
<td><p>Checks up the cache folder for the files of a given resource.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.loop_edges" title="pypath.legacy.main.PyPath.loop_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">loop_edges</span></code></a>([index, graph])</p></td>
<td><p>Returns an iterator of the indices (or the edge instances) of the edges which represent a loop (whose source and target node are the same).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.mean_reference_per_interaction" title="pypath.legacy.main.PyPath.mean_reference_per_interaction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mean_reference_per_interaction</span></code></a>([resources])</p></td>
<td><p>Computes the mean number of references per interaction of the network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.mean_reference_per_interaction_by_resource" title="pypath.legacy.main.PyPath.mean_reference_per_interaction_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mean_reference_per_interaction_by_resource</span></code></a>([...])</p></td>
<td><p>Computes the mean number of references per interaction of the network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.merge_lists" title="pypath.legacy.main.PyPath.merge_lists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merge_lists</span></code></a>(id_a, id_b[, name, and_or, ...])</p></td>
<td><p>Merges two lists from <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypat.main.PyPath.lists</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.merge_nodes" title="pypath.legacy.main.PyPath.merge_nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merge_nodes</span></code></a>(nodes[, primary, graph])</p></td>
<td><p>Merges all attributes and edges of selected nodes and assigns them to the primary node (by default the one with lowest index).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.mimp_directions" title="pypath.legacy.main.PyPath.mimp_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mimp_directions</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">mirna_entities</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">mirna_entities_by_category</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">mirna_entities_by_resource</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.mutated_edges" title="pypath.legacy.main.PyPath.mutated_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mutated_edges</span></code></a>(sample)</p></td>
<td><p>Compares the mutated residues and the modified residues in PTMs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.name_edgelist" title="pypath.legacy.main.PyPath.name_edgelist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">name_edgelist</span></code></a>([graph])</p></td>
<td><p>Returns an edge list, i.e. a list with tuples of vertex names.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">name_to_label</span></code>(name)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.names2vids" title="pypath.legacy.main.PyPath.names2vids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">names2vids</span></code></a>(names)</p></td>
<td><p>From a list of node names, returns their corresponding indices.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.negative_report" title="pypath.legacy.main.PyPath.negative_report"><code class="xref py py-obj docutils literal notranslate"><span class="pre">negative_report</span></code></a>([lst, outFile])</p></td>
<td><p>Generates a report file with the negative interactions (assumed to be already loaded).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.neighborhood" title="pypath.legacy.main.PyPath.neighborhood"><code class="xref py py-obj docutils literal notranslate"><span class="pre">neighborhood</span></code></a>(identifiers[, order, mode])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.neighbors" title="pypath.legacy.main.PyPath.neighbors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">neighbors</span></code></a>(identifier[, mode])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.neighbourhood_network" title="pypath.legacy.main.PyPath.neighbourhood_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">neighbourhood_network</span></code></a>(center[, second])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.network_by_go" title="pypath.legacy.main.PyPath.network_by_go"><code class="xref py py-obj docutils literal notranslate"><span class="pre">network_by_go</span></code></a>(node_categories[, ...])</p></td>
<td><p>Creates or filters a network based on Gene Ontology annotations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.network_filter" title="pypath.legacy.main.PyPath.network_filter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">network_filter</span></code></a>([p])</p></td>
<td><p>This function aims to cut the number of edges in the network, without losing nodes, to make the network less connected, less hairball-like, more usable for analysis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.network_stats" title="pypath.legacy.main.PyPath.network_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">network_stats</span></code></a>([outfile])</p></td>
<td><p>Calculates basic statistics for the whole network and each of sources (node and edge counts, average node degree, graph diameter, transitivity, adhesion and cohesion).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.new_edges" title="pypath.legacy.main.PyPath.new_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">new_edges</span></code></a>(edges)</p></td>
<td><p>Adds new edges from any iterable of edges to the undirected graph.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.new_nodes" title="pypath.legacy.main.PyPath.new_nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">new_nodes</span></code></a>(nodes)</p></td>
<td><p>Adds new nodes from any iterable of nodes to the undirected graph.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.node_exists" title="pypath.legacy.main.PyPath.node_exists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">node_exists</span></code></a>(name)</p></td>
<td><p>Checks if a node exists in the (undirected) network.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">nodes_by_resource</span></code>(resource)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">nodes_by_resources</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.numof_directed_edges" title="pypath.legacy.main.PyPath.numof_directed_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_directed_edges</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.numof_edges" title="pypath.legacy.main.PyPath.numof_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_edges</span></code></a>([resources])</p></td>
<td><p>Number of edges optionally limited to certain resources.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.numof_reference_interaction_pairs" title="pypath.legacy.main.PyPath.numof_reference_interaction_pairs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_reference_interaction_pairs</span></code></a>()</p></td>
<td><p>Returns the total of unique references per interaction.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_references</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_references_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.numof_references_by_resource" title="pypath.legacy.main.PyPath.numof_references_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_references_by_resource</span></code></a>([resources])</p></td>
<td><p>Counts the references for each resource, optionally limited to certain resources.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.numof_undirected_edges" title="pypath.legacy.main.PyPath.numof_undirected_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">numof_undirected_edges</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.orthology_translation" title="pypath.legacy.main.PyPath.orthology_translation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">orthology_translation</span></code></a>(target[, source, ...])</p></td>
<td><p>Translates the current object to another organism by orthology.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.p" title="pypath.legacy.main.PyPath.p"><code class="xref py py-obj docutils literal notranslate"><span class="pre">p</span></code></a>(identifier)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier is a valid vertex index in the default undirected graph, or a UniProt ID or GeneSymbol which can be found in the default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.pathway_attributes" title="pypath.legacy.main.PyPath.pathway_attributes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pathway_attributes</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.pathway_members" title="pypath.legacy.main.PyPath.pathway_members"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pathway_members</span></code></a>(pathway, source)</p></td>
<td><p>Returns an iterator with the members of a single pathway.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.pathway_names" title="pypath.legacy.main.PyPath.pathway_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pathway_names</span></code></a>(source[, graph])</p></td>
<td><p>Returns the names of all pathways having at least one member in the current graph.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.pathway_similarity" title="pypath.legacy.main.PyPath.pathway_similarity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pathway_similarity</span></code></a>([outfile])</p></td>
<td><p>Computes the Sorensen's similarity index across nodes and edges for all the available pathway sources (already loaded in the network) and saves them into table files.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.pathways_table" title="pypath.legacy.main.PyPath.pathways_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pathways_table</span></code></a>([filename, pw_sources, graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.pfam_regions" title="pypath.legacy.main.PyPath.pfam_regions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pfam_regions</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.phosphonetworks_directions" title="pypath.legacy.main.PyPath.phosphonetworks_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phosphonetworks_directions</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.phosphopoint_directions" title="pypath.legacy.main.PyPath.phosphopoint_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phosphopoint_directions</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.phosphorylation_directions" title="pypath.legacy.main.PyPath.phosphorylation_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phosphorylation_directions</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.phosphorylation_signs" title="pypath.legacy.main.PyPath.phosphorylation_signs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phosphorylation_signs</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.phosphosite_directions" title="pypath.legacy.main.PyPath.phosphosite_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phosphosite_directions</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.prdb_tissue_expr" title="pypath.legacy.main.PyPath.prdb_tissue_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prdb_tissue_expr</span></code></a>(tissue[, prdb, graph, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.process_directions" title="pypath.legacy.main.PyPath.process_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">process_directions</span></code></a>(dirs, name[, directed, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.process_dmi" title="pypath.legacy.main.PyPath.process_dmi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">process_dmi</span></code></a>(source, **kwargs)</p></td>
<td><p>This is an universal function for loading domain-motif objects like load_phospho_dmi() for PTMs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.protein" title="pypath.legacy.main.PyPath.protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">protein</span></code></a>(identifier)</p></td>
<td><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.protein_edge" title="pypath.legacy.main.PyPath.protein_edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">protein_edge</span></code></a>(source, target[, directed])</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">protein_entities</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">protein_entities_by_category</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">protein_entities_by_resource</span></code>([resources])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.proteins" title="pypath.legacy.main.PyPath.proteins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">proteins</span></code></a>(identifiers)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.ps" title="pypath.legacy.main.PyPath.ps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ps</span></code></a>(identifiers)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.random_walk_with_return" title="pypath.legacy.main.PyPath.random_walk_with_return"><code class="xref py py-obj docutils literal notranslate"><span class="pre">random_walk_with_return</span></code></a>(q[, graph, c, niter])</p></td>
<td><p>Random walk with return (RWR) starting from one or more query nodes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.random_walk_with_return2" title="pypath.legacy.main.PyPath.random_walk_with_return2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">random_walk_with_return2</span></code></a>(q[, c, niter])</p></td>
<td><p>Literally does random walks.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.read_from_cache" title="pypath.legacy.main.PyPath.read_from_cache"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_from_cache</span></code></a>(cache_file)</p></td>
<td><p>Reads a pickle file from the cache and returns it.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.read_list_file" title="pypath.legacy.main.PyPath.read_list_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_list_file</span></code></a>(settings, **kwargs)</p></td>
<td><p>Reads a list from a file and adds it to <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.lists</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.reference_edge_ratio" title="pypath.legacy.main.PyPath.reference_edge_ratio"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reference_edge_ratio</span></code></a>()</p></td>
<td><p>Computes the average number of references per edge (as in the undirected graph).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.reference_hist" title="pypath.legacy.main.PyPath.reference_hist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reference_hist</span></code></a>([filename])</p></td>
<td><p>Generates a file containing a table with information about the network's edges.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.references" title="pypath.legacy.main.PyPath.references"><code class="xref py py-obj docutils literal notranslate"><span class="pre">references</span></code></a>([resources])</p></td>
<td><p>Returns a set of references for all edges.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">references_by_category</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.references_by_resource" title="pypath.legacy.main.PyPath.references_by_resource"><code class="xref py py-obj docutils literal notranslate"><span class="pre">references_by_resource</span></code></a>([resources])</p></td>
<td><p>Creates a dict with resources as keys and sets of references as values.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">references_stats</span></code>(**kwargs)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.reload" title="pypath.legacy.main.PyPath.reload"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reload</span></code></a>()</p></td>
<td><p>Reloads the object from the module level.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.remove_htp" title="pypath.legacy.main.PyPath.remove_htp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove_htp</span></code></a>([threshold, keep_directed])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.remove_undirected" title="pypath.legacy.main.PyPath.remove_undirected"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove_undirected</span></code></a>([min_refs])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">resources_by_category</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.run_batch" title="pypath.legacy.main.PyPath.run_batch"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_batch</span></code></a>(methods[, toCall])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.save_network" title="pypath.legacy.main.PyPath.save_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_network</span></code></a>([pickle_file, pfile])</p></td>
<td><p>Saves the network object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.save_session" title="pypath.legacy.main.PyPath.save_session"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_session</span></code></a>()</p></td>
<td><p>Save the current session state into pickle dump.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.save_to_pickle" title="pypath.legacy.main.PyPath.save_to_pickle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_to_pickle</span></code></a>([pickle_file, pfile])</p></td>
<td><p>Saves the network object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.search_attr_and" title="pypath.legacy.main.PyPath.search_attr_and"><code class="xref py py-obj docutils literal notranslate"><span class="pre">search_attr_and</span></code></a>(obj, lst)</p></td>
<td><p>Searches a given collection of attributes in a given object.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.search_attr_or" title="pypath.legacy.main.PyPath.search_attr_or"><code class="xref py py-obj docutils literal notranslate"><span class="pre">search_attr_or</span></code></a>(obj, lst)</p></td>
<td><p>Searches a given collection of attributes in a given object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.second_neighbours" title="pypath.legacy.main.PyPath.second_neighbours"><code class="xref py py-obj docutils literal notranslate"><span class="pre">second_neighbours</span></code></a>(node[, indices, with_first])</p></td>
<td><p>Looks for the (first and) second neighbours of a given node and returns a list of their UniProt IDs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.select_by_go" title="pypath.legacy.main.PyPath.select_by_go"><code class="xref py py-obj docutils literal notranslate"><span class="pre">select_by_go</span></code></a>(go_terms)</p></td>
<td><p>Retrieves the vertex IDs of all vertices annotated with any Gene Ontology terms or their descendants, or evaluates string expression (see <code class="docutils literal notranslate"><span class="pre">select_by_go_expr</span></code>).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.select_by_go_all" title="pypath.legacy.main.PyPath.select_by_go_all"><code class="xref py py-obj docutils literal notranslate"><span class="pre">select_by_go_all</span></code></a>(go_terms)</p></td>
<td><p>Selects the nodes annotated by all GO terms in <code class="docutils literal notranslate"><span class="pre">go_terms</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.select_by_go_expr" title="pypath.legacy.main.PyPath.select_by_go_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">select_by_go_expr</span></code></a>(go_expr)</p></td>
<td><p>Selects vertices based on an expression of Gene Ontology terms.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.separate" title="pypath.legacy.main.PyPath.separate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">separate</span></code></a>()</p></td>
<td><p>Separates the undirected network according to the different sources.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.separate_by_category" title="pypath.legacy.main.PyPath.separate_by_category"><code class="xref py py-obj docutils literal notranslate"><span class="pre">separate_by_category</span></code></a>()</p></td>
<td><p>Separates the undirected network according to resource categories.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.sequences" title="pypath.legacy.main.PyPath.sequences"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequences</span></code></a>([isoforms, update])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_boolean_vattr" title="pypath.legacy.main.PyPath.set_boolean_vattr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_boolean_vattr</span></code></a>(attr, vids[, negate])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_categories" title="pypath.legacy.main.PyPath.set_categories"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_categories</span></code></a>()</p></td>
<td><p>Sets the category attribute on the network nodes and edges (<code class="docutils literal notranslate"><span class="pre">'cat'</span></code>) as well the edge attribute coercing the references by category (<code class="docutils literal notranslate"><span class="pre">'refs_by_cat'</span></code>).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_chembl_mysql" title="pypath.legacy.main.PyPath.set_chembl_mysql"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_chembl_mysql</span></code></a>(title[, config_file])</p></td>
<td><p>Sets the ChEMBL MySQL configuration according to the <em>title</em> section in <em>config_file</em> ini file configuration.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_disease_genes" title="pypath.legacy.main.PyPath.set_disease_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_disease_genes</span></code></a>([dataset])</p></td>
<td><p>Creates a vertex attribute named <cite>dis</cite> with boolean values <em>True</em> if the protein encoded by a disease related gene according to DisGeNet.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_druggability" title="pypath.legacy.main.PyPath.set_druggability"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_druggability</span></code></a>()</p></td>
<td><p>Creates a vertex attribute <cite>dgb</cite> with value <em>True</em> if the protein is druggable, otherwise <em>False</em>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_drugtargets" title="pypath.legacy.main.PyPath.set_drugtargets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_drugtargets</span></code></a>([pchembl])</p></td>
<td><p>Creates a vertex attribute <cite>dtg</cite> with value <em>True</em> if the protein has at least one compound binding with affinity higher than <cite>pchembl</cite>, otherwise <em>False</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_kinases" title="pypath.legacy.main.PyPath.set_kinases"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_kinases</span></code></a>()</p></td>
<td><p>Creates a vertex attribute <cite>kin</cite> with value <em>True</em> if the protein is a kinase, otherwise <em>False</em>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_plasma_membrane_proteins_cspa" title="pypath.legacy.main.PyPath.set_plasma_membrane_proteins_cspa"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_plasma_membrane_proteins_cspa</span></code></a>()</p></td>
<td><p>Creates a vertex attribute <cite>cspa</cite> with value <em>True</em> if the protein is a plasma membrane protein according to CPSA, otherwise <em>False</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_plasma_membrane_proteins_cspa_surfaceome" title="pypath.legacy.main.PyPath.set_plasma_membrane_proteins_cspa_surfaceome"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_plasma_membrane_proteins_cspa_surfaceome</span></code></a>([...])</p></td>
<td><p>Creates a vertex attribute <code class="docutils literal notranslate"><span class="pre">surf</span></code> with value <em>True</em> if the protein is a plasma membrane protein according either to the Cell Surface Protein Atlas or the In Silico Human Surfaceome.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_plasma_membrane_proteins_surfaceome" title="pypath.legacy.main.PyPath.set_plasma_membrane_proteins_surfaceome"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_plasma_membrane_proteins_surfaceome</span></code></a>([...])</p></td>
<td><p>Creates a vertex attribute <cite>ishs</cite> with value <em>True</em> if the protein is a plasma membrane protein according to the In Silico Human Surfaceome, otherwise <em>False</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_receptors" title="pypath.legacy.main.PyPath.set_receptors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_receptors</span></code></a>()</p></td>
<td><p>Creates a vertex attribute <cite>rec</cite> with value <em>True</em> if the protein is a receptor, otherwise <em>False</em>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_signaling_proteins" title="pypath.legacy.main.PyPath.set_signaling_proteins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_signaling_proteins</span></code></a>()</p></td>
<td><p>Creates a vertex attribute <cite>kin</cite> with value <em>True</em> if the protein is a kinase, otherwise <em>False</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_tfs" title="pypath.legacy.main.PyPath.set_tfs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_tfs</span></code></a>([classes])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.set_transcription_factors" title="pypath.legacy.main.PyPath.set_transcription_factors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_transcription_factors</span></code></a>([classes])</p></td>
<td><p>Creates a vertex attribute <cite>tf</cite> with value <em>True</em> if the protein is a transcription factor, otherwise <em>False</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.shortest_path_dist" title="pypath.legacy.main.PyPath.shortest_path_dist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shortest_path_dist</span></code></a>([graph, subset, outfile])</p></td>
<td><p>Computes the distribution of shortest paths for each pair of nodes in the network (or between group(s) of nodes if <em>subset</em> is provided).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.signaling_proteins_list" title="pypath.legacy.main.PyPath.signaling_proteins_list"><code class="xref py py-obj docutils literal notranslate"><span class="pre">signaling_proteins_list</span></code></a>()</p></td>
<td><p>Compiles a list of signaling proteins (as opposed to other proteins like metabolic enzymes, matrix proteins, etc), by looking up a few simple keywords in short description of GO terms.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.signor_pathways" title="pypath.legacy.main.PyPath.signor_pathways"><code class="xref py py-obj docutils literal notranslate"><span class="pre">signor_pathways</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.similarity_groups" title="pypath.legacy.main.PyPath.similarity_groups"><code class="xref py py-obj docutils literal notranslate"><span class="pre">similarity_groups</span></code></a>(groups[, index])</p></td>
<td><p>Computes the similarity index across the given <em>groups</em>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.small_plot" title="pypath.legacy.main.PyPath.small_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">small_plot</span></code></a>(graph, **kwargs)</p></td>
<td><p>This method is deprecated, do not use it.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.sorensen_pathways" title="pypath.legacy.main.PyPath.sorensen_pathways"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sorensen_pathways</span></code></a>([pwlist])</p></td>
<td><p>Computes the Sorensen's similarity index across nodes and edges for the given list of pathway sources (all loaded pathway sources by default).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.source_diagram" title="pypath.legacy.main.PyPath.source_diagram"><code class="xref py py-obj docutils literal notranslate"><span class="pre">source_diagram</span></code></a>([outf])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.source_network" title="pypath.legacy.main.PyPath.source_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">source_network</span></code></a>([font])</p></td>
<td><p>For EMBL branding, use Helvetica Neue Linotype Standard light</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.source_similarity" title="pypath.legacy.main.PyPath.source_similarity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">source_similarity</span></code></a>([outfile])</p></td>
<td><p>Computes the Sorensen's similarity index across nodes and edges for all the sources available (already loaded in the network) and saves them into table files.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.source_stats" title="pypath.legacy.main.PyPath.source_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">source_stats</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.sources_hist" title="pypath.legacy.main.PyPath.sources_hist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sources_hist</span></code></a>()</p></td>
<td><p>Counts the number of sources per interaction in the graph and saves them into a file named <code class="docutils literal notranslate"><span class="pre">source_num</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.sources_overlap" title="pypath.legacy.main.PyPath.sources_overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sources_overlap</span></code></a>([diagonal])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.sources_venn_data" title="pypath.legacy.main.PyPath.sources_venn_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sources_venn_data</span></code></a>([fname, return_data])</p></td>
<td><p>Computes the overlap in number of interactions for all pairs of sources.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.stats" title="pypath.legacy.main.PyPath.stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">stats</span></code></a>(method[, keep_collection])</p></td>
<td><p>Creates a collection of entities over the network according to <code class="docutils literal notranslate"><span class="pre">method</span></code> and counts them.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.straight_between" title="pypath.legacy.main.PyPath.straight_between"><code class="xref py py-obj docutils literal notranslate"><span class="pre">straight_between</span></code></a>(id_a, id_b)</p></td>
<td><p>Finds an edge between the provided node names.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.string_effects" title="pypath.legacy.main.PyPath.string_effects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">string_effects</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.sum_in_complex" title="pypath.legacy.main.PyPath.sum_in_complex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sum_in_complex</span></code></a>([csources, graph])</p></td>
<td><p>Returns the total number of edges in the network falling between two members of the same complex.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.summaries_tab" title="pypath.legacy.main.PyPath.summaries_tab"><code class="xref py py-obj docutils literal notranslate"><span class="pre">summaries_tab</span></code></a>([outfile, return_table])</p></td>
<td><p>Creates a table from resource vs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.table_latex" title="pypath.legacy.main.PyPath.table_latex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">table_latex</span></code></a>(fname, header, data[, sum_row, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.third_source_directions" title="pypath.legacy.main.PyPath.third_source_directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">third_source_directions</span></code></a>([graph, ...])</p></td>
<td><p>This method calls a series of methods to get additional direction &amp; effect information from sources having no literature curated references, but giving sufficient evidence about the directionality for interactions already supported by literature evidences from other sources.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.tissue_network" title="pypath.legacy.main.PyPath.tissue_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tissue_network</span></code></a>(tissue[, graph])</p></td>
<td><p>Returns a network which includes the proteins expressed in certain tissue according to ProteomicsDB.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.transcription_factors" title="pypath.legacy.main.PyPath.transcription_factors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">transcription_factors</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.uniprot" title="pypath.legacy.main.PyPath.uniprot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">uniprot</span></code></a>(uniprot)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the UniProt can be found in the default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.uniprots" title="pypath.legacy.main.PyPath.uniprots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">uniprots</span></code></a>(uniprots)</p></td>
<td><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object for a list of UniProt IDs omitting those could not be found in the default undirected graph.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.uniq_node_list" title="pypath.legacy.main.PyPath.uniq_node_list"><code class="xref py py-obj docutils literal notranslate"><span class="pre">uniq_node_list</span></code></a>(lst)</p></td>
<td><p>Returns a given list of nodes containing only the unique elements.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.uniq_ptm" title="pypath.legacy.main.PyPath.uniq_ptm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">uniq_ptm</span></code></a>(ptms)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.uniq_ptms" title="pypath.legacy.main.PyPath.uniq_ptms"><code class="xref py py-obj docutils literal notranslate"><span class="pre">uniq_ptms</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up" title="pypath.legacy.main.PyPath.up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up</span></code></a>(uniprot)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the UniProt can be found in the default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_affected_by" title="pypath.legacy.main.PyPath.up_affected_by"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_affected_by</span></code></a>(uniprot)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_affects" title="pypath.legacy.main.PyPath.up_affects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_affects</span></code></a>(uniprot)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_edge" title="pypath.legacy.main.PyPath.up_edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_edge</span></code></a>(source, target[, directed])</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_in_directed" title="pypath.legacy.main.PyPath.up_in_directed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_in_directed</span></code></a>(uniprot)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_in_undirected" title="pypath.legacy.main.PyPath.up_in_undirected"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_in_undirected</span></code></a>(uniprot)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_inhibited_by" title="pypath.legacy.main.PyPath.up_inhibited_by"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_inhibited_by</span></code></a>(uniprot)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_inhibits" title="pypath.legacy.main.PyPath.up_inhibits"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_inhibits</span></code></a>(uniprot)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_neighborhood" title="pypath.legacy.main.PyPath.up_neighborhood"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_neighborhood</span></code></a>(uniprots[, order, mode])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_neighbors" title="pypath.legacy.main.PyPath.up_neighbors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_neighbors</span></code></a>(uniprot[, mode])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_stimulated_by" title="pypath.legacy.main.PyPath.up_stimulated_by"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_stimulated_by</span></code></a>(uniprot)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.up_stimulates" title="pypath.legacy.main.PyPath.up_stimulates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_stimulates</span></code></a>(uniprot)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_adjlist" title="pypath.legacy.main.PyPath.update_adjlist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_adjlist</span></code></a>([graph, mode])</p></td>
<td><p>Creates an adjacency list in a list of sets format.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_attrs" title="pypath.legacy.main.PyPath.update_attrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_attrs</span></code></a>()</p></td>
<td><p>Updates the node and edge attributes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_cats" title="pypath.legacy.main.PyPath.update_cats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_cats</span></code></a>()</p></td>
<td><p>Makes sure that the <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.has_cats</span></code> attribute is an up to date [set] of all categories in the current network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_db_dict" title="pypath.legacy.main.PyPath.update_db_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_db_dict</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_pathway_types" title="pypath.legacy.main.PyPath.update_pathway_types"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_pathway_types</span></code></a>()</p></td>
<td><p>Updates the pathway types attribute (<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.pathway_types</span></code>) according to the loaded resources of the undirected network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_pathways" title="pypath.legacy.main.PyPath.update_pathways"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_pathways</span></code></a>()</p></td>
<td><p>Makes sure that the <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.pathways</span></code> attribute is an up to date [dict] of all pathways and their sources in the current network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_sources" title="pypath.legacy.main.PyPath.update_sources"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_sources</span></code></a>()</p></td>
<td><p>Makes sure that the <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.sources</span></code> attribute is an up to date [list] of all sources in the current network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_summaries" title="pypath.legacy.main.PyPath.update_summaries"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_summaries</span></code></a>()</p></td>
<td><p>Creates a dict with many summarizing and comparative statistics about the resources in the current network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_vertex_sources" title="pypath.legacy.main.PyPath.update_vertex_sources"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_vertex_sources</span></code></a>()</p></td>
<td><p>Updates the all the vertex attributes <code class="docutils literal notranslate"><span class="pre">'sources'</span></code> and <code class="docutils literal notranslate"><span class="pre">'references'</span></code> according to their related edges (on the undirected graph).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_vindex" title="pypath.legacy.main.PyPath.update_vindex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_vindex</span></code></a>()</p></td>
<td><p>This is deprecated.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.update_vname" title="pypath.legacy.main.PyPath.update_vname"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update_vname</span></code></a>()</p></td>
<td><p>Fast lookup of node names and indexes, these are hold in a [list] and a [dict] as well.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.ups" title="pypath.legacy.main.PyPath.ups"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ups</span></code></a>(uniprots)</p></td>
<td><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object for a list of UniProt IDs omitting those could not be found in the default undirected graph.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.v" title="pypath.legacy.main.PyPath.v"><code class="xref py py-obj docutils literal notranslate"><span class="pre">v</span></code></a>(identifier)</p></td>
<td><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier is a valid vertex index in the default undirected graph, or a UniProt ID or GeneSymbol which can be found in the default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">vertex_name</span></code>(v)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.vertex_pathways" title="pypath.legacy.main.PyPath.vertex_pathways"><code class="xref py py-obj docutils literal notranslate"><span class="pre">vertex_pathways</span></code></a>()</p></td>
<td><p>Some resources assignes interactions some others proteins to pathways.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.vsgs" title="pypath.legacy.main.PyPath.vsgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">vsgs</span></code></a>()</p></td>
<td><p>Returns a generator sequence of the node names as GeneSymbols [str] (from the undirected graph).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.vsup" title="pypath.legacy.main.PyPath.vsup"><code class="xref py py-obj docutils literal notranslate"><span class="pre">vsup</span></code></a>()</p></td>
<td><p>Returns a generator sequence of the node names as UniProt IDs [str] (from the undirected graph).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.wang_effects" title="pypath.legacy.main.PyPath.wang_effects"><code class="xref py py-obj docutils literal notranslate"><span class="pre">wang_effects</span></code></a>([graph])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.write_table" title="pypath.legacy.main.PyPath.write_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_table</span></code></a>(tbl, outfile[, sep, cut, ...])</p></td>
<td><p>Writes a given table to a file.</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_references</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">ecount</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">edge_attributes</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">es</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.legacy.main.PyPath.resources" title="pypath.legacy.main.PyPath.resources"><code class="xref py py-obj docutils literal notranslate"><span class="pre">resources</span></code></a></p></td>
<td><p>All network resources.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">vcount</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">vertex_attributes</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">vs</span></code></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.acsn_effects">
<span class="sig-name descname"><span class="pre">acsn_effects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.acsn_effects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.acsn_effects" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.add_genesets">
<span class="sig-name descname"><span class="pre">add_genesets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesets</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.add_genesets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.add_genesets" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.add_grouped_eattr">
<span class="sig-name descname"><span class="pre">add_grouped_eattr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edge</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.add_grouped_eattr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.add_grouped_eattr" title="Permalink to this definition">§</a></dt>
<dd><p>Merges (or creates) a given edge attribute as [dict] of [list]
values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> (<em>int</em>) – Edge index where the given attribute value is to be merged
or created.</p></li>
<li><p><strong>attr</strong> (<em>str</em>) – The name of the attribute. If such attribute does not exist
in the network edges, it will be created on all edges (as an
empty [dict], <em>value</em> will only be assigned to the given
<em>edge</em> and <em>group</em>).</p></li>
<li><p><strong>group</strong> (<em>str</em>) – The key of the attribute dictionary where <em>value</em> is to be
assigned.</p></li>
<li><p><strong>value</strong> (<em>list</em>) – The value of the attribute to be assigned/merged.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.add_grouped_set_eattr">
<span class="sig-name descname"><span class="pre">add_grouped_set_eattr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edge</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.add_grouped_set_eattr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.add_grouped_set_eattr" title="Permalink to this definition">§</a></dt>
<dd><p>Merges (or creates) a given edge attribute as [dict] of [set]
values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> (<em>int</em>) – Edge index where the given attribute value is to be merged
or created.</p></li>
<li><p><strong>attr</strong> (<em>str</em>) – The name of the attribute. If such attribute does not exist
in the network edges, it will be created on all edges (as an
empty [dict], <em>value</em> will only be assigned to the given
<em>edge</em> and <em>group</em>).</p></li>
<li><p><strong>group</strong> (<em>str</em>) – The key of the attribute dictionary where <em>value</em> is to be
assigned.</p></li>
<li><p><strong>value</strong> (<em>set</em>) – The value of the attribute to be assigned/merged.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.add_list_eattr">
<span class="sig-name descname"><span class="pre">add_list_eattr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edge</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.add_list_eattr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.add_list_eattr" title="Permalink to this definition">§</a></dt>
<dd><p>Merges (or creates) a given edge attribute as [list].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> (<em>int</em>) – Edge index where the given attribute value is to be merged
or created.</p></li>
<li><p><strong>attr</strong> (<em>str</em>) – The name of the attribute. If such attribute does not exist
in the network edges, it will be created on all edges (as an
empty [list], <em>value</em> will only be assigned to the given
<em>edge</em>).</p></li>
<li><p><strong>value</strong> (<em>list</em>) – The value of the attribute to be assigned/merged.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.add_set_eattr">
<span class="sig-name descname"><span class="pre">add_set_eattr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edge</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.add_set_eattr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.add_set_eattr" title="Permalink to this definition">§</a></dt>
<dd><p>Merges (or creates) a given edge attribute as [set].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> (<em>int</em>) – Edge index where the given attribute value is to be merged
or created.</p></li>
<li><p><strong>attr</strong> (<em>str</em>) – The name of the attribute. If such attribute does not exist
in the network edges, it will be created on all edges (as an
empty [set], <em>value</em> will only be assigned to the given
<em>edge</em>).</p></li>
<li><p><strong>value</strong> (<em>set</em>) – The value of the attribute to be assigned/merged.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.affects">
<span class="sig-name descname"><span class="pre">affects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.affects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.affects" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.all_between">
<span class="sig-name descname"><span class="pre">all_between</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_b</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.all_between"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.all_between" title="Permalink to this definition">§</a></dt>
<dd><p>Checks for any edges (in any direction) between the provided
nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_a</strong> (<em>str</em>) – The name of the source node.</p></li>
<li><p><strong>id_b</strong> (<em>str</em>) – The name of the target node.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>dict</em>) – Contains information on the directionality of
the requested edge. Keys are <code class="docutils literal notranslate"><span class="pre">'ab'</span></code> and <code class="docutils literal notranslate"><span class="pre">'ba'</span></code>, denoting
the straight/reverse directionalities respectively. Values
are [list] whose elements are the edge ID or <code class="docutils literal notranslate"><span class="pre">None</span></code>
according to the existance of that edge in the following
categories: undirected, straight and reverse (in that
order).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.all_neighbours">
<span class="sig-name descname"><span class="pre">all_neighbours</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indices</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.all_neighbours"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.all_neighbours" title="Permalink to this definition">§</a></dt>
<dd><p>Looks for the first neighbours of all the nodes and creates an
attribute (<code class="docutils literal notranslate"><span class="pre">'neighbours'</span></code>) on each one of them containing a
list of their UniProt IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>indices</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to list the
neighbour nodes indices or their UniProt IDs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.apply_list">
<span class="sig-name descname"><span class="pre">apply_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_or_edge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'node'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.apply_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.apply_list" title="Permalink to this definition">§</a></dt>
<dd><p>Creates vertex or edge attribute based on a list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – The name of the list to be added as attribute. Must have
been previously loaded with
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.load_list()</span></code> or other methods.
See description of <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.lists</span></code>
attribute for more information.</p></li>
<li><p><strong>node_or_edge</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'node'</span></code> by default. Whether the attribute list
is to be added to the nodes or to the edges.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.apply_negative">
<span class="sig-name descname"><span class="pre">apply_negative</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">settings</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.apply_negative"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.apply_negative" title="Permalink to this definition">§</a></dt>
<dd><p>Loads a negative interaction source (e.g.: Negatome) into the
current network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>settings</strong> (<em>pypath.input_formats.NetworkInput</em>) – <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.input_formats.NetworkInput</span></code> instance
containing the detailed definition of the input format to
the downloaded file. For instance
<code class="xref py py-data docutils literal notranslate"><span class="pre">pypath.data_formats.negative['negatome']</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.basic_stats">
<span class="sig-name descname"><span class="pre">basic_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">latex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">caption</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">latex_hdr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'HelveticaNeueLTStd-LtCn'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'%s'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_cats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['p',</span> <span class="pre">'m',</span> <span class="pre">'i',</span> <span class="pre">'r']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">urls</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.basic_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.basic_stats" title="Permalink to this definition">§</a></dt>
<dd><p>Returns basic numbers about the network resources, e.g. edge and
node counts.</p>
<dl class="simple">
<dt>latex</dt><dd><p>Return table in a LaTeX document. This can be compiled by
PDFLaTeX:
latex stats.tex</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.basic_stats_intergroup">
<span class="sig-name descname"><span class="pre">basic_stats_intergroup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">groupA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.basic_stats_intergroup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.basic_stats_intergroup" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.cancer_drivers_list">
<span class="sig-name descname"><span class="pre">cancer_drivers_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intogen_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.cancer_drivers_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.cancer_drivers_list" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the list of cancer drivers. Contains information from
COSMIC (needs user log in credentials) and IntOGen (if provided)
and adds the attribute to the undirected network nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>intogen_file</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Path to the data file. Can
also be [function] that provides the data. In general,
anything accepted by
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.input_formats.NetworkInput.input</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.cancer_gene_census_list">
<span class="sig-name descname"><span class="pre">cancer_gene_census_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.cancer_gene_census_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.cancer_gene_census_list" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the list of cancer driver proteins from the COSMIC Cancer
Gene Census.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.clean_graph">
<span class="sig-name descname"><span class="pre">clean_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organisms_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.clean_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.clean_graph" title="Permalink to this definition">§</a></dt>
<dd><p>Removes multiple edges, unknown molecules and those from wrong
taxon. Multiple edges will be combined by
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.combine_attr()</span></code> method.
Loops will be deleted unless the attribute
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.loops</span></code> is set to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>organisms_allowed</strong> (<em>set</em>) – NCBI Taxonomy identifiers [int] of the organisms allowed
in the network.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.collapse_by_name">
<span class="sig-name descname"><span class="pre">collapse_by_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.collapse_by_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.collapse_by_name" title="Permalink to this definition">§</a></dt>
<dd><p>Collapses nodes with the same name by copying and merging
all edges and attributes. Operates directly on the provided
network object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network for which the
nodes are to be collapsed. If none is provided, takes
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.graph</span></code> (undirected network) by
default.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.collect">
<span class="sig-name descname"><span class="pre">collect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.collect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.collect" title="Permalink to this definition">§</a></dt>
<dd><p>Collects various entities over the network according to <code class="docutils literal notranslate"><span class="pre">method</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.combine_attr">
<span class="sig-name descname"><span class="pre">combine_attr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_method=&lt;built-in</span> <span class="pre">function</span> <span class="pre">max&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.combine_attr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.combine_attr" title="Permalink to this definition">§</a></dt>
<dd><p>Combines multiple attributes into one. This method attempts
to find out which is the best way to combine attributes.</p>
<blockquote>
<div><ul class="simple">
<li><p>If there is only one value or one of them is None, then
returns the one available.</p></li>
<li><p>Lists: concatenates unique values of lists.</p></li>
<li><p>Numbers: returns the greater by default or calls
<em>num_method</em> if given.</p></li>
<li><p>Sets: returns the union.</p></li>
<li><p>Dictionaries: calls <code class="xref py py-func docutils literal notranslate"><span class="pre">pypath.common.merge_dicts()</span></code>.</p></li>
<li><p>Direction: calls their special
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.Direction.merge()</span></code> method.</p></li>
</ul>
</div></blockquote>
<p>Works on more than 2 attributes recursively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst</strong> (<em>list</em>) – List of one or two attribute values.</p></li>
<li><p><strong>num_method</strong> (<em>function</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">max</span></code> by default. Method to merge numeric
attributes.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.communities">
<span class="sig-name descname"><span class="pre">communities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.communities"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.communities" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.complex_comembership_network">
<span class="sig-name descname"><span class="pre">complex_comembership_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.complex_comembership_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.complex_comembership_network" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.complexes">
<span class="sig-name descname"><span class="pre">complexes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">methods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['3dcomplexes',</span> <span class="pre">'havugimana',</span> <span class="pre">'corum',</span> <span class="pre">'complexportal',</span> <span class="pre">'compleat']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.complexes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.complexes" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.complexes_in_network">
<span class="sig-name descname"><span class="pre">complexes_in_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csource</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'corum'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.complexes_in_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.complexes_in_network" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.compounds_from_chembl">
<span class="sig-name descname"><span class="pre">compounds_from_chembl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chembl_mysql</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">andor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'or'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['B',</span> <span class="pre">'F']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">relationship_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['D',</span> <span class="pre">'H']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multi_query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.compounds_from_chembl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.compounds_from_chembl" title="Permalink to this definition">§</a></dt>
<dd><p>Loads compound data from ChEMBL to the network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chebl_mysql</strong> (<em>tuple</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Contains the MySQL parameters
used by the <code class="xref py py-mod docutils literal notranslate"><span class="pre">pypath.mapping</span></code> module to loadthe
ChEMBL ID conversion tables. If none is passed, takes the
current instance <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.chembl_mysql</span></code>
attribute.</p></li>
<li><p><strong>nodes</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. List of node indices for
which the information is to be loaded. If none is provided
calls the method <code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.get_sub()</span></code> with
the provided <em>crit</em> parameter.</p></li>
<li><p><strong>crit</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Defines the critical
attributes to generate a subnetwork to extract the nodes in
case <em>nodes</em> is not provided. Keys are <code class="docutils literal notranslate"><span class="pre">'edge'</span></code> and
<code class="docutils literal notranslate"><span class="pre">'node'</span></code> and values are [dict] containing the critical
attribute names [str] and values are [set] containing those
attributes of the nodes/edges that are to be kept in the
subnetwork. If none is provided, takes the whole network.</p></li>
<li><p><strong>andor</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'or'</span></code> by default. Determines the search mode
for the subnetwork generation (if <code class="docutils literal notranslate"><span class="pre">nodes=None</span></code>). See
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.search_attr_or()</span></code> and
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.search_attr_and()</span></code> for more
details.</p></li>
<li><p><strong>assay_types</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">['B',</span> <span class="pre">'F']</span></code> by default. Types of assay to query
Options are: <code class="docutils literal notranslate"><span class="pre">'A'</span></code> (ADME), <code class="docutils literal notranslate"><span class="pre">'B'</span></code> (Binding),``’F’``
(Functional), <code class="docutils literal notranslate"><span class="pre">'P'</span></code> (Physicochemical), <code class="docutils literal notranslate"><span class="pre">'T'</span></code> (Toxicity)
and/or <code class="docutils literal notranslate"><span class="pre">'U'</span></code> (Unassigned).</p></li>
<li><p><strong>relationship_types</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">['D',</span> <span class="pre">'H']</span></code> by default. Assay relationship
types to query. Possible values are: <code class="docutils literal notranslate"><span class="pre">'D'</span></code> (Direct protein
target assigned), <code class="docutils literal notranslate"><span class="pre">'H'</span></code> (Homologous protein target
assigned), <code class="docutils literal notranslate"><span class="pre">'M'</span></code> (Molecular target other than protein
assigned), <code class="docutils literal notranslate"><span class="pre">'N'</span></code> (Non-molecular target assigned), <code class="docutils literal notranslate"><span class="pre">'S'</span></code>
(Subcellular target assigned) and/or <code class="docutils literal notranslate"><span class="pre">'U'</span></code> (Default value,
target has yet to be curated).</p></li>
<li><p><strong>multi_query</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Not used.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments for
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.chembl.Chembl.compounds_targets()</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.consistency">
<span class="sig-name descname"><span class="pre">consistency</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.consistency"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.consistency" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.copy" title="Permalink to this definition">§</a></dt>
<dd><p>Copies another <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.main.PyPath</span></code> instance into the
current one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<em>pypath.main.PyPath</em>) – The instance to be copied from.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.copy_edges">
<span class="sig-name descname"><span class="pre">copy_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sources</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">move</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.copy_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.copy_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Copies edges from <em>sources</em> node(s) to another one (<em>target</em>),
keeping attributes and directions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sources</strong> (<em>list</em>) – Contains the vertex index(es) [int] of the node(s) to be
copied or moved.</p></li>
<li><p><strong>target</strong> (<em>int</em>) – Vertex index where edges and attributes are to be copied to.</p></li>
<li><p><strong>move</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to perform copy or
move (remove or keep the source edges).</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network graph object from
which the nodes are to be merged. If none is passed, takes
the undirected network graph.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.count_sol">
<span class="sig-name descname"><span class="pre">count_sol</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.count_sol"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.count_sol" title="Permalink to this definition">§</a></dt>
<dd><p>Counts the number of nodes with zero degree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>int</em>) – The number of nodes with zero degree.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">collection_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_total</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_percent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_cat_total</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.counts" title="Permalink to this definition">§</a></dt>
<dd><p>Collects various entities over the network according to <code class="docutils literal notranslate"><span class="pre">method</span></code>
and counts them in total and by resources.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.coverage">
<span class="sig-name descname"><span class="pre">coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.coverage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.coverage" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the coverage (range [0, 1]) of a list of nodes against
the current (undirected) network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lst</strong> (<em>set</em>) – Can also be [list] (will be converted to [set]) or [str]. In
the latter case it will retrieve the list with that name (if
such list exists in <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.lists</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.curation_effort">
<span class="sig-name descname"><span class="pre">curation_effort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.curation_effort"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.curation_effort" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>set</em> of reference-interactions pairs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.curation_effort_by_resource">
<span class="sig-name descname"><span class="pre">curation_effort_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.curation_effort_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.curation_effort_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>A <em>dict</em> with resources as keys and <a href="#id21"><span class="problematic" id="id22">*</span></a>set*s of curation items
(interaction-reference pairs) as values.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.curation_stats">
<span class="sig-name descname"><span class="pre">curation_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">by_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.curation_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.curation_stats" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.curation_tab">
<span class="sig-name descname"><span class="pre">curation_tab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'curation_stats.tex'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_cats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['p',</span> <span class="pre">'m',</span> <span class="pre">'i',</span> <span class="pre">'r']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'normalsize'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.curation_tab"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.curation_tab" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.curators_work">
<span class="sig-name descname"><span class="pre">curators_work</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.curators_work"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.curators_work" title="Permalink to this definition">§</a></dt>
<dd><p>Computes and prints an estimation of how many years of curation
took to achieve the amount of information on the network.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.databases_similarity">
<span class="sig-name descname"><span class="pre">databases_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'simpson'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.databases_similarity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.databases_similarity" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the similarity across databases according to a given
index metric. Computes the similarity across the loaded
resources (listed in <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.sources</span></code> in
terms of nodes and edges separately.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'simpson'</span></code> by default. The type of index metric
to use to compute the similarity. Options are <code class="docutils literal notranslate"><span class="pre">'simpson'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'sorensen'</span></code> and <code class="docutils literal notranslate"><span class="pre">'jaccard'</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>dict</em>) – Nested dictionaries (three levels). First-level
keys are <code class="docutils literal notranslate"><span class="pre">'nodes'</span></code> and <code class="docutils literal notranslate"><span class="pre">'edges'</span></code>, then second and third
levels correspond to sources names which map to the
similarity index between those sources [float].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.degree_dist">
<span class="sig-name descname"><span class="pre">degree_dist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.degree_dist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.degree_dist" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the degree distribution over all nodes of the network.
If <em>group</em> is provided, also across nodes of that group(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix for the file name(s).</p></li>
<li><p><strong>g</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network over which to
compute the degree distribution. If none is passed, takes
the undirected network of the current instance.</p></li>
<li><p><strong>group</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Additional group(s) name(s)
[str] of node attributes to subset the network and compute
its degree distribution.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.degree_dists">
<span class="sig-name descname"><span class="pre">degree_dists</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.degree_dists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.degree_dists" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the degree distribution for all the different network
sources. This is, for each source, the subnetwork comprising all
interactions coming from it is extracted and the degree
distribution information is computed and saved into a file.
A file is created for each resource under the name
<code class="docutils literal notranslate"><span class="pre">'pwnet-&lt;session_id&gt;-degdist-&lt;resource&gt;''</span></code>. Files are stored
in <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.outdir</span></code> (<code class="docutils literal notranslate"><span class="pre">'results'</span></code> by
default).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.delete_by_organism">
<span class="sig-name descname"><span class="pre">delete_by_organism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organisms_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.delete_by_organism"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.delete_by_organism" title="Permalink to this definition">§</a></dt>
<dd><p>Removes the proteins of all organisms which are not given in
<em>tax</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>organisms_allowed</strong> (<em>list</em><em>,</em><em>set</em>) – List of NCBI Taxonomy IDs [int] of the organism(s) that are
to be kept.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.delete_by_source">
<span class="sig-name descname"><span class="pre">delete_by_source</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertexAttrsToDel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgeAttrsToDel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.delete_by_source"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.delete_by_source" title="Permalink to this definition">§</a></dt>
<dd><p>Deletes nodes and edges from the network according to a provided
source name. Optionally can also remove the given list of
attributes from nodes and/or edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<em>str</em>) – Name of the source from which the nodes and edges have to be
removed.</p></li>
<li><p><strong>vertexAttrsToDel</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Contains the names [str] of
the attributes to be removed from the nodes.</p></li>
<li><p><strong>edgeAttrsToDel</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Contains the names [str] of
the attributes to be removed from the edges.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.delete_unknown">
<span class="sig-name descname"><span class="pre">delete_unknown</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organisms_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'protein'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_name_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.delete_unknown"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.delete_unknown" title="Permalink to this definition">§</a></dt>
<dd><p>Removes those items which are not in the list of all default
IDs of the organisms. By default, it means to remove all protein
nodes not having a human UniProt ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>typ</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'protein'</span></code> by default. Determines the molecule
type. These can be <code class="docutils literal notranslate"><span class="pre">'protein'</span></code>, <code class="docutils literal notranslate"><span class="pre">'drug'</span></code>, <code class="docutils literal notranslate"><span class="pre">'lncrna'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'mirna'</span></code> or any other type defined in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.default_name_type</span></code>.</p></li>
<li><p><strong>default_name_type</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The default name type for the
given molecular species. If none is specified takes it from
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.default_name_type</span></code> (e.g.: for
<code class="docutils literal notranslate"><span class="pre">'protein'</span></code>, default is <code class="docutils literal notranslate"><span class="pre">'uniprot'</span></code>).</p></li>
<li><p><strong>organisms_allowed</strong> (<em>set</em>) – NCBI Taxonomy identifiers [int] of the organisms allowed in
the network.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.delete_unmapped">
<span class="sig-name descname"><span class="pre">delete_unmapped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.delete_unmapped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.delete_unmapped" title="Permalink to this definition">§</a></dt>
<dd><p>Checks the network for any existing unmapped node and removes
it.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dgenesymbol">
<span class="sig-name descname"><span class="pre">dgenesymbol</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.dgenesymbol"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.dgenesymbol" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default directed network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;genesymbol<span class="classifier">str</span></dt><dd><p>GeneSymbol.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dgenesymbols">
<span class="sig-name descname"><span class="pre">dgenesymbols</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbols</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.dgenesymbols"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.dgenesymbols" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dgs">
<span class="sig-name descname"><span class="pre">dgs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dgs" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default directed network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;genesymbol<span class="classifier">str</span></dt><dd><p>GeneSymbol.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dgss">
<span class="sig-name descname"><span class="pre">dgss</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbols</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dgss" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dneighbors">
<span class="sig-name descname"><span class="pre">dneighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ALL'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.dneighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.dneighbors" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dp">
<span class="sig-name descname"><span class="pre">dp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dp" title="Permalink to this definition">§</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.
Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;identifier<span class="classifier">int, str</span></dt><dd><p>Vertex index (int) or GeneSymbol (str) or UniProt ID (str) or
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dproteins">
<span class="sig-name descname"><span class="pre">dproteins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifiers</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dproteins" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dps">
<span class="sig-name descname"><span class="pre">dps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifiers</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dps" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.duniprot">
<span class="sig-name descname"><span class="pre">duniprot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.duniprot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.duniprot" title="Permalink to this definition">§</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.uniprot(),</span> <span class="pre">just</span> <span class="pre">for</span> <span class="pre">directed</span> <span class="pre">graph.</span>
<span class="pre">Returns</span> <span class="pre">``igraph.Vertex()</span></code> object if the UniProt
can be found in the default directed network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;uniprot<span class="classifier">str</span></dt><dd><p>UniProt ID.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.duniprots">
<span class="sig-name descname"><span class="pre">duniprots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprots</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.duniprots"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.duniprots" title="Permalink to this definition">§</a></dt>
<dd><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
directed graph.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dup">
<span class="sig-name descname"><span class="pre">dup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dup" title="Permalink to this definition">§</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.uniprot(),</span> <span class="pre">just</span> <span class="pre">for</span> <span class="pre">directed</span> <span class="pre">graph.</span>
<span class="pre">Returns</span> <span class="pre">``igraph.Vertex()</span></code> object if the UniProt
can be found in the default directed network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;uniprot<span class="classifier">str</span></dt><dd><p>UniProt ID.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dups">
<span class="sig-name descname"><span class="pre">dups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprots</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dups" title="Permalink to this definition">§</a></dt>
<dd><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
directed graph.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dv">
<span class="sig-name descname"><span class="pre">dv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dv" title="Permalink to this definition">§</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.
Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;identifier<span class="classifier">int, str</span></dt><dd><p>Vertex index (int) or GeneSymbol (str) or UniProt ID (str) or
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.dvs">
<span class="sig-name descname"><span class="pre">dvs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifiers</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.dvs" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edge_exists">
<span class="sig-name descname"><span class="pre">edge_exists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_b</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edge_exists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edge_exists" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a tuple of vertex indices if edge doesn’t exist,
otherwise, the edge ID. Not sensitive to direction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_a</strong> (<em>str</em>) – Name of the source node.</p></li>
<li><p><strong>id_b</strong> (<em>str</em>) – Name of the target node.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>int</em>) – The edge index, if exists such edge. Otherwise,
[tuple] of [int] corresponding to the node IDs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edge_loc">
<span class="sig-name descname"><span class="pre">edge_loc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">topn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edge_loc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edge_loc" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edge_names">
<span class="sig-name descname"><span class="pre">edge_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">e</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edge_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edge_names" title="Permalink to this definition">§</a></dt>
<dd><p>Returns the node names of a given edge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>e</strong> (<em>int</em>) – The edge index.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>tuple</em>) – Contains the source and target node names of
the edge [str].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edges_3d">
<span class="sig-name descname"><span class="pre">edges_3d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">methods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['inputs.instruct.get_instruct',</span> <span class="pre">'inputs.i3d.get_i3d']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edges_3d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edges_3d" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edges_between">
<span class="sig-name descname"><span class="pre">edges_between</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">group1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edges_between"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edges_between" title="Permalink to this definition">§</a></dt>
<dd><p>Selects edges between two groups of vertex IDs.
Returns set of edge IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>group1</strong><strong>,</strong><strong>group2</strong> (<em>set</em>) – List, set or tuple of vertex IDs.</p></li>
<li><p><strong>directed</strong> (<em>bool</em>) – Only edges with direction <cite>group1 -&gt; group2</cite> selected.</p></li>
<li><p><strong>strict</strong> (<em>bool</em>) – Edges with no direction information still selected even if
<code class="docutils literal notranslate"><span class="pre">directed</span></code> is <cite>False</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edges_expression">
<span class="sig-name descname"><span class="pre">edges_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func=&lt;function</span> <span class="pre">PyPath.&lt;lambda&gt;&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edges_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edges_expression" title="Permalink to this definition">§</a></dt>
<dd><p>Executes function <cite>func</cite> for each pairs of connected proteins in the
network, for every expression dataset. By default, <cite>func</cite> simply
gives the product the (normalized) expression values.</p>
<dl class="simple">
<dt>func<span class="classifier">callable</span></dt><dd><p>Function to handle 2 vectors (pandas.Series() objects), should
return one vector of the same length.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edges_in_complexes">
<span class="sig-name descname"><span class="pre">edges_in_complexes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['corum']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edges_in_complexes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edges_in_complexes" title="Permalink to this definition">§</a></dt>
<dd><p>Creates edge attributes <code class="docutils literal notranslate"><span class="pre">complexes</span></code> and <code class="docutils literal notranslate"><span class="pre">in_complex</span></code>.
These are both dicts where the keys are complex resources.
The values in <code class="docutils literal notranslate"><span class="pre">complexes</span></code> are the list of complex names
both the source and the target vertices belong to.
The values <code class="docutils literal notranslate"><span class="pre">in_complex</span></code> are boolean values whether there
is at least one complex in the given resources both the
source and the target vertex of the edge belong to.</p>
<dl class="simple">
<dt>&#64;csources<span class="classifier">list</span></dt><dd><p>List of complex resources. Should be already loaded.</p>
</dd>
<dt>&#64;graph<span class="classifier">igraph.Graph()</span></dt><dd><p>The graph object to do the calculations on.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edges_ptms">
<span class="sig-name descname"><span class="pre">edges_ptms</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edges_ptms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edges_ptms" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.edgeseq_inverse">
<span class="sig-name descname"><span class="pre">edgeseq_inverse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.edgeseq_inverse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.edgeseq_inverse" title="Permalink to this definition">§</a></dt>
<dd><p>Returns the sequence of all edge indexes that are not in
the argument <em>edges</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>edges</strong> (<em>set</em>) – Sequence of edge indices [int] that will not be returned.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – Contains all edge indices [int] of the
undirected network except the ones on <em>edges</em> argument.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.entities_by_resource">
<span class="sig-name descname"><span class="pre">entities_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.entities_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.entities_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>dict</em> of <a href="#id23"><span class="problematic" id="id24">*</span></a>set*s with resources as keys and sets of
entities as values.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.entities_by_resources">
<span class="sig-name descname"><span class="pre">entities_by_resources</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.entities_by_resources"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.entities_by_resources" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a dict of sets with resources as keys and sets of entity IDs
as values.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.export_dot">
<span class="sig-name descname"><span class="pre">export_dot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nodes=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels='genesymbol'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges_filter=&lt;function</span> <span class="pre">PyPath.&lt;lambda&gt;&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes_filter=&lt;function</span> <span class="pre">PyPath.&lt;lambda&gt;&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_sources=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir_sources=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_object=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dot=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_graphics=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prog='neato'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auto_edges=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_nodes=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defaults={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">**kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.export_dot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.export_dot" title="Permalink to this definition">§</a></dt>
<dd><p>Builds a pygraphviz.AGraph() object with filtering the edges
and vertices along arbitrary criteria.
Returns the Agraph object if requesred, or exports the dot
file, or saves the graphics.</p>
<p>&#64;nodes : list
List of vertex ids to be included.
&#64;edges : list
List of edge ids to be included.
&#64;directed : bool
Create a directed or undirected graph.
&#64;labels : str
Name type to be used as id/label in the dot format.
&#64;edges_filter : function
Function to filter edges, accepting igraph.Edge as argument.
&#64;nodes_filter : function
Function to filter vertices, accepting igraph.Vertex as argument.
&#64;edge_sources : list
Sources to be included.
&#64;dir_sources : list
Direction and effect sources to be included.
&#64;graph : igraph.Graph
The graph object to export.
&#64;return_object : bool
Whether to return the pygraphviz.AGraph object.
&#64;save_dot : str
Filename to export the dot file to.
&#64;save_graphics : str
Filename to export the graphics, the extension defines the format.
&#64;prog : str
The graphviz layout algorithm to use.
&#64;format : str
The graphics format passed to pygraphviz.AGrapg().draw().
&#64;hide : bool
Hide filtered edges instead of omit them.
&#64;hide nodes : list
Nodes to hide. List of vertex ids.
&#64;auto_edges : str
Automatic, built-in style for edges.
‘DIRECTIONS’ or ‘RESOURCE_CATEGORIES’ are supported.
&#64;font : str
Font to use for labels.
For using more than one fonts refer to graphviz attributes with constant values
or define callbacks or mapping dictionaries.
&#64;defaults : dict
Default values for graphviz attributes, labeled with the entity, e.g.
<cite>{‘edge_penwidth’: 0.2}</cite>.
&#64;**kwargs : constant, callable or dict
Graphviz attributes, labeled by the target entity. E.g. <cite>edge_penwidth</cite>,
‘vertex_shape` or <cite>graph_label</cite>.
If the value is constant, this value will be used.
If the value is dict, and has <cite>_name</cite> as key, for every instance of the
given entity, the value of the attribute defined by <cite>_name</cite> will be looked
up in the dict, and the corresponding value will be given to the graphviz
attribute. If the key <cite>_name</cite> is missing from the dict, igraph vertex and
edge indices will be looked up among the keys.
If the value is callable, it will be called with the current instance of
the entity and the returned value will be used for the graphviz attribute.
E.g. <cite>edge_arrowhead(edge)</cite> or <cite>vertex_fillcolor(vertex)</cite>
.. rubric:: Example</p>
<p>import pypath
from pypath import data_formats
net = pypath.PyPath()
net.init_network(pfile = ‘cache/default.pickle’)
#net.init_network({‘arn’: data_formats.omnipath[‘arn’]})
tgf = [v.index for v in net.graph.vs if ‘TGF’ in v[‘slk_pathways’]]
dot = net.export_dot(nodes = tgf, save_graphics = ‘tgf_slk.pdf’, prog = ‘dot’,</p>
<blockquote>
<div><p>main_title = ‘TGF-beta pathway’, return_object = True,
label_font = ‘HelveticaNeueLTStd Med Cn’,
edge_sources = [‘SignaLink3’],
dir_sources = [‘SignaLink3’], hide = True)</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.export_edgelist">
<span class="sig-name descname"><span class="pre">export_edgelist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['name']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_attributes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\t'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.export_edgelist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.export_edgelist" title="Permalink to this definition">§</a></dt>
<dd><p>Write edge list to text file with attributes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> – the name of the file or a stream to read from.</p></li>
<li><p><strong>graph</strong> – the igraph object containing the network</p></li>
<li><p><strong>names</strong> – list with the vertex attribute names to be printed
for source and target vertices</p></li>
<li><p><strong>edge_attributes</strong> – list with the edge attribute names
to be printed</p></li>
<li><p><strong>sep</strong> – string used to separate columns</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.export_graphml">
<span class="sig-name descname"><span class="pre">export_graphml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'main'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.export_graphml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.export_graphml" title="Permalink to this definition">§</a></dt>
<dd><p>Saves the network in a <code class="docutils literal notranslate"><span class="pre">.graphml</span></code> file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Name/path of the output file.
If none is passed,
<code class="docutils literal notranslate"><span class="pre">'results/netrowk-&lt;session_id&gt;.graphml'</span></code> is used.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network object to be
saved. If none is passed, takes the undirected network of
the current instance.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'main'</span></code> by default. The graph name for the
output file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.export_ptms_tab">
<span class="sig-name descname"><span class="pre">export_ptms_tab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.export_ptms_tab"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.export_ptms_tab" title="Permalink to this definition">§</a></dt>
<dd><p>Exports a tab file containing the PTM interaction information
loaded in the network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The output file nama/path to
store the PTM information. If none is provided, the default
is <code class="docutils literal notranslate"><span class="pre">'results/network-&lt;session_id&gt;.tab'</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – Contains the edge indices [int] of all PTM
interactions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.export_sif">
<span class="sig-name descname"><span class="pre">export_sif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.export_sif"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.export_sif" title="Permalink to this definition">§</a></dt>
<dd><p>Exports the network interactions in <code class="docutils literal notranslate"><span class="pre">.sif</span></code> format (Simple
Interaction Format).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Name/path of the output file.
If none is passed, <code class="docutils literal notranslate"><span class="pre">'results/netrowk-&lt;session_id&gt;.sif'</span></code>
is used.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.export_struct_tab">
<span class="sig-name descname"><span class="pre">export_struct_tab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.export_struct_tab"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.export_struct_tab" title="Permalink to this definition">§</a></dt>
<dd><p>Exports a tab file containing the domain interaction information
and PTM regulation loaded in the network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The output file nama/path to
store the PTM information. If none is provided, the default
is <code class="docutils literal notranslate"><span class="pre">'results/network-&lt;session_id&gt;.tab'</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – Contains the edge indices [int] of all PTM
interactions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.export_tab">
<span class="sig-name descname"><span class="pre">export_tab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_node_attrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_edge_attrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique_pairs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.export_tab"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.export_tab" title="Permalink to this definition">§</a></dt>
<dd><p>Exports the network in a tabular format. By default UniProt IDs,
Gene Symbols, source databases, literature references,
directionality and sign information and interaction type are
included.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Name/path of the output file.
If none is passed, <code class="docutils literal notranslate"><span class="pre">'results/netrowk-&lt;session_id&gt;.tab'</span></code>
is used.</p></li>
<li><p><strong>extra_node_attrs</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">{}</span></code> by default. Additional node attributes to
be included in the exported table. Keys are column names
used in the header while values are names of vertex
attributes. In the header <code class="docutils literal notranslate"><span class="pre">'_A'</span></code> and <code class="docutils literal notranslate"><span class="pre">'_B'</span></code> suffixes
will be appended to the column names so the values can be
assigned to A and B side interaction partners.</p></li>
<li><p><strong>extra_edge_attrs</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">{}</span></code> by default. Additional edge attributes to
be included in the exported table. Keys are column names
used in the header while values are names of edge
attributes.</p></li>
<li><p><strong>unique_pairs</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. If set to <code class="docutils literal notranslate"><span class="pre">True</span></code> each line
corresponds to a unique pair of molecules, all
directionality and sign information are covered in other
columns. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, order of <code class="docutils literal notranslate"><span class="pre">'A'</span></code> and <code class="docutils literal notranslate"><span class="pre">'B'</span></code> IDs
corresponds to the direction while sign covered in further
columns.</p></li>
<li><p><strong>kwargs</strong> (<em>**</em>) – Additional keyword arguments passed to
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.export.Export</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.find_all_paths">
<span class="sig-name descname"><span class="pre">find_all_paths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'OUT'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">silent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_adjlist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.find_all_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.find_all_paths" title="Permalink to this definition">§</a></dt>
<dd><p>Finds all paths up to length <cite>maxlen</cite> between groups of
vertices. This function is needed only becaues igraph`s
get_all_shortest_paths() finds only the shortest, not any
path up to a defined length.</p>
<dl class="simple">
<dt>start<span class="classifier">int or list</span></dt><dd><p>Indices of the starting node(s) of the paths.</p>
</dd>
<dt>end<span class="classifier">int or list</span></dt><dd><p>Indices of the target node(s) of the paths.</p>
</dd>
<dt>attr<span class="classifier">str</span></dt><dd><p>Name of the vertex attribute to identify the vertices by.
Necessary if <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> are not igraph vertex ids
but for example vertex names or labels.</p>
</dd>
<dt>mode<span class="classifier">‘IN’, ‘OUT’, ‘ALL’</span></dt><dd><p>Passed to igraph.Graph.neighbors()</p>
</dd>
<dt>maxlen<span class="classifier">int</span></dt><dd><p>Maximum length of paths in steps, i.e. if maxlen = 3, then
the longest path may consist of 3 edges and 4 nodes.</p>
</dd>
<dt>graph<span class="classifier">igraph.Graph object</span></dt><dd><p>The graph you want to find paths in. self.graph by default.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.find_all_paths2">
<span class="sig-name descname"><span class="pre">find_all_paths2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'OUT'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_adjlist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.find_all_paths2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.find_all_paths2" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.find_complex">
<span class="sig-name descname"><span class="pre">find_complex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">search</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.find_complex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.find_complex" title="Permalink to this definition">§</a></dt>
<dd><p>Finds complexes by their non standard names.
E.g. to find DNA polymerases you can use the search
term <cite>DNA pol</cite> which will be tested against complex names
in CORUM.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.first_neighbours">
<span class="sig-name descname"><span class="pre">first_neighbours</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indices</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.first_neighbours"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.first_neighbours" title="Permalink to this definition">§</a></dt>
<dd><p>Looks for the first neighbours of a given node and returns a
list of their UniProt IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>node</strong> (<em>str</em>) – The UniProt ID of the node of interest. Can also be the
index of such node [int].</p></li>
<li><p><strong>indices</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to return the
neighbour nodes indices or their UniProt IDs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – The list containing the first neighbours of the
queried node.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.fisher_enrichment">
<span class="sig-name descname"><span class="pre">fisher_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'proteome'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.fisher_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.fisher_enrichment" title="Permalink to this definition">§</a></dt>
<dd><p>Computes an enrichment analysis using Fisher’s exact test. The
contingency table is built as follows:
First row contains the number of nodes in the <em>ref</em> list (such
list is considered to be loaded in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.lists</span></code>) and the number of nodes in
the current (undirected) network. Second row contains the number
of nodes in <em>lst</em> list (also considered to be already loaded)
and the number of nodes in the network with a non-empty
attribute <em>attr</em>. Uses <code class="xref py py-func docutils literal notranslate"><span class="pre">scipy.stats.fisher_exact()</span></code>, see
the documentation of the corresponding package for more
information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst</strong> (<em>str</em>) – Name of the list in <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.lists</span></code>
whose number of elements will be the first element in the
second row of the contingency table.</p></li>
<li><p><strong>attr</strong> (<em>str</em>) – The node attribute name for which the number of nodes in the
network with such attribute will be the second element of
the second row of the contingency table.</p></li>
<li><p><strong>ref</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'proteome'</span></code> by default. The name of the list in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.lists</span></code> whose number of elements
will be the first element of the first row of the
contingency table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>(<em>float</em>) – Prior odds ratio.</p></li>
<li><p>(<em>float</em>) – P-value or probability of obtaining a
distribution as extreme as the observed, assuming that the
null hypothesis is true.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.geneset_enrichment">
<span class="sig-name descname"><span class="pre">geneset_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">proteins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_proteins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geneset_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hommel'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.geneset_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.geneset_enrichment" title="Permalink to this definition">§</a></dt>
<dd><p>Does not work at the moment because cfisher module should be
replaced with scipy.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.genesymbol">
<span class="sig-name descname"><span class="pre">genesymbol</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.genesymbol"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.genesymbol" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default undirected network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;genesymbol<span class="classifier">str</span></dt><dd><p>GeneSymbol.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.genesymbol_labels">
<span class="sig-name descname"><span class="pre">genesymbol_labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remap_all</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.genesymbol_labels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.genesymbol_labels" title="Permalink to this definition">§</a></dt>
<dd><p>Creats vertex attribute <code class="docutils literal notranslate"><span class="pre">'label'</span></code> and fills up with the
corresponding GeneSymbols of all proteins where the GeneSymbol
can be looked up based on the default name of the protein
vertex (UniProt ID by default). If the attribute <code class="docutils literal notranslate"><span class="pre">'label'</span></code> has
been already initialized, updates this attribute or recreates if
<em>remap_all</em> is set to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network graph object
where the GeneSymbol labels are to be set/updated. If none
is passed, takes the current network undirected graph by
default (<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.graph</span></code>).</p></li>
<li><p><strong>remap_all</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to map anew the
GeneSymbol labels if those were already initialized.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.genesymbols">
<span class="sig-name descname"><span class="pre">genesymbols</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbols</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.genesymbols"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.genesymbols" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_attrs">
<span class="sig-name descname"><span class="pre">get_attrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">line</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lnum</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_attrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_attrs" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_directed">
<span class="sig-name descname"><span class="pre">get_directed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conv_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ret</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_directed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_directed" title="Permalink to this definition">§</a></dt>
<dd><p>Converts a copy of <em>graph</em> undirected <em>igraph.Graph</em> object to a
directed one. By default it converts the current network
instance in <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.graph</span></code> and places the
copy of the directed instance in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.dgraph</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Undirected graph object. If
none is passed, takes the current undirected network
instance and saves the directed network under the attribute
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.dgraph</span></code>. Otherwise, the
directed graph will be returned instead.</p></li>
<li><p><strong>conv_edges</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to convert
undirected edges (those without explicit direction
information) to an arbitrary direction edge or
a pair of opposite edges. Otherwise those will be deleted.</p></li>
<li><p><strong>mutual</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. If <em>conv_edges</em> is <code class="docutils literal notranslate"><span class="pre">True</span></code>,
whether to convert the undirected edges to a single,
arbitrary directed edge, or a pair of opposite directed
edges.</p></li>
<li><p><strong>ret</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to return the
directed graph instance, or not. If a <em>graph</em> is provided,
its directed version will be returned anyway.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>igraph.Graph</em>) – If <em>graph</em> is passed or <em>ret</em> is
<code class="docutils literal notranslate"><span class="pre">True</span></code>, returns the copy of the directed graph. otherwise
returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_dirs_signs">
<span class="sig-name descname"><span class="pre">get_dirs_signs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_dirs_signs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_dirs_signs" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_edge">
<span class="sig-name descname"><span class="pre">get_edge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_edge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_edge" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<em>int</em><em>,</em><em>str</em>) – Vertex index or UniProt ID or GeneSymbol or <code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code>
object.</p></li>
<li><p><strong>target</strong> (<em>int</em><em>,</em><em>str</em>) – Vertex index or UniProt ID or GeneSymbol or <code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code>
object.</p></li>
<li><p><strong>directed</strong> (<em>bool</em>) – To be passed to igraph.Graph.get_eid()</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_edges">
<span class="sig-name descname"><span class="pre">get_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sources</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">targets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a generator with all edges between source and target vertices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sources</strong> (<em>iterable</em>) – Source vertex IDs, names, labels, or any iterable yielding
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> objects.</p></li>
<li><p><strong>targets</strong> (<em>iterable</em>) – Target vertec IDs, names, labels, or any iterable yielding
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> objects.</p></li>
<li><p><strong>directed</strong> (<em>bool</em>) – Passed to <code class="docutils literal notranslate"><span class="pre">igraph.get_eid()</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_function">
<span class="sig-name descname"><span class="pre">get_function</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fun</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_function"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_function" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_giant">
<span class="sig-name descname"><span class="pre">get_giant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_giant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_giant" title="Permalink to this definition">§</a></dt>
<dd><p>Returns the giant component of the <em>graph</em>, or replaces the
<code class="xref py py-class docutils literal notranslate"><span class="pre">igraph.Graph</span></code> instance with only the giant component
if specified.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>replace</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Specifies whether to replace
the <code class="xref py py-class docutils literal notranslate"><span class="pre">igraph.Graph</span></code> instance. This can be either
the undirected network of the current
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.main.PyPath</span></code> instance (default) or the one
passed under the keyword argument <em>graph</em>.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The graph object from which
the giant component is to be computed. If none is specified,
takes the undirected network of the current
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.main.PyPath</span></code> instance.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>igraph.Graph</em>) – If <code class="docutils literal notranslate"><span class="pre">replace=False</span></code>, returns a copy of
the giant component graph.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_go">
<span class="sig-name descname"><span class="pre">get_go</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_go"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_go" title="Permalink to this definition">§</a></dt>
<dd><p>Returns the <code class="docutils literal notranslate"><span class="pre">GOAnnotation</span></code> object for the organism requested
(or the default one).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_max">
<span class="sig-name descname"><span class="pre">get_max</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attrList</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_max"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_max" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_network">
<span class="sig-name descname"><span class="pre">get_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">andor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'or'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_network" title="Permalink to this definition">§</a></dt>
<dd><p>Retrieves a subnetwork according to a set of user-defined
attributes. Basically applies
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.get_sub()</span></code> on a given <em>graph</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>crit</strong> (<em>dict</em>) – Defines the critical attributes to generate the subnetwork.
Keys are <code class="docutils literal notranslate"><span class="pre">'edge'</span></code> and <code class="docutils literal notranslate"><span class="pre">'node'</span></code> and values are [dict]
containing the critical attribute names [str] and values
are [set] containing those attributes of the nodes/edges
that are to be kept.</p></li>
<li><p><strong>andor</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'or'</span></code> by default. Determines the search mode.
See <code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.search_attr_or()</span></code> and
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.search_attr_and()</span></code> for more
details.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The graph object where to
extract the subnetwork. If none is passed, takes the current
network (undirected) graph
(<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.graph</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>igraph.Graph</em>) – The subgraph obtained from filtering
according to the attributes defined in <em>crit</em>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_node">
<span class="sig-name descname"><span class="pre">get_node</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_node"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_node" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;identifier<span class="classifier">int, str</span></dt><dd><p>Vertex index (int) or GeneSymbol (str) or UniProt ID (str) or
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_node_d">
<span class="sig-name descname"><span class="pre">get_node_d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_node_d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_node_d" title="Permalink to this definition">§</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.
Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;identifier<span class="classifier">int, str</span></dt><dd><p>Vertex index (int) or GeneSymbol (str) or UniProt ID (str) or
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_node_pair">
<span class="sig-name descname"><span class="pre">get_node_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_node_pair"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_node_pair" title="Permalink to this definition">§</a></dt>
<dd><p>Retrieves the node IDs from a pair of node names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_a</strong> (<em>str</em>) – Name of the source node.</p></li>
<li><p><strong>id_b</strong> (<em>str</em>) – Name of the target node.</p></li>
<li><p><strong>directed</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to return the node
indices from the directed or undirected graph.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>tuple</em>) – The pair of node IDs of the selected graph.
If not found, returns <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_nodes">
<span class="sig-name descname"><span class="pre">get_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifiers</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_nodes" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_nodes_d">
<span class="sig-name descname"><span class="pre">get_nodes_d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifiers</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_nodes_d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_nodes_d" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_pathways">
<span class="sig-name descname"><span class="pre">get_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_pathways" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_proteomicsdb">
<span class="sig-name descname"><span class="pre">get_proteomicsdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">user</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">passwd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tissues</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pickle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_proteomicsdb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_proteomicsdb" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_sub">
<span class="sig-name descname"><span class="pre">get_sub</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">crit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">andor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'or'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_sub"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_sub" title="Permalink to this definition">§</a></dt>
<dd><p>Selects the nodes from <em>graph</em> (and edges to be removed)
according to a set of user-defined attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>crit</strong> (<em>dict</em>) – Defines the critical attributes to generate the subnetwork.
Keys are <code class="docutils literal notranslate"><span class="pre">'edge'</span></code> and <code class="docutils literal notranslate"><span class="pre">'node'</span></code> and values are [dict]
containing the critical attribute names [str] and values
are [set] containing those attributes of the nodes/edges
that are to be kept.</p></li>
<li><p><strong>andor</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'or'</span></code> by default. Determines the search mode.
See <code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.search_attr_or()</span></code> and
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.search_attr_and()</span></code> for more
details.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The graph object where to
extract the subnetwork. If none is passed, takes the current
network (undirected) graph
(<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.graph</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>dict</em>) – Keys are <code class="docutils literal notranslate"><span class="pre">'nodes'</span></code> and <code class="docutils literal notranslate"><span class="pre">'edges'</span></code> whose
values are [lst] of elements (as indexes [int]). Nodes are
those to be kept and edges to be removed on the extracted
subnetwork.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.get_taxon">
<span class="sig-name descname"><span class="pre">get_taxon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tax_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fields</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.get_taxon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.get_taxon" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.go_annotate_graph">
<span class="sig-name descname"><span class="pre">go_annotate_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aspects</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('C',</span> <span class="pre">'F',</span> <span class="pre">'P')</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.go_annotate_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.go_annotate_graph" title="Permalink to this definition">§</a></dt>
<dd><p>Annotates protein nodes with GO terms. In the <code class="docutils literal notranslate"><span class="pre">go</span></code> vertex
attribute each node is annotated by a dict of sets where keys are
one letter codes of GO aspects and values are sets of GO accessions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.go_enrichment">
<span class="sig-name descname"><span class="pre">go_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">proteins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aspect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'P'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hommel'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_proteins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.go_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.go_enrichment" title="Permalink to this definition">§</a></dt>
<dd><p>Does not work at the moment because cfisher module should be
replaced with scipy.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs">
<span class="sig-name descname"><span class="pre">gs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.gs" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default undirected network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;genesymbol<span class="classifier">str</span></dt><dd><p>GeneSymbol.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_affected_by">
<span class="sig-name descname"><span class="pre">gs_affected_by</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_affected_by"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_affected_by" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_affects">
<span class="sig-name descname"><span class="pre">gs_affects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_affects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_affects" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_edge">
<span class="sig-name descname"><span class="pre">gs_edge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_edge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_edge" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;source<span class="classifier">str</span></dt><dd><p>GeneSymbol</p>
</dd>
<dt>&#64;target<span class="classifier">str</span></dt><dd><p>GeneSymbol</p>
</dd>
<dt>&#64;directed<span class="classifier">bool</span></dt><dd><p>To be passed to igraph.Graph.get_eid()</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_in_directed">
<span class="sig-name descname"><span class="pre">gs_in_directed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_in_directed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_in_directed" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_in_undirected">
<span class="sig-name descname"><span class="pre">gs_in_undirected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_in_undirected"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_in_undirected" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_inhibited_by">
<span class="sig-name descname"><span class="pre">gs_inhibited_by</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_inhibited_by"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_inhibited_by" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_inhibits">
<span class="sig-name descname"><span class="pre">gs_inhibits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_inhibits"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_inhibits" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_neighborhood">
<span class="sig-name descname"><span class="pre">gs_neighborhood</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ALL'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_neighborhood"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_neighborhood" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_neighbors">
<span class="sig-name descname"><span class="pre">gs_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ALL'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_neighbors" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_stimulated_by">
<span class="sig-name descname"><span class="pre">gs_stimulated_by</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_stimulated_by"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_stimulated_by" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gs_stimulates">
<span class="sig-name descname"><span class="pre">gs_stimulates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.gs_stimulates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.gs_stimulates" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.gss">
<span class="sig-name descname"><span class="pre">gss</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genesymbols</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.gss" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.guide2pharma">
<span class="sig-name descname"><span class="pre">guide2pharma</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.guide2pharma"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.guide2pharma" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.having_attr">
<span class="sig-name descname"><span class="pre">having_attr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.having_attr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.having_attr" title="Permalink to this definition">§</a></dt>
<dd><p>Checks if edges or nodes of the network have a specific
attribute and returns an iterator of the indices (or the
edge/node instances) of edges/nodes having such attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attr</strong> (<em>str</em>) – The name of the attribute to look for.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The graph object where the
edge/node attribute is to be searched. If none is passed,
takes the undirected network of the current instance.</p></li>
<li><p><strong>index</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Whether to return the
iterator of the indices or the node/edge instances.</p></li>
<li><p><strong>edges</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Whether to look for the
attribute in the networks edges or nodes instead.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>generator</em>) – Generator object containing the edge/node
indices (or instances) having the specified attribute.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.having_eattr">
<span class="sig-name descname"><span class="pre">having_eattr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.having_eattr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.having_eattr" title="Permalink to this definition">§</a></dt>
<dd><p>Checks if edges of the network have a specific attribute and
returns an iterator of the indices (or the edge instances) of
edges having such attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attr</strong> (<em>str</em>) – The name of the attribute to look for.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The graph object where the
edge/node attribute is to be searched. If none is passed,
takes the undirected network of the current instance.</p></li>
<li><p><strong>index</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Whether to return the
iterator of the indices or the node/edge instances.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>generator</em>) – Generator object containing the edge
indices (or instances) having the specified attribute.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.having_ptm">
<span class="sig-name descname"><span class="pre">having_ptm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.having_ptm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.having_ptm" title="Permalink to this definition">§</a></dt>
<dd><p>Checks if edges of the network have the <code class="docutils literal notranslate"><span class="pre">'ptm'</span></code> attribute and
returns an iterator of the indices (or the edge instances) of
edges having such attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Whether to return the
iterator of the indices or the node/edge instances.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The graph object where the
edge/node attribute is to be searched. If none is passed,
takes the undirected network of the current instance.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>generator</em>) – Generator object containing the edge
indices (or instances) having the <code class="docutils literal notranslate"><span class="pre">ptm''</span></code> attribute.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.having_vattr">
<span class="sig-name descname"><span class="pre">having_vattr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.having_vattr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.having_vattr" title="Permalink to this definition">§</a></dt>
<dd><p>Checks if nodes of the network have a specific attribute and
returns an iterator of the indices (or the node instances) of
nodes having such attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attr</strong> (<em>str</em>) – The name of the attribute to look for.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The graph object where the
edge/node attribute is to be searched. If none is passed,
takes the undirected network of the current instance.</p></li>
<li><p><strong>index</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Whether to return the
iterator of the indices or the node/edge instances.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>generator</em>) – Generator object containing the node
indices (or instances) having the specified attribute.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.homology_translation">
<span class="sig-name descname"><span class="pre">homology_translation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_swissprot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.homology_translation" title="Permalink to this definition">§</a></dt>
<dd><p>Translates the current object to another organism by orthology.
Proteins without known ortholog will be deleted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>target</strong> (<em>int</em>) – NCBI Taxonomy ID of the target organism. E.g. 10090 for mouse.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.htp_stats">
<span class="sig-name descname"><span class="pre">htp_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.htp_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.htp_stats" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.in_complex">
<span class="sig-name descname"><span class="pre">in_complex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['corum']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.in_complex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.in_complex" title="Permalink to this definition">§</a></dt>
<dd><p>Deprecated, will be removed.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.in_directed">
<span class="sig-name descname"><span class="pre">in_directed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vertex</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.in_directed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.in_directed" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.in_undirected">
<span class="sig-name descname"><span class="pre">in_undirected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vertex</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.in_undirected"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.in_undirected" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.info">
<span class="sig-name descname"><span class="pre">info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.info" title="Permalink to this definition">§</a></dt>
<dd><p>Given the name of a resource, prints out the information about
that source/database. You can check the list of available
resource descriptions in
<code class="xref py py-func docutils literal notranslate"><span class="pre">ypath.descriptions.descriptions.keys()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – The name of the resource from which to print the
information.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.init_complex_attr">
<span class="sig-name descname"><span class="pre">init_complex_attr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.init_complex_attr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.init_complex_attr" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.init_edge_attr">
<span class="sig-name descname"><span class="pre">init_edge_attr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.init_edge_attr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.init_edge_attr" title="Permalink to this definition">§</a></dt>
<dd><p>Fills all edges attribute <em>attr</em> with its default type (if
such attribute value is <code class="docutils literal notranslate"><span class="pre">None</span></code>), creates [list] if in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.edgeAttrs</span></code> such attribute is
registered as [list].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>attr</strong> (<em>str</em>) – The attribute name to be initialized on the network edges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.init_gsea">
<span class="sig-name descname"><span class="pre">init_gsea</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">user</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.init_gsea"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.init_gsea" title="Permalink to this definition">§</a></dt>
<dd><p>Initializes a <code class="docutils literal notranslate"><span class="pre">pypath.gsea.GSEA</span></code> object and shows the list of the
collections in MSigDB.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.init_network">
<span class="sig-name descname"><span class="pre">init_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pickle_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reread</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">redownload</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.init_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.init_network" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the network data.</p>
<p>This is a lazy way to start the module, load data and build the
high confidence, literature curated part of the signaling
network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies the data input
formats for the different resources (keys) [str]. Values
are <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.input_formats.NetworkInput</span></code> instances
containing the information. By default uses the set of
resources of OmniPath.</p></li>
<li><p><strong>exclude</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">[]</span></code> by default. List of resources [str] to
exclude from the network.</p></li>
<li><p><strong>cache_files</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">{}</span></code> by default. Contains the resource name(s)
[str] (keys) and the corresponding cached file name [str].
If provided (and file exists) bypasses the download of the
data for that resource and uses the cache file instead.</p></li>
<li><p><strong>pfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. If any, provides the file
name or path to a previously saved network pickle file.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> is passed, takes the default path from
<a class="reference internal" href="#pypath.legacy.main.PyPath.save_network" title="pypath.legacy.main.PyPath.save_network"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PyPath.save_network()</span></code></a>
(<code class="docutils literal notranslate"><span class="pre">'cache/default_network.pickle'</span></code>).</p></li>
<li><p><strong>save</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. If set to <code class="docutils literal notranslate"><span class="pre">True</span></code>, saves
the loaded network to its default location
(<code class="docutils literal notranslate"><span class="pre">'cache/default_network.pickle'</span></code>).</p></li>
<li><p><strong>reread</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Specifies whether to reread
the data files from the cache or omit them (similar to
<em>redownload</em>).</p></li>
<li><p><strong>redownload</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Specifies whether to
re-download the data and ignore the cache.</p></li>
<li><p><strong>**kwargs</strong> – Not used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.init_vertex_attr">
<span class="sig-name descname"><span class="pre">init_vertex_attr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.init_vertex_attr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.init_vertex_attr" title="Permalink to this definition">§</a></dt>
<dd><p>Fills all vertices attribute <em>attr</em> with its default type (if
such attribute value is <code class="docutils literal notranslate"><span class="pre">None</span></code>), creates [list] if in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.vertexAttrs</span></code> such attribute is
registered as [list].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>attr</strong> (<em>str</em>) – The attribute name to be initialized on the network
vertices.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_all">
<span class="sig-name descname"><span class="pre">interactions_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_all"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_all" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>set</em> of tuples of node name pairs representing
interactions, both directed and undirected. Directed interactions
will be present according to their direction, mutual directed
interactions are represented by two tuples. The directed and
undirected interactions are not distinguished here.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_directed">
<span class="sig-name descname"><span class="pre">interactions_directed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_directed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_directed" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>set</em> of tuples of node name pairs with being aware
of the directions.
Undirected interactions will be discarded.
Pairs of node names represent the directions: first is the source,
second is the target.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_directed_by_resource">
<span class="sig-name descname"><span class="pre">interactions_directed_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">effect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_directed_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_directed_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>dict</em> of <a href="#id25"><span class="problematic" id="id26">*</span></a>set*s of tuples of node name pairs with being
aware of the directions.
Undirected interactions will be discarded.
Pairs of node names represent the directions: first is the source,
second is the target.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_inhibitory">
<span class="sig-name descname"><span class="pre">interactions_inhibitory</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_inhibitory"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_inhibitory" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>set</em> of tuples of node name pairs only for inhibitory
interactions.
Pairs of node names represent the directions: first is the source,
second is the target.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_inhibitory_by_resource">
<span class="sig-name descname"><span class="pre">interactions_inhibitory_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_inhibitory_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_inhibitory_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>dict</em> of <a href="#id27"><span class="problematic" id="id28">*</span></a>set*s of tuples of node name pairs with being
aware of the directions.
Undirected interactions will be discarded.
Pairs of node names represent the directions: first is the source,
second is the target.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_mutual">
<span class="sig-name descname"><span class="pre">interactions_mutual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_mutual"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_mutual" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>set</em> of tuples of node name pairs representing
mutual interactions (i.e. A–&gt;B and B–&gt;A).
Pairs of node names will be sorted alphabetically.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_mutual_by_resource">
<span class="sig-name descname"><span class="pre">interactions_mutual_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_mutual_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_mutual_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>dict</em> of <a href="#id29"><span class="problematic" id="id30">*</span></a>set*s of tuples of node name pairs representing
mutual interactions (i.e. A–&gt;B and B–&gt;A).
Pairs of node names will be sorted alphabetically.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_signed">
<span class="sig-name descname"><span class="pre">interactions_signed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_signed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_signed" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>set</em> of tuples of node name pairs only for signed
interactions.
Pairs of node names represent the directions: first is the source,
second is the target.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_signed_by_resource">
<span class="sig-name descname"><span class="pre">interactions_signed_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_signed_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_signed_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>dict</em> of <a href="#id31"><span class="problematic" id="id32">*</span></a>set*s of tuples of node name pairs with being
aware of the directions.
Undirected interactions will be discarded.
Pairs of node names represent the directions: first is the source,
second is the target.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_stimulatory">
<span class="sig-name descname"><span class="pre">interactions_stimulatory</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_stimulatory"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_stimulatory" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>set</em> of tuples of node name pairs only for stimulatory
interactions.
Pairs of node names represent the directions: first is the source,
second is the target.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_stimulatory_by_resource">
<span class="sig-name descname"><span class="pre">interactions_stimulatory_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_stimulatory_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_stimulatory_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>dict</em> of <a href="#id33"><span class="problematic" id="id34">*</span></a>set*s of tuples of node name pairs with being
aware of the directions.
Undirected interactions will be discarded.
Pairs of node names represent the directions: first is the source,
second is the target.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_undirected">
<span class="sig-name descname"><span class="pre">interactions_undirected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_undirected"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_undirected" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>set</em> of tuples of node name pairs without being aware
of the directions.
Pairs of node names will be sorted alphabetically.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.interactions_undirected_by_resource">
<span class="sig-name descname"><span class="pre">interactions_undirected_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.interactions_undirected_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.interactions_undirected_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a <em>dict</em> of <a href="#id35"><span class="problematic" id="id36">*</span></a>set*s of tuples of node name pairs without
being aware of the directions.
Pairs of node names will be sorted alphabetically.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.intergroup_shortest_paths">
<span class="sig-name descname"><span class="pre">intergroup_shortest_paths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">groupA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.intergroup_shortest_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.intergroup_shortest_paths" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.intogen_cancer_drivers_list">
<span class="sig-name descname"><span class="pre">intogen_cancer_drivers_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intogen_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.intogen_cancer_drivers_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.intogen_cancer_drivers_list" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the list of cancer driver proteins from IntOGen data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>intogen_file</strong> (<em>str</em>) – Path to the data file. Can also be [function] that provides
the data. In general, anything accepted by
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.input_formats.NetworkInput.input</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.iter_edges">
<span class="sig-name descname"><span class="pre">iter_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.iter_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.iter_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Iterates the edges in the graph optionally limited to certain
resources. Yields <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.iter_interactions">
<span class="sig-name descname"><span class="pre">iter_interactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">signs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_undirected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_references</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.iter_interactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.iter_interactions" title="Permalink to this definition">§</a></dt>
<dd><p>Iterates over edges and yields interaction records.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>signs</strong> (<em>bool</em>) – Ignoring signs if <code class="docutils literal notranslate"><span class="pre">False</span></code>. This way each directed interaction
will yield a single record even if it’s ambiguously labeled
both positive and negative. The default behaviour is to yield
two records in this case, one with positive and one with negative
sign.</p></li>
<li><p><strong>all_undirected</strong> (<em>bool</em>) – Yield records for undirected interactions even if certain sources
provide direction. If <code class="docutils literal notranslate"><span class="pre">False</span></code> only the directed records will
be provided and the undirected sources and references will be
added to the directed records.</p></li>
<li><p><strong>by_source</strong> (<em>bool</em>) – Yield separate records by resources. This way the node pairs
will be redundant and you need to group later if you want
unique interacting pairs. By default is <code class="docutils literal notranslate"><span class="pre">False</span></code> because for
most applications unique interactions are preferred.
If <code class="docutils literal notranslate"><span class="pre">False</span></code> the <em>refrences</em> field will still be present
but with <code class="docutils literal notranslate"><span class="pre">None</span></code> values.</p></li>
<li><p><strong>with_references</strong> (<em>bool</em>) – Include the literature references. By default is <code class="docutils literal notranslate"><span class="pre">False</span></code>
because you rarely need these and they increase the data size
significantly.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.jaccard_edges">
<span class="sig-name descname"><span class="pre">jaccard_edges</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.jaccard_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.jaccard_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the Jaccard similarity index between the sets of first
neighbours of all node pairs. <strong>NOTE:</strong> this method can take a
while to compute, e.g.: if the network has 10K nodes, the total
number of possible pairs to compute is:</p>
<div class="math notranslate nohighlight">
\[\binom{10^4}{2} = 49995000\]</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>list</em>) – Large list of [tuple] elements containing the
node pair names [str] and their corresponding first
neighbours Jaccard index [float].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.jaccard_meta">
<span class="sig-name descname"><span class="pre">jaccard_meta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jedges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">critical</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.jaccard_meta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.jaccard_meta" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a (undirected) graph from a list of edges filtering by
their Jaccard index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>jedges</strong> (<em>list</em>) – List of [tuple] containing the edges node names [str] and
their Jaccard index. Basically, the output of
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.jaccard_edges()</span></code>.</p></li>
<li><p><strong>critical</strong> (<em>float</em>) – Specifies the threshold of the Jaccard index from above
which an edge will be included in the graph.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>igraph.Graph</em>) – The Undirected graph instance containing
only the edges whose Jaccard similarity index is above the
threshold specified by <em>critical</em>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.kegg_directions">
<span class="sig-name descname"><span class="pre">kegg_directions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.kegg_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.kegg_directions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.kegg_pathways">
<span class="sig-name descname"><span class="pre">kegg_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.kegg_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.kegg_pathways" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.kinase_stats">
<span class="sig-name descname"><span class="pre">kinase_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.kinase_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.kinase_stats" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.label">
<span class="sig-name descname"><span class="pre">label</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">what</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'vertices'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.label"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.label" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a boolean attribute <code class="docutils literal notranslate"><span class="pre">label</span></code> True for the
vertex or edge IDs in the set <code class="docutils literal notranslate"><span class="pre">idx</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.label_by_go">
<span class="sig-name descname"><span class="pre">label_by_go</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_terms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ANY'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.label_by_go"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.label_by_go" title="Permalink to this definition">§</a></dt>
<dd><p>Assigns a boolean vertex attribute to nodes which tells whether
the node is annotated by all or any of the GO terms.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.label_edges">
<span class="sig-name descname"><span class="pre">label_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.label_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.label_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a boolean edge attribute <code class="docutils literal notranslate"><span class="pre">label</span></code> True for the
edge IDs in the set <code class="docutils literal notranslate"><span class="pre">edges</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.label_vertices">
<span class="sig-name descname"><span class="pre">label_vertices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertices</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.label_vertices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.label_vertices" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a boolean vertex attribute <code class="docutils literal notranslate"><span class="pre">label</span></code> True for the
vertex IDs in the set <code class="docutils literal notranslate"><span class="pre">vertices</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.laudanna_directions">
<span class="sig-name descname"><span class="pre">laudanna_directions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.laudanna_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.laudanna_directions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.laudanna_effects">
<span class="sig-name descname"><span class="pre">laudanna_effects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.laudanna_effects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.laudanna_effects" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.license">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">license</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.license"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.license" title="Permalink to this definition">§</a></dt>
<dd><p>Prints information about data licences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.list_resources">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">list_resources</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.list_resources"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.list_resources" title="Permalink to this definition">§</a></dt>
<dd><p>Prints the list of resources through the standard output.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_3dcomplexes">
<span class="sig-name descname"><span class="pre">load_3dcomplexes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_3dcomplexes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_3dcomplexes" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_3did_ddi">
<span class="sig-name descname"><span class="pre">load_3did_ddi</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_3did_ddi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_3did_ddi" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_3did_ddi2">
<span class="sig-name descname"><span class="pre">load_3did_ddi2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ddi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interfaces</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_3did_ddi2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_3did_ddi2" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_3did_dmi">
<span class="sig-name descname"><span class="pre">load_3did_dmi</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_3did_dmi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_3did_dmi" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_3did_interfaces">
<span class="sig-name descname"><span class="pre">load_3did_interfaces</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_3did_interfaces"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_3did_interfaces" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_all_pathways">
<span class="sig-name descname"><span class="pre">load_all_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_all_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_all_pathways" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_compleat">
<span class="sig-name descname"><span class="pre">load_compleat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_compleat"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_compleat" title="Permalink to this definition">§</a></dt>
<dd><p>Loads complexes from Compleat. Loads data into vertex attribute
<cite>graph.vs[‘complexes’][‘compleat’]</cite>.
This resource is human only.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_complexportal">
<span class="sig-name descname"><span class="pre">load_complexportal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_complexportal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_complexportal" title="Permalink to this definition">§</a></dt>
<dd><p>Loads complexes from ComplexPortal. Loads data into vertex attribute
<cite>graph.vs[‘complexes’][‘complexportal’]</cite>.
This resource is human only.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_comppi">
<span class="sig-name descname"><span class="pre">load_comppi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_comppi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_comppi" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_corum">
<span class="sig-name descname"><span class="pre">load_corum</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_corum"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_corum" title="Permalink to this definition">§</a></dt>
<dd><p>Loads complexes from CORUM database. Loads data into vertex attribute
<cite>graph.vs[‘complexes’][‘corum’]</cite>.
This resource is human only.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_dbptm">
<span class="sig-name descname"><span class="pre">load_dbptm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">non_matching</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_dbptm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_dbptm" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_ddi">
<span class="sig-name descname"><span class="pre">load_ddi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ddi</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_ddi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_ddi" title="Permalink to this definition">§</a></dt>
<dd><p>ddi is either a list of intera.DomainDomain objects,
or a function resulting this list</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_ddis">
<span class="sig-name descname"><span class="pre">load_ddis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">methods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['inputs.threedcomplex.threedcomplex_ddi',</span> <span class="pre">'inputs.domino.domino_ddi',</span> <span class="pre">'self.load_3did_ddi2']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_ddis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_ddis" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_depod_dmi">
<span class="sig-name descname"><span class="pre">load_depod_dmi</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_depod_dmi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_depod_dmi" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_disgenet">
<span class="sig-name descname"><span class="pre">load_disgenet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'curated'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umls</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_disgenet"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_disgenet" title="Permalink to this definition">§</a></dt>
<dd><p>Assigns DisGeNet disease-gene associations to the proteins
in the network. Disease annotations will be added to the <cite>dis</cite>
vertex attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>score</strong> (<em>float</em>) – Confidence score from DisGeNet. Only associations
above the score provided will be considered.</p></li>
<li><p><strong>ulms</strong> (<em>bool</em>) – By default we assign a list of disease names to
each protein. To use Unified Medical Language System IDs instead
set this to <cite>True</cite>.</p></li>
<li><p><strong>full_data</strong> (<em>bool</em>) – By default we load only disease names. Set this
to <cite>True</cite> if you wish to load additional annotations like number
of PubMed IDs, number of SNPs and original sources.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_dmi">
<span class="sig-name descname"><span class="pre">load_dmi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dmi</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_dmi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_dmi" title="Permalink to this definition">§</a></dt>
<dd><p>dmi is either a list of intera.DomainMotif objects,
or a function resulting this list</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_dmis">
<span class="sig-name descname"><span class="pre">load_dmis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">methods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['self.pfam_regions',</span> <span class="pre">'self.load_depod_dmi',</span> <span class="pre">'self.load_dbptm',</span> <span class="pre">'self.load_mimp_dmi',</span> <span class="pre">'self.load_pnetworks_dmi',</span> <span class="pre">'self.load_domino_dmi',</span> <span class="pre">'self.load_pepcyber',</span> <span class="pre">'self.load_psite_reg',</span> <span class="pre">'self.load_psite_phos',</span> <span class="pre">'self.load_ielm',</span> <span class="pre">'self.load_phosphoelm',</span> <span class="pre">'self.load_elm',</span> <span class="pre">'self.load_3did_dmi']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_dmis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_dmis" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_domino_dmi">
<span class="sig-name descname"><span class="pre">load_domino_dmi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_domino_dmi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_domino_dmi" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_dorothea">
<span class="sig-name descname"><span class="pre">load_dorothea</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">levels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'A',</span> <span class="pre">'B'}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_curated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_dorothea"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_dorothea" title="Permalink to this definition">§</a></dt>
<dd><p>Adds TF-target interactions from TF regulons to the network.
DoRothEA is a comprehensive resource of TF-target
interactions combining multiple lines of evidences: literature
curated databases, ChIP-Seq data, PWM based prediction using
HOCOMOCO and JASPAR matrices and prediction from GTEx expression
data by ARACNe.</p>
<p>For details see <a class="github reference external" href="https://github.com/saezlab/DoRothEA">saezlab/DoRothEA</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>levels</strong> (<em>set</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">{'A',</span> <span class="pre">'B'}</span></code> by default. Confidence levels to be
loaded (from A to E) [str].</p></li>
<li><p><strong>only_curated</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to retrieve only the
literature curated interactions or not.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_elm">
<span class="sig-name descname"><span class="pre">load_elm</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_elm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_elm" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_exocarta_attrs">
<span class="sig-name descname"><span class="pre">load_exocarta_attrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">load_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_refs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_exocarta_attrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_exocarta_attrs" title="Permalink to this definition">§</a></dt>
<dd><p>Creates vertex attributes from ExoCarta data. Creates a boolean
attribute <code class="docutils literal notranslate"><span class="pre">exocarts_exosomal</span></code> which tells whether a protein is
in ExoCarta i.e. has been found in exosomes. Optionally creates
attributes <code class="docutils literal notranslate"><span class="pre">exocarta_samples</span></code> and <code class="docutils literal notranslate"><span class="pre">exocarta_refs</span></code> listing the
sample tissue and the PubMed references, respectively.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_expression">
<span class="sig-name descname"><span class="pre">load_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_expression" title="Permalink to this definition">§</a></dt>
<dd><p>Expression data can be loaded into vertex attributes,
or into a pandas DataFrame – the latter offers faster
ways to process and use these huge matrices.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_from_pickle">
<span class="sig-name descname"><span class="pre">load_from_pickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pickle_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_from_pickle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_from_pickle" title="Permalink to this definition">§</a></dt>
<dd><p>Shortcut for loading a network from a pickle dump.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_go">
<span class="sig-name descname"><span class="pre">load_go</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_go"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_go" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a <code class="docutils literal notranslate"><span class="pre">pypath.go.GOAnnotation</span></code> object for one organism in the
dict under <code class="docutils literal notranslate"><span class="pre">go</span></code> attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>organism</strong> (<em>int</em>) – NCBI Taxonomy ID of the organism.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_havugimana">
<span class="sig-name descname"><span class="pre">load_havugimana</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_havugimana"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_havugimana" title="Permalink to this definition">§</a></dt>
<dd><p>Loads complexes from Havugimana 2012. Loads data into vertex attribute
<cite>graph.vs[‘complexes’][‘havugimana’]</cite>.
This resource is human only.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_hpa">
<span class="sig-name descname"><span class="pre">load_hpa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">normal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathology</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cancer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarize_pathology</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tissues</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'Approved',</span> <span class="pre">'Supported'}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">levels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'High':</span> <span class="pre">3,</span> <span class="pre">'Low':</span> <span class="pre">1,</span> <span class="pre">'Medium':</span> <span class="pre">2,</span> <span class="pre">'Not</span> <span class="pre">detected':</span> <span class="pre">0}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">na_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_hpa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_hpa" title="Permalink to this definition">§</a></dt>
<dd><p>Loads Human Protein Atlas data into vertex attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_hprd_ptms">
<span class="sig-name descname"><span class="pre">load_hprd_ptms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">non_matching</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_hprd_ptms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_hprd_ptms" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_ielm">
<span class="sig-name descname"><span class="pre">load_ielm</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_ielm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_ielm" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_interfaces">
<span class="sig-name descname"><span class="pre">load_interfaces</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_interfaces"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_interfaces" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_li2012_ptms">
<span class="sig-name descname"><span class="pre">load_li2012_ptms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">non_matching</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_li2012_ptms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_li2012_ptms" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_ligand_receptor_network">
<span class="sig-name descname"><span class="pre">load_ligand_receptor_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lig_rec_resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inference_from_go</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_undirected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_rec_rec</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_lig_lig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_ligand_receptor_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_ligand_receptor_network" title="Permalink to this definition">§</a></dt>
<dd><p>Initializes a ligand-receptor network.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_lmpid">
<span class="sig-name descname"><span class="pre">load_lmpid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_lmpid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_lmpid" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_matrisome_attrs">
<span class="sig-name descname"><span class="pre">load_matrisome_attrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_matrisome_attrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_matrisome_attrs" title="Permalink to this definition">§</a></dt>
<dd><p>Loads vertex attributes from MatrisomeDB 2.0. Attributes are
<code class="docutils literal notranslate"><span class="pre">matrisome_class</span></code>, <code class="docutils literal notranslate"><span class="pre">matrisome_subclass</span></code> and <code class="docutils literal notranslate"><span class="pre">matrisome_notes</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_membranome_attrs">
<span class="sig-name descname"><span class="pre">load_membranome_attrs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_membranome_attrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_membranome_attrs" title="Permalink to this definition">§</a></dt>
<dd><p>Loads attributes from Membranome, a database of single-helix
transmembrane proteins.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_mimp_dmi">
<span class="sig-name descname"><span class="pre">load_mimp_dmi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">non_matching</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_mimp_dmi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_mimp_dmi" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_mutations">
<span class="sig-name descname"><span class="pre">load_mutations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attributes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gdsc_datadir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_mutations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_mutations" title="Permalink to this definition">§</a></dt>
<dd><p>Mutations are listed in vertex attributes. Mutation() objects
offers methods to identify residues and look up in Ptm(), Motif()
and Domain() objects, to check if those residues are
modified, or are in some short motif or domain.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_negatives">
<span class="sig-name descname"><span class="pre">load_negatives</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_negatives"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_negatives" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_old_omnipath">
<span class="sig-name descname"><span class="pre">load_old_omnipath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kinase_substrate_extra</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_htp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">htp_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_refs_undirected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_old_omnipath"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_old_omnipath" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the OmniPath network as it was before August 2016.
Furthermore it gives some more options.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_omnipath">
<span class="sig-name descname"><span class="pre">load_omnipath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">omnipath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kinase_substrate_extra</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_receptor_extra</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathway_extra</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_htp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">htp_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_refs_undirected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">old_omnipath_resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pickle_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_omnipath"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_omnipath" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the OmniPath network.
Note, if <code class="docutils literal notranslate"><span class="pre">pickle_file</span></code> provided the network will be loaded directly
from there regardless of its content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_pathways">
<span class="sig-name descname"><span class="pre">load_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_pathways" title="Permalink to this definition">§</a></dt>
<dd><p>Generic method to load pathway annotations from a resource.
We don’t recommend calling this method but either specific
methods for a single source e.g. <cite>kegg_pathways()</cite>
or <cite>sirnor_pathways()</cite> or call <cite>load_all_pathways()</cite> to
load all resources.</p>
<dl class="simple">
<dt>:param<span class="classifier">Name of the source, this need to match a method in the dict</span></dt><dd><p>in <cite>get_pathways()</cite> method and the edge and vertex attributes
with pathway annotations will be called “&lt;source&gt;_pathways”.</p>
</dd>
</dl>
<p>:type : param str source:
:param : A graph, by default the default the <cite>graph</cite> attribute of the</p>
<blockquote>
<div><p>current instance.</p>
</div></blockquote>
<p>:type : param igraph.Graph graph:</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_pdb">
<span class="sig-name descname"><span class="pre">load_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_pdb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_pdb" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the 3D structure information from PDB into the network.
Creates the node attribute <code class="docutils literal notranslate"><span class="pre">'pdb'</span></code> containing a [dict] whose
keys are the PDB identifier [str] and values are [tuple] of two
elements denoting the experimental method [str] (e.g.:
<code class="docutils literal notranslate"><span class="pre">'X-ray'</span></code>, <code class="docutils literal notranslate"><span class="pre">'NMR'</span></code>, …) and the resolution [float] (if
applicable).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network object for which
the information is to be loaded. If none is passed, takes
the undirected network of the current instance.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_pepcyber">
<span class="sig-name descname"><span class="pre">load_pepcyber</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_pepcyber"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_pepcyber" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_pfam">
<span class="sig-name descname"><span class="pre">load_pfam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_pfam"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_pfam" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the protein family information from UniProt into the
network. Creates the node attribute <code class="docutils literal notranslate"><span class="pre">'pfam'</span></code> containing a
[list] of protein family identifier(s) [str].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network object for which
the information is to be loaded. If none is passed, takes
the undirected network of the current instance.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_pfam2">
<span class="sig-name descname"><span class="pre">load_pfam2</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_pfam2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_pfam2" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the protein family information from Pfam into the network.
Creates the node attribute <code class="docutils literal notranslate"><span class="pre">'pfam'</span></code> containing a [list] of
[dict] whose keys are protein family identifier(s) [str] and
corresponding values are [list] of [dict] containing detailed
information about the protein family(ies) for regions and
isoforms of the protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network object for which
the information is to be loaded. If none is passed, takes
the undirected network of the current instance.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_pfam3">
<span class="sig-name descname"><span class="pre">load_pfam3</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_pfam3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_pfam3" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the protein domain information from Pfam into the network.
Creates the node attribute <code class="docutils literal notranslate"><span class="pre">'doms'</span></code> containing a [list] of
<code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.intera.Domain</span></code> instances with information
about each domain of the protein (see the corresponding class
documentation for more information).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_phospho_dmi">
<span class="sig-name descname"><span class="pre">load_phospho_dmi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_phospho_dmi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_phospho_dmi" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_phosphoelm">
<span class="sig-name descname"><span class="pre">load_phosphoelm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_phosphoelm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_phosphoelm" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_pisa">
<span class="sig-name descname"><span class="pre">load_pisa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_pisa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_pisa" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_pnetworks_dmi">
<span class="sig-name descname"><span class="pre">load_pnetworks_dmi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_pnetworks_dmi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_pnetworks_dmi" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_psite_phos">
<span class="sig-name descname"><span class="pre">load_psite_phos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_psite_phos"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_psite_phos" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_psite_reg">
<span class="sig-name descname"><span class="pre">load_psite_reg</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_psite_reg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_psite_reg" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_ptms">
<span class="sig-name descname"><span class="pre">load_ptms</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_ptms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_ptms" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_ptms2">
<span class="sig-name descname"><span class="pre">load_ptms2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_methods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">map_by_homology_from</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[9606]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homology_only_swissprot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_homology_strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nonhuman_direct_lookup</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inputargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_load</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_ptms2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_ptms2" title="Permalink to this definition">§</a></dt>
<dd><p>This is a new method which will replace <cite>load_ptms</cite>.
It uses <cite>pypath.enz_sub.EnzymeSubstrateAggregator</cite>, a newly
introduced module for combining enzyme-substrate data from multiple
resources using homology translation on users demand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_methods</strong> (<em>list</em>) – Resources to collect enzyme-substrate
interactions from. E.g. <cite>[‘Signor’, ‘phosphoELM’]</cite>. By default
it contains Signor, PhosphoSitePlus, HPRD, phosphoELM, dbPTM,
PhosphoNetworks, Li2012 and MIMP.</p></li>
<li><p><strong>map_by_homology_from</strong> (<em>list</em>) – List of NCBI Taxonomy IDs of
source taxons used for homology translation of enzyme-substrate
interactions. If you have a human network and you add here
<cite>[10090, 10116]</cite> then mouse and rat interactions from the source
databases will be translated to human.</p></li>
<li><p><strong>homology_only_swissprot</strong> (<em>bool</em>) – <cite>True</cite> by default which means
only SwissProt IDs are accepted at homology translateion, Trembl
IDs will be dropped.</p></li>
<li><p><strong>ptm_homology_strict</strong> (<em>bool</em>) – For homology translation use
PhosphoSite’s PTM homology table. This guarantees that only
truely homologous sites will be included. Otherwise we only
check if at the same numeric offset in the homologous sequence
the appropriate residue can be find.</p></li>
<li><p><strong>nonhuman_direct_lookup</strong> (<em>bool</em>) – Fetch also directly nonhuman
data from the resources whereever it’s available. PhosphoSite
contains mouse enzyme-substrate interactions and it is possible
to extract these directly beside translating the human ones
to mouse.</p></li>
<li><p><strong>inputargs</strong> (<em>dict</em>) – Additional arguments passed to <cite>PtmProcessor</cite>.
A <cite>dict</cite> can be supplied for each resource, e.g.
<cite>{‘Signor’: {…}, ‘PhosphoSite’: {…}, …}</cite>.
Those not used by <cite>PtmProcessor</cite> are forwarded to the
<cite>pypath.inputs</cite> methods.</p></li>
<li><p><strong>database</strong> – A <code class="docutils literal notranslate"><span class="pre">PtmAggregator</span></code> object. If provided no new database will be
created.</p></li>
<li><p><strong>force_load</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> the database will be loaded with the parameters
provided here; otherwise if the <code class="docutils literal notranslate"><span class="pre">ptm</span></code> module already has a
database no new database will be created. This means the
parameters specified in other arguments might have no effect.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_resource">
<span class="sig-name descname"><span class="pre">load_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">settings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reread</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">redownload</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Loads the data from a single resource and attaches it to the
network</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>settings</strong> (<em>pypath.input_formats.NetworkInput</em>) – <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.input_formats.NetworkInput</span></code> instance
containing the detailed definition of the input format to
the downloaded file.</p></li>
<li><p><strong>clean</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Whether to clean the graph
after importing the data or not. See
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.clean_graph()</span></code> for more
information.</p></li>
<li><p><strong>cache_files</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">{}</span></code> by default. Contains the resource name(s)
[str] (keys) and the corresponding cached file name [str].
If provided (and file exists) bypasses the download of the
data for that resource and uses the cache file instead.</p></li>
<li><p><strong>reread</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Specifies whether to reread
the data files from the cache or omit them (similar to
<em>redownload</em>).</p></li>
<li><p><strong>redownload</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Specifies whether to
re-download the data and ignore the cache.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_resources">
<span class="sig-name descname"><span class="pre">load_resources</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reread</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">redownload</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_resources"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_resources" title="Permalink to this definition">§</a></dt>
<dd><p>Loads multiple resources, and cleans up after. Looks up ID
types, and loads all ID conversion tables from UniProt if
necessary. This is much faster than loading the ID conversion
and the resources one by one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies the data input
formats for the different resources (keys) [str]. Values
are <code class="xref py py-class docutils literal notranslate"><span class="pre">pypath.input_formats.NetworkInput</span></code> instances
containing the information. By default uses the set of
resources of OmniPath.</p></li>
<li><p><strong>exclude</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">[]</span></code> by default. List of resources [str] to
exclude from the network.</p></li>
<li><p><strong>cache_files</strong> (<em>dict</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">{}</span></code> by default. Contains the resource name(s)
[str] (keys) and the corresponding cached file name [str].
If provided (and file exists) bypasses the download of the
data for that resource and uses the cache file instead.</p></li>
<li><p><strong>reread</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Specifies whether to reread
the data files from the cache or omit them (similar to
<em>redownload</em>).</p></li>
<li><p><strong>redownload</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Specifies whether to
re-download the data and ignore the cache.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_signor_ptms">
<span class="sig-name descname"><span class="pre">load_signor_ptms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">non_matching</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_signor_ptms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_signor_ptms" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_surfaceome_attrs">
<span class="sig-name descname"><span class="pre">load_surfaceome_attrs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_surfaceome_attrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_surfaceome_attrs" title="Permalink to this definition">§</a></dt>
<dd><p>Loads vertex attributes from the In Silico Human Surfaceome.
Attributes are <code class="docutils literal notranslate"><span class="pre">surfaceome_score</span></code>, <code class="docutils literal notranslate"><span class="pre">surfaceome_class</span></code> and
<code class="docutils literal notranslate"><span class="pre">surfaceome_subclass</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_tfregulons">
<span class="sig-name descname"><span class="pre">load_tfregulons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">levels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'A',</span> <span class="pre">'B'}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_curated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.load_tfregulons" title="Permalink to this definition">§</a></dt>
<dd><p>Adds TF-target interactions from TF regulons to the network.
DoRothEA is a comprehensive resource of TF-target
interactions combining multiple lines of evidences: literature
curated databases, ChIP-Seq data, PWM based prediction using
HOCOMOCO and JASPAR matrices and prediction from GTEx expression
data by ARACNe.</p>
<p>For details see <a class="github reference external" href="https://github.com/saezlab/DoRothEA">saezlab/DoRothEA</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>levels</strong> (<em>set</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">{'A',</span> <span class="pre">'B'}</span></code> by default. Confidence levels to be
loaded (from A to E) [str].</p></li>
<li><p><strong>only_curated</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to retrieve only the
literature curated interactions or not.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.load_vesiclepedia_attrs">
<span class="sig-name descname"><span class="pre">load_vesiclepedia_attrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">load_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_refs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_vesicle_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.load_vesiclepedia_attrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.load_vesiclepedia_attrs" title="Permalink to this definition">§</a></dt>
<dd><p>Creates vertex attributes from Vesiclepedia data. Creates a boolean
attribute <code class="docutils literal notranslate"><span class="pre">vesiclepedia_in_vesicle</span></code> which tells whether a protein is
in ExoCarta i.e. has been found in exosomes. Optionally creates
attributes <code class="docutils literal notranslate"><span class="pre">vesiclepedia_samples</span></code>, <code class="docutils literal notranslate"><span class="pre">vesiclepedia_refs</span></code> and
<code class="docutils literal notranslate"><span class="pre">vesiclepedia_vesicles</span></code> listing the sample tissue, the PubMed
references and the vesicle types, respectively.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.lookup_cache">
<span class="sig-name descname"><span class="pre">lookup_cache</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int_cache</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges_cache</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.lookup_cache"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.lookup_cache" title="Permalink to this definition">§</a></dt>
<dd><p>Checks up the cache folder for the files of a given resource.
First checks if <em>name</em> is on the <em>cache_files</em> dictionary.
If so, loads either the interactions or edges otherwise. If
not, checks <em>edges_cache</em> or <em>int_cache</em> otherwise.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Name of the resource (lower-case).</p></li>
<li><p><strong>cache_files</strong> (<em>dict</em>) – Contains the resource name(s) [str] (keys) and the
corresponding cached file name [str] (values).</p></li>
<li><p><strong>int_cache</strong> (<em>str</em>) – Path to the interactions cache file of the resource.</p></li>
<li><p><strong>edges_cache</strong> (<em>str</em>) – Path to the edges cache file of the resource.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>(<em>file</em>) – The loaded pickle file from the cache if the
file is contains the interactions. <code class="docutils literal notranslate"><span class="pre">None</span></code> otherwise.</p></li>
<li><p>(<em>list</em>) – List of mapped edges if the file contains the
information from the edges. <code class="docutils literal notranslate"><span class="pre">[]</span></code> otherwise.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.loop_edges">
<span class="sig-name descname"><span class="pre">loop_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.loop_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.loop_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Returns an iterator of the indices (or the edge instances) of
the edges which represent a loop (whose source and target node
are the same).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Whether to return the
iterator of the indices or the edge instances.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The graph object where the
edge loops are to be searched. If none is passed, takes the
undirected network of the current instance.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>generator</em>) – Generator object containing the edge
indices (or instances) containing loops.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.mean_reference_per_interaction">
<span class="sig-name descname"><span class="pre">mean_reference_per_interaction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.mean_reference_per_interaction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.mean_reference_per_interaction" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the mean number of references per interaction of the
network.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>float</em>) – Mean number of interactions per edge.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.mean_reference_per_interaction_by_resource">
<span class="sig-name descname"><span class="pre">mean_reference_per_interaction_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.mean_reference_per_interaction_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.mean_reference_per_interaction_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the mean number of references per interaction of the
network.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>float</em>) – Mean number of interactions per edge.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.merge_lists">
<span class="sig-name descname"><span class="pre">merge_lists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">and_or</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'and'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delete</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.merge_lists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.merge_lists" title="Permalink to this definition">§</a></dt>
<dd><p>Merges two lists from <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypat.main.PyPath.lists</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_a</strong> (<em>str</em>) – Name of the first list to be merged.</p></li>
<li><p><strong>id_b</strong> (<em>str</em>) – Name of the second list to be merged.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies a new name for the
merged list. If none is passed, name will be set to
<em>id_a*_*id_b</em>.</p></li>
<li><p><strong>and_or</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'and'</span></code> by default. The logic operation perfomed
in the merging: <code class="docutils literal notranslate"><span class="pre">'and'</span></code> performs an union, <code class="docutils literal notranslate"><span class="pre">'or'</span></code> for
the intersection.</p></li>
<li><p><strong>delete</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to delete the
former lists or not.</p></li>
<li><p><strong>func</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'max'</span></code> by default. Not used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.merge_nodes">
<span class="sig-name descname"><span class="pre">merge_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nodes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.merge_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.merge_nodes" title="Permalink to this definition">§</a></dt>
<dd><p>Merges all attributes and edges of selected nodes and assigns
them to the primary node (by default the one with lowest index).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<em>list</em>) – List of node indexes [int] that are to be collapsed.</p></li>
<li><p><strong>primary</strong> (<em>int</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. ID of the primary edge, if
none is passed, the node with lowest index on <em>nodes</em> is
selected.</p></li>
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network graph object from
which the nodes are to be merged. If none is passed, takes
the undirected network graph.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.mimp_directions">
<span class="sig-name descname"><span class="pre">mimp_directions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.mimp_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.mimp_directions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.mutated_edges">
<span class="sig-name descname"><span class="pre">mutated_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.mutated_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.mutated_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Compares the mutated residues and the modified residues in PTMs.
Interactions are marked as mutated if the target residue in the
underlying PTM is mutated.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.name_edgelist">
<span class="sig-name descname"><span class="pre">name_edgelist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.name_edgelist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.name_edgelist" title="Permalink to this definition">§</a></dt>
<dd><p>Returns an edge list, i.e. a list with tuples of vertex names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.names2vids">
<span class="sig-name descname"><span class="pre">names2vids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.names2vids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.names2vids" title="Permalink to this definition">§</a></dt>
<dd><p>From a list of node names, returns their corresponding indices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>names</strong> (<em>list</em>) – Contains the node names [str] for which the IDs are to be
searched.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – The queried node IDs [int].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.negative_report">
<span class="sig-name descname"><span class="pre">negative_report</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outFile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.negative_report"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.negative_report" title="Permalink to this definition">§</a></dt>
<dd><p>Generates a report file with the negative interactions (assumed
to be already loaded).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Whether to retun a list of
edges containing the edge instances which have negative
references.</p></li>
<li><p><strong>outFile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The output file name/path. If
none is passed, the default is
<code class="docutils literal notranslate"><span class="pre">'results/&lt;session_id&gt;-negatives'</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – If <em>lst</em> is set to <code class="docutils literal notranslate"><span class="pre">True</span></code>, returns a [list]
is returned with the <code class="xref py py-class docutils literal notranslate"><span class="pre">igraph.Edge</span></code> instances that
contain at least a negative reference.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.neighborhood">
<span class="sig-name descname"><span class="pre">neighborhood</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifiers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ALL'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.neighborhood"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.neighborhood" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.neighbors">
<span class="sig-name descname"><span class="pre">neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ALL'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.neighbors" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.neighbourhood_network">
<span class="sig-name descname"><span class="pre">neighbourhood_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">center</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.neighbourhood_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.neighbourhood_network" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.network_by_go">
<span class="sig-name descname"><span class="pre">network_by_go</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node_categories</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_sources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_undirected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GO'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delete</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertex_attrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_attrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.network_by_go"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.network_by_go" title="Permalink to this definition">§</a></dt>
<dd><p>Creates or filters a network based on Gene Ontology annotations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>node_categories</strong> (<em>dict</em>) – <p>A dict with custom category labels as keys and expressions of
GO terms as values. E.g.
<a href="#id37"><span class="problematic" id="id38">``</span></a>{‘extracell’: ‘<a class="reference external" href="GO:0005576">GO:0005576</a> and not <a class="reference external" href="GO:0070062">GO:0070062</a>’,</p>
<blockquote>
<div><p>’plasmamem’: ‘<a class="reference external" href="GO:0005887">GO:0005887</a>’}``.</p>
</div></blockquote>
</p></li>
<li><p><strong>network_sources</strong> (<em>dict</em>) – A dict with anything as keys and network input format definintions
(<code class="docutils literal notranslate"><span class="pre">input_formats.NetworkInput</span></code> instances) as values.</p></li>
<li><p><strong>include</strong> (<em>list</em>) – A list of tuples of category label pairs. By default we keep all
edges connecting proteins annotated with any of the defined
categories. If <code class="docutils literal notranslate"><span class="pre">include</span></code> is defined then only edges between
category pairs defined here will be kept and all others deleted.</p></li>
<li><p><strong>exclude</strong> (<em>list</em>) – Similarly to include, all edges will be kept but the ones listed
in <code class="docutils literal notranslate"><span class="pre">exclude</span></code> will be deleted.</p></li>
<li><p><strong>directed</strong> (<em>bool</em>) – If True <code class="docutils literal notranslate"><span class="pre">include</span></code> and <code class="docutils literal notranslate"><span class="pre">exclude</span></code> relations will be processed
with directed (source, target) else direction won’t be considered.</p></li>
<li><p><strong>keep_undirected</strong> (<em>bool</em>) – If True the interactions without direction information will be
kept even if <code class="docutils literal notranslate"><span class="pre">directed</span></code> is True. Passed to <code class="docutils literal notranslate"><span class="pre">edges_between</span></code>
as <code class="docutils literal notranslate"><span class="pre">strict</span></code> argument.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix for all vertex and edge attributes created in this
operation. E.g. if you have a category label ‘bar’ and prefix
is ‘foo’ then you will have a new vertex attribute ‘foo__bar’.</p></li>
<li><p><strong>delete</strong> (<em>bool</em>) – Delete the vertices and edges which don’t belong to any of the
categories.</p></li>
<li><p><strong>copy</strong> (<em>bool</em>) – Return a copy of the entire <code class="docutils literal notranslate"><span class="pre">PyPath</span></code> object with the graph
filtered by GO terms. By default the object is modified in place
and <code class="docutils literal notranslate"><span class="pre">None</span></code> is returned.</p></li>
<li><p><strong>vertex_attrs</strong> (<em>bool</em>) – Create vertex attributes.</p></li>
<li><p><strong>edge_attrs</strong> (<em>bool</em>) – Create edge attributes.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.network_filter">
<span class="sig-name descname"><span class="pre">network_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.network_filter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.network_filter" title="Permalink to this definition">§</a></dt>
<dd><p>This function aims to cut the number of edges in the network,
without losing nodes, to make the network less connected,
less hairball-like, more usable for analysis.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.network_stats">
<span class="sig-name descname"><span class="pre">network_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.network_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.network_stats" title="Permalink to this definition">§</a></dt>
<dd><p>Calculates basic statistics for the whole network and each of
sources (node and edge counts, average node degree, graph
diameter, transitivity, adhesion and cohesion). Writes the
results in a tab file. File is stored in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.outdir</span></code> (<code class="docutils literal notranslate"><span class="pre">'results'</span></code> by default).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies the file name. If
none is specified, this will be
<code class="docutils literal notranslate"><span class="pre">'pwnet-&lt;session_id&gt;-stats'</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.new_edges">
<span class="sig-name descname"><span class="pre">new_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.new_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.new_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Adds new edges from any iterable of edges to the undirected
graph. Basically, calls <code class="xref py py-meth docutils literal notranslate"><span class="pre">igraph.Graph.add_edges()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>edges</strong> (<em>list</em>) – Contains the edges that are to be added to the network.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.new_nodes">
<span class="sig-name descname"><span class="pre">new_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nodes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.new_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.new_nodes" title="Permalink to this definition">§</a></dt>
<dd><p>Adds new nodes from any iterable of nodes to the undirected
graph. Basically, calls <code class="xref py py-meth docutils literal notranslate"><span class="pre">igraph.Graph.add_vertices()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nodes</strong> (<em>list</em>) – Contains the nodes that are to be added to the network.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.node_exists">
<span class="sig-name descname"><span class="pre">node_exists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.node_exists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.node_exists" title="Permalink to this definition">§</a></dt>
<dd><p>Checks if a node exists in the (undirected) network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – The name of the node to be searched.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>bool</em>) – Whether the node exists in the network or not.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.numof_directed_edges">
<span class="sig-name descname"><span class="pre">numof_directed_edges</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.numof_directed_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.numof_directed_edges" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.numof_edges">
<span class="sig-name descname"><span class="pre">numof_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.numof_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.numof_edges" title="Permalink to this definition">§</a></dt>
<dd><p>Number of edges optionally limited to certain resources.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.numof_reference_interaction_pairs">
<span class="sig-name descname"><span class="pre">numof_reference_interaction_pairs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.numof_reference_interaction_pairs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.numof_reference_interaction_pairs" title="Permalink to this definition">§</a></dt>
<dd><p>Returns the total of unique references per interaction.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>int</em>) – Total number of unique references per
interaction.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.numof_references_by_resource">
<span class="sig-name descname"><span class="pre">numof_references_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.numof_references_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.numof_references_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Counts the references for each resource, optionally limited
to certain resources.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.numof_undirected_edges">
<span class="sig-name descname"><span class="pre">numof_undirected_edges</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.numof_undirected_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.numof_undirected_edges" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.orthology_translation">
<span class="sig-name descname"><span class="pre">orthology_translation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_swissprot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.orthology_translation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.orthology_translation" title="Permalink to this definition">§</a></dt>
<dd><p>Translates the current object to another organism by orthology.
Proteins without known ortholog will be deleted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>target</strong> (<em>int</em>) – NCBI Taxonomy ID of the target organism. E.g. 10090 for mouse.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.p">
<span class="sig-name descname"><span class="pre">p</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.p" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;identifier<span class="classifier">int, str</span></dt><dd><p>Vertex index (int) or GeneSymbol (str) or UniProt ID (str) or
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.pathway_attributes">
<span class="sig-name descname"><span class="pre">pathway_attributes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.pathway_attributes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.pathway_attributes" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.pathway_members">
<span class="sig-name descname"><span class="pre">pathway_members</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pathway</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.pathway_members"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.pathway_members" title="Permalink to this definition">§</a></dt>
<dd><p>Returns an iterator with the members of a single pathway.
Apart from the pathway name you need to supply its source
database too.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.pathway_names">
<span class="sig-name descname"><span class="pre">pathway_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.pathway_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.pathway_names" title="Permalink to this definition">§</a></dt>
<dd><p>Returns the names of all pathways having at least one member
in the current graph.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.pathway_similarity">
<span class="sig-name descname"><span class="pre">pathway_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.pathway_similarity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.pathway_similarity" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the Sorensen’s similarity index across nodes and edges
for all the available pathway sources (already loaded in the
network) and saves them into table files. Files are stored in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.outdir</span></code> (<code class="docutils literal notranslate"><span class="pre">'results'</span></code> by default).
See <code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.sorensen_pathways()</span></code> for more
information..</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies the file name
prefix (suffixes will be <code class="docutils literal notranslate"><span class="pre">'-nodes'</span></code> and <code class="docutils literal notranslate"><span class="pre">'-edges'</span></code>). If
none is specified, this will be
<code class="docutils literal notranslate"><span class="pre">'pwnet-&lt;session_id&gt;-sim-pw'</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.pathways_table">
<span class="sig-name descname"><span class="pre">pathways_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_pathways.list'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pw_sources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['signalink',</span> <span class="pre">'signor',</span> <span class="pre">'netpath',</span> <span class="pre">'kegg']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.pathways_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.pathways_table" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.pfam_regions">
<span class="sig-name descname"><span class="pre">pfam_regions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.pfam_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.pfam_regions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.phosphonetworks_directions">
<span class="sig-name descname"><span class="pre">phosphonetworks_directions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.phosphonetworks_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.phosphonetworks_directions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.phosphopoint_directions">
<span class="sig-name descname"><span class="pre">phosphopoint_directions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.phosphopoint_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.phosphopoint_directions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.phosphorylation_directions">
<span class="sig-name descname"><span class="pre">phosphorylation_directions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.phosphorylation_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.phosphorylation_directions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.phosphorylation_signs">
<span class="sig-name descname"><span class="pre">phosphorylation_signs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.phosphorylation_signs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.phosphorylation_signs" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.phosphosite_directions">
<span class="sig-name descname"><span class="pre">phosphosite_directions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.phosphosite_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.phosphosite_directions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.prdb_tissue_expr">
<span class="sig-name descname"><span class="pre">prdb_tissue_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tissue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prdb=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">occurrence=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_function=&lt;function</span> <span class="pre">PyPath.&lt;lambda&gt;&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">na_value=0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.prdb_tissue_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.prdb_tissue_expr" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.process_directions">
<span class="sig-name descname"><span class="pre">process_directions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stimulation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inhibition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirs_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.process_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.process_directions" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.process_dmi">
<span class="sig-name descname"><span class="pre">process_dmi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.process_dmi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.process_dmi" title="Permalink to this definition">§</a></dt>
<dd><p>This is an universal function
for loading domain-motif objects
like load_phospho_dmi() for PTMs.
TODO this will replace load_elm, load_ielm, etc</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.protein">
<span class="sig-name descname"><span class="pre">protein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.protein" title="Permalink to this definition">§</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.
Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;identifier<span class="classifier">int, str</span></dt><dd><p>Vertex index (int) or GeneSymbol (str) or UniProt ID (str) or
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.protein_edge">
<span class="sig-name descname"><span class="pre">protein_edge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.protein_edge" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<em>int</em><em>,</em><em>str</em>) – Vertex index or UniProt ID or GeneSymbol or <code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code>
object.</p></li>
<li><p><strong>target</strong> (<em>int</em><em>,</em><em>str</em>) – Vertex index or UniProt ID or GeneSymbol or <code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code>
object.</p></li>
<li><p><strong>directed</strong> (<em>bool</em>) – To be passed to igraph.Graph.get_eid()</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.proteins">
<span class="sig-name descname"><span class="pre">proteins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifiers</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.proteins" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.ps">
<span class="sig-name descname"><span class="pre">ps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifiers</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.ps" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.random_walk_with_return">
<span class="sig-name descname"><span class="pre">random_walk_with_return</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">niter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.random_walk_with_return"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.random_walk_with_return" title="Permalink to this definition">§</a></dt>
<dd><p>Random walk with return (RWR) starting from one or more query nodes.
Returns affinity (probability) vector of all nodes in the graph.</p>
<p>:param : Vertex IDs of query nodes.
:type : param int,list q:
:param : An <cite>igraph.Graph</cite> object.
:type : param igraph.Graph graph:
:param : Probability of restart.
:type : param float c:
:param : Number of iterations.
:type : param int niter:</p>
<section id="example">
<h2>Example:<a class="headerlink" href="#example" title="Permalink to this heading">§</a></h2>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">igraph</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pypath</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pa</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">PyPath</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pa</span><span class="o">.</span><span class="n">init_network</span><span class="p">({</span>
<span class="go">        &#39;signor&#39;: pypath.data_formats.pathway[&#39;signor&#39;]</span>
<span class="go">    })</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q</span> <span class="o">=</span> <span class="p">[</span>
<span class="go">        pa.gs(&#39;EGFR&#39;).index,</span>
<span class="go">        pa.gs(&#39;ATG4B&#39;).index</span>
<span class="go">    ]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rwr</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">random_walk_with_return</span><span class="p">(</span><span class="n">q</span> <span class="o">=</span> <span class="n">q</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">palette</span> <span class="o">=</span> <span class="n">igraph</span><span class="o">.</span><span class="n">RainbowPalette</span><span class="p">(</span><span class="n">n</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">colors</span>  <span class="o">=</span> <span class="p">[</span><span class="n">palette</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">i</span><span class="p">)))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">rwr</span> <span class="o">/</span> <span class="nb">max</span><span class="p">(</span><span class="n">rwr</span><span class="p">)</span> <span class="o">*</span> <span class="mi">99</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">igraph</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="p">,</span> <span class="n">vertex_color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">)</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.random_walk_with_return2">
<span class="sig-name descname"><span class="pre">random_walk_with_return2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">niter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.random_walk_with_return2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.random_walk_with_return2" title="Permalink to this definition">§</a></dt>
<dd><p>Literally does random walks.
Only for testing of the other method, to be deleted later.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.read_from_cache">
<span class="sig-name descname"><span class="pre">read_from_cache</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cache_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.read_from_cache"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.read_from_cache" title="Permalink to this definition">§</a></dt>
<dd><p>Reads a pickle file from the cache and returns it. It is assumed
that the subfolder <code class="docutils literal notranslate"><span class="pre">cache/</span></code> is on the supplied path.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cache_file</strong> (<em>str</em>) – Path to the cache file that is to be loaded.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>file</em>) – The loaded pickle file from the cache. Type will
depend on the file itself (e.g.: if the pickle was saved
from a dictionary, the type will be [dict]).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.read_list_file">
<span class="sig-name descname"><span class="pre">read_list_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">settings</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.read_list_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.read_list_file" title="Permalink to this definition">§</a></dt>
<dd><p>Reads a list from a file and adds it to
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.lists</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>settings</strong> (<em>pypath.input_formats.ReadList</em>) – <code class="xref py py-class docutils literal notranslate"><span class="pre">python.data_formats.ReadList</span></code> instance specifying
the settings of the file to be read. See the class
documentation for more details.</p></li>
<li><p><strong>**kwargs</strong> – Extra arguments passed to the file reading function. Such
function name is outlined in the
<code class="xref py py-attr docutils literal notranslate"><span class="pre">python.data_formats.ReadList.input</span></code> attribute and
defined in <a class="reference internal" href="pypath.inputs.html#module-pypath.inputs" title="pypath.inputs"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pypath.inputs</span></code></a>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.reference_edge_ratio">
<span class="sig-name descname"><span class="pre">reference_edge_ratio</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.reference_edge_ratio"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.reference_edge_ratio" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the average number of references per edge (as in the
undirected graph).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>float</em>) – Average number of references per edge.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.reference_hist">
<span class="sig-name descname"><span class="pre">reference_hist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.reference_hist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.reference_hist" title="Permalink to this definition">§</a></dt>
<dd><p>Generates a file containing a table with information about the
network’s edges. First column contains the source node ID,
followed by the target’s ID, third column contains the number of
references for that interaction and finally the number of
sources. Writes the results in a tab file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies the file name and
path to save the table. If none is passed, file will be
saved in <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.outdir</span></code> (<code class="docutils literal notranslate"><span class="pre">'results'</span></code>
by default) with the name <code class="docutils literal notranslate"><span class="pre">'&lt;session_id&gt;-refs-hist'</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.references">
<span class="sig-name descname"><span class="pre">references</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.references"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.references" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a set of references for all edges.</p>
<dl class="simple">
<dt>resources<span class="classifier">None,str,set</span></dt><dd><p>Limits the query to one or more resources.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.references_by_resource">
<span class="sig-name descname"><span class="pre">references_by_resource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.references_by_resource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.references_by_resource" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a dict with resources as keys and sets of references
as values.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.reload">
<span class="sig-name descname"><span class="pre">reload</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.reload"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.reload" title="Permalink to this definition">§</a></dt>
<dd><p>Reloads the object from the module level.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.remove_htp">
<span class="sig-name descname"><span class="pre">remove_htp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.remove_htp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.remove_htp" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.remove_undirected">
<span class="sig-name descname"><span class="pre">remove_undirected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_refs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.remove_undirected"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.remove_undirected" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.resources">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">resources</span></span><a class="headerlink" href="#pypath.legacy.main.PyPath.resources" title="Permalink to this definition">§</a></dt>
<dd><p>All network resources. Returns <em>set</em> of strings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.run_batch">
<span class="sig-name descname"><span class="pre">run_batch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">methods</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">toCall</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.run_batch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.run_batch" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.save_network">
<span class="sig-name descname"><span class="pre">save_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pickle_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.save_network" title="Permalink to this definition">§</a></dt>
<dd><p>Saves the network object.</p>
<p>Stores the instance into a pickle (binary) file which can be
reloaded in the future.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>pickle_file</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The path/file name where to
store the pcikle file. If not specified, saves the network
to its default location
(<code class="docutils literal notranslate"><span class="pre">'cache/default_network.pickle'</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.save_session">
<span class="sig-name descname"><span class="pre">save_session</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.save_session"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.save_session" title="Permalink to this definition">§</a></dt>
<dd><p>Save the current session state into pickle dump. The file will
be saved in the current working directory as
<code class="docutils literal notranslate"><span class="pre">'pypath-&lt;session_id&gt;.pickle'</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.save_to_pickle">
<span class="sig-name descname"><span class="pre">save_to_pickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pickle_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.save_to_pickle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.save_to_pickle" title="Permalink to this definition">§</a></dt>
<dd><p>Saves the network object.</p>
<p>Stores the instance into a pickle (binary) file which can be
reloaded in the future.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>pickle_file</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The path/file name where to
store the pcikle file. If not specified, saves the network
to its default location
(<code class="docutils literal notranslate"><span class="pre">'cache/default_network.pickle'</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.search_attr_and">
<span class="sig-name descname"><span class="pre">search_attr_and</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lst</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.search_attr_and"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.search_attr_and" title="Permalink to this definition">§</a></dt>
<dd><p>Searches a given collection of attributes in a given object.
Only returns <code class="docutils literal notranslate"><span class="pre">True</span></code>, if all elements of <em>lst</em> can be found in
<em>obj</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<em>object</em>) – Object (dictionary-like) where to search for elements of
<em>lst</em>.</p></li>
<li><p><strong>lst</strong> (<em>dict</em>) – Keys are the attribute names [str] and values the collection
of elements to be searched in such attribute [set].</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>bool</em>) – <code class="docutils literal notranslate"><span class="pre">True</span></code> only if <em>lst</em> is empty or all of its
elements are found in <em>obj</em>. Returns <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise (as
soon as one element of <em>lst</em> is not found).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.search_attr_or">
<span class="sig-name descname"><span class="pre">search_attr_or</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lst</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.search_attr_or"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.search_attr_or" title="Permalink to this definition">§</a></dt>
<dd><p>Searches a given collection of attributes in a given object. As
soon as one item is found, returns <code class="docutils literal notranslate"><span class="pre">True</span></code>, if none could be
found then returns <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<em>object</em>) – Object (dictionary-like) where to search for elements of
<em>lst</em>.</p></li>
<li><p><strong>lst</strong> (<em>dict</em>) – Keys are the attribute names [str] and values the collection
of elements to be searched in such attribute [set].</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>bool</em>) – <code class="docutils literal notranslate"><span class="pre">True</span></code> if <em>lst</em> is empty or any of its
elements is found in <em>obj</em>. Returns only <code class="docutils literal notranslate"><span class="pre">False</span></code> if cannot
find anything.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.second_neighbours">
<span class="sig-name descname"><span class="pre">second_neighbours</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indices</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_first</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.second_neighbours"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.second_neighbours" title="Permalink to this definition">§</a></dt>
<dd><p>Looks for the (first and) second neighbours of a given node and
returns a list of their UniProt IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>node</strong> (<em>str</em>) – The UniProt ID of the node of interest. Can also be the
index of such node [int].</p></li>
<li><p><strong>indices</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to return the
neighbour nodes indices or their UniProt IDs.</p></li>
<li><p><strong>wit_first</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to return also the
first neighbours or not.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – The list containing the second neighbours of the
queried node (including the first ones if specified).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.select_by_go">
<span class="sig-name descname"><span class="pre">select_by_go</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">go_terms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.select_by_go"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.select_by_go" title="Permalink to this definition">§</a></dt>
<dd><p>Retrieves the vertex IDs of all vertices annotated with any
Gene Ontology terms or their descendants, or evaluates string
expression (see <code class="docutils literal notranslate"><span class="pre">select_by_go_expr</span></code>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>go_terms</strong> (<em>str</em><em>,</em><em>set</em>) – A single GO term, a set of GO terms or an expression with
GO terms.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.select_by_go_all">
<span class="sig-name descname"><span class="pre">select_by_go_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">go_terms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.select_by_go_all"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.select_by_go_all" title="Permalink to this definition">§</a></dt>
<dd><p>Selects the nodes annotated by all GO terms in <code class="docutils literal notranslate"><span class="pre">go_terms</span></code>.</p>
<p>Returns set of vertex IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>go_terms</strong> (<em>list</em>) – List, set or tuple of GO terms.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.select_by_go_expr">
<span class="sig-name descname"><span class="pre">select_by_go_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">go_expr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.select_by_go_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.select_by_go_expr" title="Permalink to this definition">§</a></dt>
<dd><p>Selects vertices based on an expression of Gene Ontology terms.
Operator precedence not considered, please use parentheses.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>go_expr</strong> (<em>str</em>) – An expression of Gene Ontology terms. E.g.
<code class="docutils literal notranslate"><span class="pre">'(GO:0005576</span> <span class="pre">and</span> <span class="pre">not</span> <span class="pre">GO:0070062)</span> <span class="pre">or</span> <span class="pre">GO:0005887'</span></code>. Parentheses
and operators <code class="docutils literal notranslate"><span class="pre">and</span></code>, <code class="docutils literal notranslate"><span class="pre">or</span></code> and <code class="docutils literal notranslate"><span class="pre">not</span></code> can be used.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.separate">
<span class="sig-name descname"><span class="pre">separate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.separate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.separate" title="Permalink to this definition">§</a></dt>
<dd><p>Separates the undirected network according to the different
sources. Basically applies
<code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.get_network()</span></code> for each resource.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>dict</em>) – Keys are resource names [str] whose values are
the subnetwork [igraph.Graph] containing the elements of
that source.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.separate_by_category">
<span class="sig-name descname"><span class="pre">separate_by_category</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.separate_by_category"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.separate_by_category" title="Permalink to this definition">§</a></dt>
<dd><p>Separates the undirected network according to resource
categories. Possible categories are:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">'m'</span></code>: PTM/enzyme-substrate resources.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'p'</span></code>: Pathway/activity flow resources.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'i'</span></code>: Undirected/PPI resources.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'r'</span></code>: Process description/reaction resources.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'t'</span></code>: Transcription resources.</p></li>
</ul>
</div></blockquote>
<p>Works in the same way as <code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.separate()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>dict</em>) – Keys are category names [str] whose values are
the subnetwork [igraph.Graph] containing the elements of
those resources corresponding to that category.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.sequences">
<span class="sig-name descname"><span class="pre">sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">isoforms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.sequences" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_boolean_vattr">
<span class="sig-name descname"><span class="pre">set_boolean_vattr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_boolean_vattr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_boolean_vattr" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_categories">
<span class="sig-name descname"><span class="pre">set_categories</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_categories"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_categories" title="Permalink to this definition">§</a></dt>
<dd><p>Sets the category attribute on the network nodes and edges
(<code class="docutils literal notranslate"><span class="pre">'cat'</span></code>) as well the edge attribute coercing the references
by category (<code class="docutils literal notranslate"><span class="pre">'refs_by_cat'</span></code>). The possible categories are
as follows:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">'m'</span></code>: PTM/enzyme-substrate resources.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'p'</span></code>: Pathway/activity flow resources.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'i'</span></code>: Undirected/PPI resources.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'r'</span></code>: Process description/reaction resources.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'t'</span></code>: Transcription resources.</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_chembl_mysql">
<span class="sig-name descname"><span class="pre">set_chembl_mysql</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">title</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_chembl_mysql"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_chembl_mysql" title="Permalink to this definition">§</a></dt>
<dd><p>Sets the ChEMBL MySQL configuration according to the <em>title</em>
section in <em>config_file</em> ini file configuration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>title</strong> (<em>str</em>) – Section title of the ini file.</p></li>
<li><p><strong>config_file</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies the configuration
file name if none is passed, <code class="docutils literal notranslate"><span class="pre">mysql_config/defaults.mysql</span></code>
will be used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_disease_genes">
<span class="sig-name descname"><span class="pre">set_disease_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'curated'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_disease_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_disease_genes" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute named <cite>dis</cite> with boolean values <em>True</em>
if the protein encoded by a disease related gene according to
DisGeNet.</p>
<p>:param : Which dataset to use from DisGeNet. Default is <cite>curated</cite>.
:type : param str dataset:</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_druggability">
<span class="sig-name descname"><span class="pre">set_druggability</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_druggability"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_druggability" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <cite>dgb</cite> with value <em>True</em> if
the protein is druggable, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_drugtargets">
<span class="sig-name descname"><span class="pre">set_drugtargets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pchembl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_drugtargets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_drugtargets" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <cite>dtg</cite> with value <em>True</em> if
the protein has at least one compound binding with
affinity higher than <cite>pchembl</cite>, otherwise <em>False</em>.</p>
<p>:param : Pchembl threshold.
:type : param float pchembl:</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_kinases">
<span class="sig-name descname"><span class="pre">set_kinases</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_kinases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_kinases" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <cite>kin</cite> with value <em>True</em> if
the protein is a kinase, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_plasma_membrane_proteins_cspa">
<span class="sig-name descname"><span class="pre">set_plasma_membrane_proteins_cspa</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_plasma_membrane_proteins_cspa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_plasma_membrane_proteins_cspa" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <cite>cspa</cite> with value <em>True</em> if
the protein is a plasma membrane protein according to CPSA,
otherwise <em>False</em>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_plasma_membrane_proteins_cspa_surfaceome">
<span class="sig-name descname"><span class="pre">set_plasma_membrane_proteins_cspa_surfaceome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">score_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_plasma_membrane_proteins_cspa_surfaceome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_plasma_membrane_proteins_cspa_surfaceome" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <code class="docutils literal notranslate"><span class="pre">surf</span></code> with value <em>True</em> if
the protein is a plasma membrane protein according either to the
Cell Surface Protein Atlas or the In Silico Human Surfaceome.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_plasma_membrane_proteins_surfaceome">
<span class="sig-name descname"><span class="pre">set_plasma_membrane_proteins_surfaceome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">score_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_plasma_membrane_proteins_surfaceome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_plasma_membrane_proteins_surfaceome" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <cite>ishs</cite> with value <em>True</em> if
the protein is a plasma membrane protein according to the In Silico
Human Surfaceome, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_receptors">
<span class="sig-name descname"><span class="pre">set_receptors</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_receptors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_receptors" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <cite>rec</cite> with value <em>True</em> if
the protein is a receptor, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_signaling_proteins">
<span class="sig-name descname"><span class="pre">set_signaling_proteins</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_signaling_proteins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_signaling_proteins" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <cite>kin</cite> with value <em>True</em> if
the protein is a kinase, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_tfs">
<span class="sig-name descname"><span class="pre">set_tfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['a',</span> <span class="pre">'b',</span> <span class="pre">'other']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_tfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_tfs" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.set_transcription_factors">
<span class="sig-name descname"><span class="pre">set_transcription_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['a',</span> <span class="pre">'b',</span> <span class="pre">'other']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.set_transcription_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.set_transcription_factors" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a vertex attribute <cite>tf</cite> with value <em>True</em> if
the protein is a transcription factor, otherwise <em>False</em>.</p>
<p>:param : Classes to use from TF Census. Default is <cite>[‘a’, ‘b’, ‘other’]</cite>.
:type : param list classes:</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.shortest_path_dist">
<span class="sig-name descname"><span class="pre">shortest_path_dist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.shortest_path_dist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.shortest_path_dist" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the distribution of shortest paths for each pair of
nodes in the network (or between group(s) of nodes if <em>subset</em>
is provided). <strong>NOTE:</strong> this method can take a while to compute,
e.g.: if the network has 10K nodes, the total number of possible
pairs to compute is:</p>
<div class="math notranslate nohighlight">
\[\binom{10^4}{2} = 49995000\]</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<em>igraph.Graph</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The network object for which
the shortest path distribution is to be computed. If none is
passed, takes the undirected network of the current
instance.</p></li>
<li><p><strong>susbet</strong> (<em>tuple</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Contains two lists of node
indices defining two groups between which the distribution
is to be computed. Can also be [list] if the shortest paths
are to be searched whithin the group. If none is passed, the
whole network is taken by default.</p></li>
<li><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. File name/path to save the
shortest path distribution. If none is passed, no file is
generated.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments passed to
<code class="xref py py-meth docutils literal notranslate"><span class="pre">igraph.Graph.get_shortest_paths()</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – The length of the shortest paths for each pair
of nodes of the network (or whithin/between group/s if
<em>subset</em> is provided).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.signaling_proteins_list">
<span class="sig-name descname"><span class="pre">signaling_proteins_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.signaling_proteins_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.signaling_proteins_list" title="Permalink to this definition">§</a></dt>
<dd><p>Compiles a list of signaling proteins (as opposed to other
proteins like metabolic enzymes, matrix proteins, etc), by
looking up a few simple keywords in short description of GO
terms.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.signor_pathways">
<span class="sig-name descname"><span class="pre">signor_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.signor_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.signor_pathways" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.similarity_groups">
<span class="sig-name descname"><span class="pre">similarity_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'simpson'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.similarity_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.similarity_groups" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the similarity index across the given <em>groups</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>groups</strong> (<em>dict</em>) – Contains the different group names [str] as keys and their
corresponding elements [set].</p></li>
<li><p><strong>index</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'simpson'</span></code> by default. The type of index metric
to use to compute the similarity. Options are <code class="docutils literal notranslate"><span class="pre">'simpson'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'sorensen'</span></code> and <code class="docutils literal notranslate"><span class="pre">'jaccard'</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>dict</em>) – Dictionary of dictionaries containing the groups
names [str] as keys (for both inner and outer dictionaries)
and the index metric as inner value [float] between those
groups.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.small_plot">
<span class="sig-name descname"><span class="pre">small_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.small_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.small_plot" title="Permalink to this definition">§</a></dt>
<dd><p>This method is deprecated, do not use it.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.sorensen_pathways">
<span class="sig-name descname"><span class="pre">sorensen_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pwlist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.sorensen_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.sorensen_pathways" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the Sorensen’s similarity index across nodes and edges
for the given list of pathway sources (all loaded pathway
sources by default).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>pwlist</strong> (<em>list</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. The list of pathway sources
to be compared.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>dict</em>) – Nested dictionaries (three levels). First-level
keys are <code class="docutils literal notranslate"><span class="pre">'nodes'</span></code> and <code class="docutils literal notranslate"><span class="pre">'edges'</span></code>, then second and third
levels correspond to <code class="docutils literal notranslate"><span class="pre">&lt;source&gt;__&lt;patwhay&gt;</span></code> names which map
to the similarity index between those pathways [float].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.source_diagram">
<span class="sig-name descname"><span class="pre">source_diagram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outf</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.source_diagram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.source_diagram" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.source_network">
<span class="sig-name descname"><span class="pre">source_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">font</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'HelveticaNeueLTStd'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.source_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.source_network" title="Permalink to this definition">§</a></dt>
<dd><p>For EMBL branding, use Helvetica Neue Linotype Standard light</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.source_similarity">
<span class="sig-name descname"><span class="pre">source_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.source_similarity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.source_similarity" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the Sorensen’s similarity index across nodes and edges
for all the sources available (already loaded in the network)
and saves them into table files. Files are stored in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.outdir</span></code> (<code class="docutils literal notranslate"><span class="pre">'results'</span></code> by default).
See <code class="xref py py-meth docutils literal notranslate"><span class="pre">pypath.main.PyPath.databases_similarity()</span></code> for more
information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>outfile</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies the file name
prefix (suffixes will be <code class="docutils literal notranslate"><span class="pre">'-nodes'</span></code> and <code class="docutils literal notranslate"><span class="pre">'-edges'</span></code>). If
none is specified, this will be
<code class="docutils literal notranslate"><span class="pre">'pwnet-&lt;session_id&gt;-sim-src'</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.source_stats">
<span class="sig-name descname"><span class="pre">source_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.source_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.source_stats" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.sources_hist">
<span class="sig-name descname"><span class="pre">sources_hist</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.sources_hist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.sources_hist" title="Permalink to this definition">§</a></dt>
<dd><p>Counts the number of sources per interaction in the graph and
saves them into a file named <code class="docutils literal notranslate"><span class="pre">source_num</span></code>. File is stored in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.outdir</span></code> (<code class="docutils literal notranslate"><span class="pre">'results'</span></code> by
default).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.sources_overlap">
<span class="sig-name descname"><span class="pre">sources_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">diagonal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.sources_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.sources_overlap" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.sources_venn_data">
<span class="sig-name descname"><span class="pre">sources_venn_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.sources_venn_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.sources_venn_data" title="Permalink to this definition">§</a></dt>
<dd><p>Computes the overlap in number of interactions for all pairs of
sources.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. If provided, saves the
results into a table file. File is stored in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.outdir</span></code> (<code class="docutils literal notranslate"><span class="pre">'results'</span></code> by
default).</p></li>
<li><p><strong>return_data</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. Whether to return the
results as a [list].</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – Only if <em>return_data</em> is set to <code class="docutils literal notranslate"><span class="pre">True</span></code>. List
of lists containing the counts for each pair of resources.
This is, for instance, number of interactions only in
resource A, number of interactions only in resource B and
number of common interactions between A and B.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.stats">
<span class="sig-name descname"><span class="pre">stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_collection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.stats" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a collection of entities over the network according to
<code class="docutils literal notranslate"><span class="pre">method</span></code> and counts them. By default the collection won’t be
returned but only the counts.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.straight_between">
<span class="sig-name descname"><span class="pre">straight_between</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_b</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.straight_between"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.straight_between" title="Permalink to this definition">§</a></dt>
<dd><p>Finds an edge between the provided node names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_a</strong> (<em>str</em>) – The name of the source node.</p></li>
<li><p><strong>id_b</strong> (<em>str</em>) – The name of the target node.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>int</em>) – The edge ID. If the edge doesn’t exist, returns
[list] with the node indices [int].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.string_effects">
<span class="sig-name descname"><span class="pre">string_effects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.string_effects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.string_effects" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.sum_in_complex">
<span class="sig-name descname"><span class="pre">sum_in_complex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['corum']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.sum_in_complex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.sum_in_complex" title="Permalink to this definition">§</a></dt>
<dd><p>Returns the total number of edges in the network falling
between two members of the same complex.
Returns as a dict by complex resources.
Calls :py:func:pypath.pypath.Pypath.edges_in_comlexes()
to do the calculations.</p>
<dl class="simple">
<dt>&#64;csources<span class="classifier">list</span></dt><dd><p>List of complex resources. Should be already loaded.</p>
</dd>
<dt>&#64;graph<span class="classifier">igraph.Graph()</span></dt><dd><p>The graph object to do the calculations on.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.summaries_tab">
<span class="sig-name descname"><span class="pre">summaries_tab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.summaries_tab"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.summaries_tab" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a table from resource vs. entity counts and optionally
writes it to <code class="docutils literal notranslate"><span class="pre">outfile</span></code> and returns it.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.table_latex">
<span class="sig-name descname"><span class="pre">table_latex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sum_row</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">latex_hdr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">caption</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'HelveticaNeueLTStd-LtCn'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sum_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Total'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sum_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'%s'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.table_latex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.table_latex" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.third_source_directions">
<span class="sig-name descname"><span class="pre">third_source_directions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_string_effects</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_laudanna_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.third_source_directions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.third_source_directions" title="Permalink to this definition">§</a></dt>
<dd><p>This method calls a series of methods to get
additional direction &amp; effect information
from sources having no literature curated references,
but giving sufficient evidence about the directionality
for interactions already supported by literature
evidences from other sources.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.tissue_network">
<span class="sig-name descname"><span class="pre">tissue_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tissue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.tissue_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.tissue_network" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a network which includes the proteins expressed in
certain tissue according to ProteomicsDB.</p>
<p>:param : Tissue name as used in ProteomicsDB.
:type : param str tissue:
:param : A graph object, by default the <cite>graph</cite> attribute of</p>
<blockquote>
<div><p>the current instance.</p>
</div></blockquote>
<p>:type : param igraph.Graph graph:</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.transcription_factors">
<span class="sig-name descname"><span class="pre">transcription_factors</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.transcription_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.transcription_factors" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.uniprot">
<span class="sig-name descname"><span class="pre">uniprot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.uniprot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.uniprot" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the UniProt
can be found in the default undirected network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;uniprot<span class="classifier">str</span></dt><dd><p>UniProt ID.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.uniprots">
<span class="sig-name descname"><span class="pre">uniprots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprots</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.uniprots"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.uniprots" title="Permalink to this definition">§</a></dt>
<dd><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
undirected graph.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.uniq_node_list">
<span class="sig-name descname"><span class="pre">uniq_node_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.uniq_node_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.uniq_node_list" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a given list of nodes containing only the unique
elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lst</strong> (<em>list</em>) – List of nodes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(<em>list</em>) – Copy of <em>lst</em> containing only unique nodes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.uniq_ptm">
<span class="sig-name descname"><span class="pre">uniq_ptm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.uniq_ptm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.uniq_ptm" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.uniq_ptms">
<span class="sig-name descname"><span class="pre">uniq_ptms</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.uniq_ptms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.uniq_ptms" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up">
<span class="sig-name descname"><span class="pre">up</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.up" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the UniProt
can be found in the default undirected network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;uniprot<span class="classifier">str</span></dt><dd><p>UniProt ID.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_affected_by">
<span class="sig-name descname"><span class="pre">up_affected_by</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_affected_by"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_affected_by" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_affects">
<span class="sig-name descname"><span class="pre">up_affects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_affects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_affects" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_edge">
<span class="sig-name descname"><span class="pre">up_edge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_edge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_edge" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;source<span class="classifier">str</span></dt><dd><p>UniProt ID</p>
</dd>
<dt>&#64;target<span class="classifier">str</span></dt><dd><p>UniProt ID</p>
</dd>
<dt>&#64;directed<span class="classifier">bool</span></dt><dd><p>To be passed to igraph.Graph.get_eid()</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_in_directed">
<span class="sig-name descname"><span class="pre">up_in_directed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_in_directed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_in_directed" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_in_undirected">
<span class="sig-name descname"><span class="pre">up_in_undirected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_in_undirected"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_in_undirected" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_inhibited_by">
<span class="sig-name descname"><span class="pre">up_inhibited_by</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_inhibited_by"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_inhibited_by" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_inhibits">
<span class="sig-name descname"><span class="pre">up_inhibits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_inhibits"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_inhibits" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_neighborhood">
<span class="sig-name descname"><span class="pre">up_neighborhood</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprots</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ALL'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_neighborhood"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_neighborhood" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_neighbors">
<span class="sig-name descname"><span class="pre">up_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ALL'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_neighbors" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_stimulated_by">
<span class="sig-name descname"><span class="pre">up_stimulated_by</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_stimulated_by"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_stimulated_by" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.up_stimulates">
<span class="sig-name descname"><span class="pre">up_stimulates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprot</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.up_stimulates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.up_stimulates" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_adjlist">
<span class="sig-name descname"><span class="pre">update_adjlist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ALL'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_adjlist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_adjlist" title="Permalink to this definition">§</a></dt>
<dd><p>Creates an adjacency list in a list of sets format.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_attrs">
<span class="sig-name descname"><span class="pre">update_attrs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_attrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_attrs" title="Permalink to this definition">§</a></dt>
<dd><p>Updates the node and edge attributes. Note that no data is
donwloaded, mainly updates the dictionaries of attributes
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.edgeAttrs</span></code> and
<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.vertexAttrs</span></code> containing the
attributes names and their correspoding types and initializes
such attributes in the network nodes/edges if they weren’t.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_cats">
<span class="sig-name descname"><span class="pre">update_cats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_cats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_cats" title="Permalink to this definition">§</a></dt>
<dd><p>Makes sure that the <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.has_cats</span></code>
attribute is an up to date [set] of all categories in the
current network.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_db_dict">
<span class="sig-name descname"><span class="pre">update_db_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_db_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_db_dict" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_pathway_types">
<span class="sig-name descname"><span class="pre">update_pathway_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_pathway_types"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_pathway_types" title="Permalink to this definition">§</a></dt>
<dd><p>Updates the pathway types attribute
(<code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.pathway_types</span></code>) according to the
loaded resources of the undirected network.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_pathways">
<span class="sig-name descname"><span class="pre">update_pathways</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_pathways" title="Permalink to this definition">§</a></dt>
<dd><p>Makes sure that the <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.pathways</span></code>
attribute is an up to date [dict] of all pathways and their
sources in the current network.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_sources">
<span class="sig-name descname"><span class="pre">update_sources</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_sources"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_sources" title="Permalink to this definition">§</a></dt>
<dd><p>Makes sure that the <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.sources</span></code>
attribute is an up to date [list] of all sources in the current
network.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_summaries">
<span class="sig-name descname"><span class="pre">update_summaries</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_summaries"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_summaries" title="Permalink to this definition">§</a></dt>
<dd><p>Creates a dict with many summarizing and comparative statistics
about the resources in the current network.
The result will be assigned to the attribute <code class="docutils literal notranslate"><span class="pre">summaries</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_vertex_sources">
<span class="sig-name descname"><span class="pre">update_vertex_sources</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_vertex_sources"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_vertex_sources" title="Permalink to this definition">§</a></dt>
<dd><p>Updates the all the vertex attributes <code class="docutils literal notranslate"><span class="pre">'sources'</span></code> and
<code class="docutils literal notranslate"><span class="pre">'references'</span></code> according to their related edges (on the
undirected graph).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_vindex">
<span class="sig-name descname"><span class="pre">update_vindex</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_vindex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_vindex" title="Permalink to this definition">§</a></dt>
<dd><p>This is deprecated.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.update_vname">
<span class="sig-name descname"><span class="pre">update_vname</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.update_vname"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.update_vname" title="Permalink to this definition">§</a></dt>
<dd><p>Fast lookup of node names and indexes, these are hold in a
[list] and a [dict] as well. However, every time new nodes are
added, these should be updated. This function is automatically
called after all operations affecting node indices.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.ups">
<span class="sig-name descname"><span class="pre">ups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">uniprots</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.ups" title="Permalink to this definition">§</a></dt>
<dd><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
undirected graph.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.v">
<span class="sig-name descname"><span class="pre">v</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.legacy.main.PyPath.v" title="Permalink to this definition">§</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="simple">
<dt>&#64;identifier<span class="classifier">int, str</span></dt><dd><p>Vertex index (int) or GeneSymbol (str) or UniProt ID (str) or
<code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.vertex_pathways">
<span class="sig-name descname"><span class="pre">vertex_pathways</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.vertex_pathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.vertex_pathways" title="Permalink to this definition">§</a></dt>
<dd><p>Some resources assignes interactions some others proteins to
pathways. This function copies pathway annotations from edge
attributes to vertex attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.vsgs">
<span class="sig-name descname"><span class="pre">vsgs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.vsgs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.vsgs" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a generator sequence of the node names as GeneSymbols
[str] (from the undirected graph).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>generator</em>) – Sequence containing the node names as
GeneSymbols [str].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.vsup">
<span class="sig-name descname"><span class="pre">vsup</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.vsup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.vsup" title="Permalink to this definition">§</a></dt>
<dd><p>Returns a generator sequence of the node names as UniProt IDs
[str] (from the undirected graph).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<em>generator</em>) – Sequence containing the node names as
UniProt IDs [str].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.wang_effects">
<span class="sig-name descname"><span class="pre">wang_effects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.wang_effects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.wang_effects" title="Permalink to this definition">§</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.legacy.main.PyPath.write_table">
<span class="sig-name descname"><span class="pre">write_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tbl</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\t'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colnames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rownames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/legacy/main.html#PyPath.write_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.legacy.main.PyPath.write_table" title="Permalink to this definition">§</a></dt>
<dd><p>Writes a given table to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tbl</strong> (<em>dict</em>) – Contains the data of the table. It is assumed that keys are
the row names [str] and the values, well, values. Column
names (if any) are defined with the key <code class="docutils literal notranslate"><span class="pre">'header'</span></code>.</p></li>
<li><p><strong>outfile</strong> (<em>str</em>) – File name where to save the table. The file will be saved
under the object’s <code class="xref py py-attr docutils literal notranslate"><span class="pre">pypath.main.PyPath.outdir</span></code>
(<code class="docutils literal notranslate"><span class="pre">'results'</span></code> by default).</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">'</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">'</span></code> (tab) by default. Specifies the separator
for the file.</p></li>
<li><p><strong>cut</strong> (<em>int</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Specifies the maximum number
of characters for the row names.</p></li>
<li><p><strong>colnames</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Specifies whether to write
the column names in the file or not.</p></li>
<li><p><strong>rownames</strong> (<em>bool</em>) – Optional, <code class="docutils literal notranslate"><span class="pre">True</span></code> by default. Specifies whether to write
the row names in the file or not.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="pypath.legacy.main.NetworkStatsRecord.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">pypath.legacy.main.NetworkStatsRecord</p>
      </div>
    </a>
    <a class="right-next"
       href="pypath.omnipath.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">pypath.omnipath</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../index.html">Welcome to pypath!</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../webservice.html">Webservice</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../releasehistory.html">Release history</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../reference.html">Reference</a></li>
<li class="toctree-l1 toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="../notebooks/manual.html">The pypath book</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">
  <p class="copyright">
    
      © Copyright 2023, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Melih Darcan, Erva Ulusoy, Tennur Kilic, Omer Kaan Vural, Ahmet Rifaioglu, Charles Tapley Hoyt.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">
  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 6.1.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item"><p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.13.3.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>